<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.551">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>cfRNA POP Study - 2&nbsp; RNA-seq data processing</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./cross_validation.html" rel="next">
<link href="./sample.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./sample.html">Methods</a></li><li class="breadcrumb-item"><a href="./quant.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">RNA-seq data processing</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">cfRNA POP Study</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/obsgynaecam/cell-free-rna-2024" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Methods</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./sample.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Sample Information</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./quant.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">RNA-seq data processing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cross_validation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Cross Validation</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Supplementary Text</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./SI_select_quant_method.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Selection of RNA-seq quantification method</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./SI_trimming.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Assessing the effect of trimming the adaptor sequences</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./figure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figures</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./main_figure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Main Figures</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./suppl_figure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Supplementary Text Figures</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#sec-mtd-rna-seq-quant" id="toc-sec-mtd-rna-seq-quant" class="nav-link active" data-scroll-target="#sec-mtd-rna-seq-quant"><span class="header-section-number">2.1</span> Transcript quantification</a>
  <ul class="collapse">
  <li><a href="#sec-mtd-salmon-index" id="toc-sec-mtd-salmon-index" class="nav-link" data-scroll-target="#sec-mtd-salmon-index"><span class="header-section-number">2.1.1</span> Salmon index</a></li>
  <li><a href="#sec-mtd-salmon-quant" id="toc-sec-mtd-salmon-quant" class="nav-link" data-scroll-target="#sec-mtd-salmon-quant"><span class="header-section-number">2.1.2</span> Salmon quant</a></li>
  </ul></li>
  <li><a href="#sec-mtd-deg" id="toc-sec-mtd-deg" class="nav-link" data-scroll-target="#sec-mtd-deg"><span class="header-section-number">2.2</span> Differentially expressed gene analysis</a>
  <ul class="collapse">
  <li><a href="#sec-mtd-deg-deseq2" id="toc-sec-mtd-deg-deseq2" class="nav-link" data-scroll-target="#sec-mtd-deg-deseq2"><span class="header-section-number">2.2.1</span> <code>DESeq2</code></a></li>
  <li><a href="#sec-mtd-deg-edger" id="toc-sec-mtd-deg-edger" class="nav-link" data-scroll-target="#sec-mtd-deg-edger"><span class="header-section-number">2.2.2</span> <code>edgeR</code></a></li>
  <li><a href="#sec-zscore-mat" id="toc-sec-zscore-mat" class="nav-link" data-scroll-target="#sec-zscore-mat"><span class="header-section-number">2.2.3</span> The z-score matrix</a></li>
  <li><a href="#sec-logreg" id="toc-sec-logreg" class="nav-link" data-scroll-target="#sec-logreg"><span class="header-section-number">2.2.4</span> Univariate logistic regression</a></li>
  </ul></li>
  <li><a href="#sec-core-deg" id="toc-sec-core-deg" class="nav-link" data-scroll-target="#sec-core-deg"><span class="header-section-number">2.3</span> Selection of core differentially expressed genes</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/obsgynaecam/cell-free-rna-2024/edit/main/quant.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/obsgynaecam/cell-free-rna-2024/blob/main/quant.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li><li><a href="https://github.com/obsgynaecam/cell-free-rna-2024/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./sample.html">Methods</a></li><li class="breadcrumb-item"><a href="./quant.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">RNA-seq data processing</span></a></li></ol></nav>
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-mtd-rna-seq" class="quarto-section-identifier"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">RNA-seq data processing</span></span></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="sec-mtd-rna-seq-quant" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="sec-mtd-rna-seq-quant"><span class="header-section-number">2.1</span> Transcript quantification</h2>
<p>For an RNA-seq quantification method applied in this study, we chose the most reliable approach based on the performance of predicting fetal sex by measuring the extent of chromosome Y encoded transcripts (see <a href="SI_select_quant_method.html" class="quarto-xref"><span>Chapter 4</span></a> for detail). Based on this benchmark result, we chose <a href="https://combine-lab.github.io/salmon/"><code>Salmon</code></a> (v1.5.2) in mapping-mode to process our RNA-seq datasets.</p>
<section id="sec-mtd-salmon-index" class="level3" data-number="2.1.1">
<h3 data-number="2.1.1" class="anchored" data-anchor-id="sec-mtd-salmon-index"><span class="header-section-number">2.1.1</span> Salmon index</h3>
<div id="lst-salmon-index" class="bash listing quarto-float" data-filename="static/shell/salmon.decoy.sh">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-salmon-index-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.1: A pseudo-code to run `salmon index`
</figcaption>
<div aria-describedby="lst-salmon-index-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>static/shell/salmon.decoy.sh</strong></pre>
</div>
<div class="sourceCode" id="lst-salmon-index" data-filename="static/shell/salmon.decoy.sh"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="lst-salmon-index-1"><a href="#lst-salmon-index-1"></a><span class="va">$HOME</span><span class="ex">/Install/SalmonTools/scripts/generateDecoyTranscriptome.sh</span> <span class="dt">\</span></span>
<span id="lst-salmon-index-2"><a href="#lst-salmon-index-2"></a>  <span class="at">-j</span> 32 <span class="dt">\</span></span>
<span id="lst-salmon-index-3"><a href="#lst-salmon-index-3"></a>  <span class="at">-m</span> <span class="va">$HOME</span>/Install/mashmap/mashmap <span class="dt">\</span></span>
<span id="lst-salmon-index-4"><a href="#lst-salmon-index-4"></a>  <span class="at">-g</span> <span class="va">$HOME</span>/data/genome/Homo_sapiens/Ensembl/GRCh38/Sequence/WholeGenomeFasta/genome.fa <span class="dt">\</span></span>
<span id="lst-salmon-index-5"><a href="#lst-salmon-index-5"></a>  <span class="at">-a</span> <span class="va">$HOME</span>/results/RNA-Seq/Placentome/gffcompare/POPS-2022/POPS-Placenta-Transcriptome/POPS-2022.GRCh38.88.Novel.Known.Freq.0.1.TPM.0.1.tr.reconstruction.gffread.gtf <span class="dt">\</span></span>
<span id="lst-salmon-index-6"><a href="#lst-salmon-index-6"></a>  <span class="at">-t</span> <span class="va">$HOME</span>/results/RNA-Seq/Placentome/gffcompare/POPS-2022/POPS-Placenta-Transcriptome/POPS-2022.GRCh38.88.Novel.Known.Freq.0.1.TPM.0.1.tr.reconstruction.gffread.gtf.fa <span class="dt">\</span></span>
<span id="lst-salmon-index-7"><a href="#lst-salmon-index-7"></a>  <span class="at">-o</span> <span class="va">$HOME</span>/data/Salmon/decoy/Homo_sapiens/Ensembl/GRCh38/POPS-2022.GRCh38.88.Novel.Known.Freq.0.1.TPM.0.1.tr.reconstruction</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</figure>
</div>
</section>
<section id="sec-mtd-salmon-quant" class="level3" data-number="2.1.2">
<h3 data-number="2.1.2" class="anchored" data-anchor-id="sec-mtd-salmon-quant"><span class="header-section-number">2.1.2</span> Salmon quant</h3>
<div id="lst-salmon-quant" class="bash listing quarto-float" data-filename="static/shell/salmon-quant.sh">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-salmon-quant-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.2: A pseudo-code to run `salmon quant` in mapping mode
</figcaption>
<div aria-describedby="lst-salmon-quant-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>static/shell/salmon-quant.sh</strong></pre>
</div>
<div class="sourceCode" id="lst-salmon-quant" data-filename="static/shell/salmon-quant.sh"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="lst-salmon-quant-1"><a href="#lst-salmon-quant-1"></a><span class="ex">salmon</span> quant <span class="at">-p</span> 32 <span class="dt">\</span></span>
<span id="lst-salmon-quant-2"><a href="#lst-salmon-quant-2"></a>             <span class="at">-i</span> POPS_TR_INDEX  <span class="dt">\</span></span>
<span id="lst-salmon-quant-3"><a href="#lst-salmon-quant-3"></a>             <span class="at">-l</span> A <span class="dt">\</span></span>
<span id="lst-salmon-quant-4"><a href="#lst-salmon-quant-4"></a>             <span class="at">-1</span> S1_FQ_1.fq <span class="dt">\</span></span>
<span id="lst-salmon-quant-5"><a href="#lst-salmon-quant-5"></a>             <span class="at">-2</span> S1_FQ_2.fq <span class="dt">\</span></span>
<span id="lst-salmon-quant-6"><a href="#lst-salmon-quant-6"></a>             <span class="at">--seqBias</span> <span class="dt">\</span></span>
<span id="lst-salmon-quant-7"><a href="#lst-salmon-quant-7"></a>             <span class="at">--gcBias</span> <span class="dt">\</span></span>
<span id="lst-salmon-quant-8"><a href="#lst-salmon-quant-8"></a>             <span class="at">--posBias</span> <span class="dt">\</span></span>
<span id="lst-salmon-quant-9"><a href="#lst-salmon-quant-9"></a>             <span class="at">--discardOrphanQuasi</span> <span class="dt">\</span></span>
<span id="lst-salmon-quant-10"><a href="#lst-salmon-quant-10"></a>             <span class="at">--writeUnmappedNames</span> <span class="dt">\</span></span>
<span id="lst-salmon-quant-11"><a href="#lst-salmon-quant-11"></a>             <span class="at">--writeMapping</span> <span class="dt">\</span></span>
<span id="lst-salmon-quant-12"><a href="#lst-salmon-quant-12"></a>             <span class="at">-o</span> OUTPUT_DIR <span class="kw">|</span> <span class="ex">samtools</span> view <span class="at">-bS</span> <span class="op">&gt;</span> OUTPUT_DIR/S1_salmon.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</figure>
</div>
</section>
</section>
<section id="sec-mtd-deg" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="sec-mtd-deg"><span class="header-section-number">2.2</span> Differentially expressed gene analysis</h2>
<p>The differentially expressed gene analysis was conducted for each gestational epoch (12wk, 20wk, 28wk, and 36wk) separately, except for the 36wkGA gestation samples of the pre-term dataset which has only one PE case.</p>
<p>Firstly, the following parameters were defined:</p>
<div id="lst-set-config" class="r listing quarto-float" data-filename="static/R/config/cfRNA-2024.R">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-set-config-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.3: Set the parameters
</figcaption>
<div aria-describedby="lst-set-config-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>static/R/config/cfRNA-2024.R</strong></pre>
</div>
<div class="sourceCode" id="lst-set-config" data-filename="static/R/config/cfRNA-2024.R"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="lst-set-config-1"><a href="#lst-set-config-1"></a><span class="co"># set parameters</span></span>
<span id="lst-set-config-2"><a href="#lst-set-config-2"></a>my.disease<span class="ot">=</span><span class="st">"PE"</span></span>
<span id="lst-set-config-3"><a href="#lst-set-config-3"></a>my.type<span class="ot">=</span><span class="st">"preterm"</span></span>
<span id="lst-set-config-4"><a href="#lst-set-config-4"></a>my.salmon<span class="ot">=</span><span class="st">"Salmon"</span></span>
<span id="lst-set-config-5"><a href="#lst-set-config-5"></a>my.salmon.index<span class="ot">=</span><span class="st">"POPS-2022.GRCh38.88"</span></span>
<span id="lst-set-config-6"><a href="#lst-set-config-6"></a>my.slx<span class="ot">&lt;-</span><span class="st">"SLX-ILM-Plasma2021.Homo_sapiens.v1"</span></span>
<span id="lst-set-config-7"><a href="#lst-set-config-7"></a></span>
<span id="lst-set-config-8"><a href="#lst-set-config-8"></a><span class="co"># set filter</span></span>
<span id="lst-set-config-9"><a href="#lst-set-config-9"></a>minCPM<span class="ot">=</span><span class="fl">0.1</span>; minRead<span class="ot">=</span><span class="dv">10</span>; minFreq<span class="ot">=</span><span class="fl">0.1</span>; minFC<span class="ot">=</span><span class="fl">1.2</span></span>
<span id="lst-set-config-10"><a href="#lst-set-config-10"></a></span>
<span id="lst-set-config-11"><a href="#lst-set-config-11"></a><span class="co"># which p.adjust methods?</span></span>
<span id="lst-set-config-12"><a href="#lst-set-config-12"></a>adjust.methods<span class="ot">=</span><span class="fu">c</span>(<span class="st">`</span><span class="at">Benjamini &amp; Yekutieli</span><span class="st">`</span><span class="ot">=</span><span class="st">"BY"</span>,</span>
<span id="lst-set-config-13"><a href="#lst-set-config-13"></a>            <span class="st">`</span><span class="at">Benjamini &amp; Hochberg</span><span class="st">`</span><span class="ot">=</span><span class="st">"BH"</span>,</span>
<span id="lst-set-config-14"><a href="#lst-set-config-14"></a>            <span class="st">`</span><span class="at">Bonferroni</span><span class="st">`</span><span class="ot">=</span><span class="st">"bonferroni"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</figure>
</div>
<p>The transcript-level read count matrices (e.g.&nbsp;quant.sf files) were imported using <a href="https://bioconductor.org/packages/release/bioc/html/tximeta.html"><code>tximeta</code></a> (v1.8.5) Bioconductor package and merged at the gene-level.</p>
<div class="cell">
<div id="lst-prep-quant-sf" class="r cell-code listing quarto-float" data-code-summary="Code to set the `Salmon` quant files of the pre-term dataset (i.e. Discovery dataset)">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-prep-quant-sf-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.4: Code to set the <code>Salmon</code> quant files of the pre-term dataset (i.e.&nbsp;Discovery dataset)
</figcaption>
<div aria-describedby="lst-prep-quant-sf-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sourceCode cell-code" id="lst-prep-quant-sf" data-code-summary="Code to set the `Salmon` quant files of the pre-term dataset (i.e. Discovery dataset)"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="lst-prep-quant-sf-1"><a href="#lst-prep-quant-sf-1" aria-hidden="true" tabindex="-1"></a>dt.foo<span class="ot">&lt;-</span><span class="fu">data.table</span>(<span class="st">`</span><span class="at">files</span><span class="st">`</span><span class="ot">=</span><span class="fu">system</span>(<span class="fu">paste0</span>(<span class="st">"ls "</span>, <span class="st">"~/results/"</span>,my.slx,<span class="st">"/"</span>,my.salmon,<span class="st">"/"</span>,my.salmon.index,<span class="st">"/*/quant.sf"</span>), <span class="at">intern=</span>T))</span>
<span id="lst-prep-quant-sf-2"><a href="#lst-prep-quant-sf-2" aria-hidden="true" tabindex="-1"></a>dt.foo[,names<span class="sc">:</span><span class="er">=</span><span class="fu">tstrsplit</span>(files,<span class="st">"/"</span>,<span class="at">keep=</span><span class="dv">8</span>)]</span>
<span id="lst-prep-quant-sf-3"><a href="#lst-prep-quant-sf-3" aria-hidden="true" tabindex="-1"></a>dt.colDataAll<span class="ot">&lt;-</span><span class="fu">merge</span>(dt.foo,dt.samples,<span class="at">by.x=</span><span class="st">"names"</span>,<span class="at">by.y=</span><span class="st">"SampleID"</span>) <span class="co"># n=755</span></span>
<span id="lst-prep-quant-sf-4"><a href="#lst-prep-quant-sf-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-prep-quant-sf-5"><a href="#lst-prep-quant-sf-5" aria-hidden="true" tabindex="-1"></a><span class="co"># pre-term</span></span>
<span id="lst-prep-quant-sf-6"><a href="#lst-prep-quant-sf-6" aria-hidden="true" tabindex="-1"></a>dt.colData<span class="ot">&lt;-</span>dt.colDataAll[Type<span class="sc">==</span>my.type] <span class="co"># n=279</span></span>
<span id="lst-prep-quant-sf-7"><a href="#lst-prep-quant-sf-7" aria-hidden="true" tabindex="-1"></a><span class="co">#dt.colData[grepl("-b$",names)] # CX (CX-b): both failed; HQ (HQ-b): only HQ-b passed QC according to illumina</span></span>
<span id="lst-prep-quant-sf-8"><a href="#lst-prep-quant-sf-8" aria-hidden="true" tabindex="-1"></a>dt.colData<span class="ot">&lt;-</span>dt.colData[<span class="sc">!</span>names <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"GS-59-CX-b"</span>,<span class="st">"GS-179-HQ"</span>)] <span class="co"># remove these two samples (n=277)</span></span>
<span id="lst-prep-quant-sf-9"><a href="#lst-prep-quant-sf-9" aria-hidden="true" tabindex="-1"></a>li.GA<span class="ot">&lt;-</span><span class="fu">split</span>(dt.colData, dt.colData<span class="sc">$</span>GA)</span>
<span id="lst-prep-quant-sf-10"><a href="#lst-prep-quant-sf-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-prep-quant-sf-11"><a href="#lst-prep-quant-sf-11" aria-hidden="true" tabindex="-1"></a><span class="co">#tx2gene</span></span>
<span id="lst-prep-quant-sf-12"><a href="#lst-prep-quant-sf-12" aria-hidden="true" tabindex="-1"></a>dt.tx2gene<span class="ot">&lt;-</span><span class="fu">fread</span>(<span class="st">"~/results/RNA-Seq/Placentome/gffcompare/POPS-2022/POPS-Placenta-Transcriptome/POPS-2022.GRCh38.88.Novel.Known.Freq.0.1.TPM.0.1.tr.reconstruction.tx2gene.txt"</span>, <span class="at">header=</span>F,<span class="at">col.names=</span><span class="fu">c</span>(<span class="st">"transcript_id"</span>,<span class="st">"gene_id"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</figure>
</div>
</div>
<div class="cell">
<div id="lst-prep-dds" class="r cell-code listing quarto-float" data-code-summary="Code to make `dds` (DESeq Data Set) object">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-prep-dds-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.5: Code to make <code>dds</code> (DESeq Data Set) object
</figcaption>
<div aria-describedby="lst-prep-dds-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sourceCode cell-code" id="lst-prep-dds" data-code-summary="Code to make `dds` (DESeq Data Set) object"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="lst-prep-dds-1"><a href="#lst-prep-dds-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="lst-prep-dds-2"><a href="#lst-prep-dds-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(edgeR)</span>
<span id="lst-prep-dds-3"><a href="#lst-prep-dds-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"BiocParallel"</span>)</span>
<span id="lst-prep-dds-4"><a href="#lst-prep-dds-4" aria-hidden="true" tabindex="-1"></a><span class="fu">register</span>(<span class="fu">MulticoreParam</span>(<span class="dv">12</span>))</span>
<span id="lst-prep-dds-5"><a href="#lst-prep-dds-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-prep-dds-6"><a href="#lst-prep-dds-6" aria-hidden="true" tabindex="-1"></a><span class="co">#tximeta via tx2gene</span></span>
<span id="lst-prep-dds-7"><a href="#lst-prep-dds-7" aria-hidden="true" tabindex="-1"></a><span class="co"># 36wk dataset excluded from the beginnin</span></span>
<span id="lst-prep-dds-8"><a href="#lst-prep-dds-8" aria-hidden="true" tabindex="-1"></a>gse<span class="ot">&lt;-</span>tximeta<span class="sc">::</span><span class="fu">tximeta</span>(dt.colData[GA<span class="sc">!=</span><span class="st">"36wk"</span>],</span>
<span id="lst-prep-dds-9"><a href="#lst-prep-dds-9" aria-hidden="true" tabindex="-1"></a>                      <span class="at">skipMeta=</span>T,</span>
<span id="lst-prep-dds-10"><a href="#lst-prep-dds-10" aria-hidden="true" tabindex="-1"></a>                      <span class="at">tx2gene=</span>dt.tx2gene[,.(transcript_id,gene_id)],<span class="at">txOut=</span>F)</span>
<span id="lst-prep-dds-11"><a href="#lst-prep-dds-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-prep-dds-12"><a href="#lst-prep-dds-12" aria-hidden="true" tabindex="-1"></a><span class="co"># set up `dds` at gene-level</span></span>
<span id="lst-prep-dds-13"><a href="#lst-prep-dds-13" aria-hidden="true" tabindex="-1"></a>my.design <span class="ot">&lt;-</span> <span class="fu">formula</span>(<span class="sc">~</span> Batch <span class="sc">+</span> GA <span class="sc">+</span> Sex <span class="sc">+</span> Condition)    <span class="co"># isa 'formula'</span></span>
<span id="lst-prep-dds-14"><a href="#lst-prep-dds-14" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSet</span>(<span class="at">se=</span>gse, <span class="at">design=</span>my.design)</span>
<span id="lst-prep-dds-15"><a href="#lst-prep-dds-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-prep-dds-16"><a href="#lst-prep-dds-16" aria-hidden="true" tabindex="-1"></a>dds<span class="sc">$</span>Group<span class="ot">&lt;-</span><span class="fu">factor</span>(<span class="fu">paste0</span>(dds<span class="sc">$</span>GA,dds<span class="sc">$</span>Condition)) <span class="co"># add 'Group'</span></span>
<span id="lst-prep-dds-17"><a href="#lst-prep-dds-17" aria-hidden="true" tabindex="-1"></a><span class="fu">design</span>(dds) <span class="ot">&lt;-</span> <span class="fu">formula</span>(<span class="sc">~</span> Batch <span class="sc">+</span> Sex <span class="sc">+</span> Group)     <span class="co"># isa 'formula'</span></span>
<span id="lst-prep-dds-18"><a href="#lst-prep-dds-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-prep-dds-19"><a href="#lst-prep-dds-19" aria-hidden="true" tabindex="-1"></a>dds<span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds, <span class="at">parallel=</span><span class="cn">TRUE</span>) <span class="co"># isa 'DESeqDataSet'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</figure>
</div>
</div>
<section id="sec-mtd-deg-deseq2" class="level3" data-number="2.2.1">
<h3 data-number="2.2.1" class="anchored" data-anchor-id="sec-mtd-deg-deseq2"><span class="header-section-number">2.2.1</span> <code>DESeq2</code></h3>
<p>We only considered genes found in ≥10% of samples (i.e.&nbsp;≥ 23) having ≥10 reads, and discarded genes detected as dispersion outliers by <a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html"><code>DESeq2</code></a> (v1.30.0).</p>
<div class="cell">
<div id="lst-set-dds" class="r cell-code listing quarto-float" data-code-summary="Code to set up `dds` object">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-set-dds-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.6: Code to set up <code>dds</code> object
</figcaption>
<div aria-describedby="lst-set-dds-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sourceCode cell-code" id="lst-set-dds" data-code-summary="Code to set up `dds` object"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="lst-set-dds-1"><a href="#lst-set-dds-1" aria-hidden="true" tabindex="-1"></a><span class="co"># at leat 10 reads for at leat 10  % of the samples </span></span>
<span id="lst-set-dds-2"><a href="#lst-set-dds-2" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">counts</span>(dds) <span class="sc">&gt;=</span> minRead) <span class="sc">&gt;=</span> <span class="fu">ncol</span>(dds)<span class="sc">*</span>minFreq </span>
<span id="lst-set-dds-3"><a href="#lst-set-dds-3" aria-hidden="true" tabindex="-1"></a>dds.f<span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds[keep,], <span class="at">parallel=</span><span class="cn">TRUE</span>) <span class="co"># isa 'DESeqDataSet'</span></span>
<span id="lst-set-dds-4"><a href="#lst-set-dds-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dds.f) <span class="co"># 15380 x 221</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</figure>
</div>
</div>
<p>A total of 15,150 genes and 221 samples were used to find differentially expressed genes by adjusting the batch number, the fetal sex, and the gestion of the samples in the design matrix of DESeq2. The p-values were calculated from the null hypothesis that the fold changes were less than or equal to 20% (i.e.&nbsp;lfcThreshold=log2(1.2)) in case and control groups.</p>
<div class="cell">
<div id="lst-set-deg" class="r cell-code listing quarto-float" data-code-summary="Code to find DEGs by DESeq">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-set-deg-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.7: Code to find DEGs by DESeq
</figcaption>
<div aria-describedby="lst-set-deg-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sourceCode cell-code" id="lst-set-deg" data-code-summary="Code to find DEGs by DESeq"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="lst-set-deg-1"><a href="#lst-set-deg-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove dispOutlier genes</span></span>
<span id="lst-set-deg-2"><a href="#lst-set-deg-2" aria-hidden="true" tabindex="-1"></a>keep2<span class="ot">&lt;-</span><span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">rowData</span>(dds.f)[,<span class="st">"dispOutlier"</span>]) <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">rowData</span>(dds.f)[,<span class="st">"dispOutlier"</span>]</span>
<span id="lst-set-deg-3"><a href="#lst-set-deg-3" aria-hidden="true" tabindex="-1"></a>dds.f2<span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds.f[keep2], <span class="at">parallel=</span>T)</span>
<span id="lst-set-deg-4"><a href="#lst-set-deg-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-set-deg-5"><a href="#lst-set-deg-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dds.f2) <span class="co"># 15150 x 221</span></span>
<span id="lst-set-deg-6"><a href="#lst-set-deg-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-set-deg-7"><a href="#lst-set-deg-7" aria-hidden="true" tabindex="-1"></a><span class="co"># apply shink separately for 12wk, 20wk, and 28wk (li.GA[1:3])</span></span>
<span id="lst-set-deg-8"><a href="#lst-set-deg-8" aria-hidden="true" tabindex="-1"></a>li.resLFC<span class="ot">&lt;-</span><span class="fu">lapply</span>(<span class="fu">names</span>(li.GA)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>],<span class="cf">function</span>(my.GA){ </span>
<span id="lst-set-deg-9"><a href="#lst-set-deg-9" aria-hidden="true" tabindex="-1"></a>  my.res<span class="ot">&lt;-</span><span class="fu">results</span>(dds.f2, </span>
<span id="lst-set-deg-10"><a href="#lst-set-deg-10" aria-hidden="true" tabindex="-1"></a>          <span class="at">independentFiltering=</span><span class="cn">FALSE</span>, <span class="co">#by default independant filtering at FDR (alpha) 0.1 (10%) </span></span>
<span id="lst-set-deg-11"><a href="#lst-set-deg-11" aria-hidden="true" tabindex="-1"></a>          <span class="at">lfcThreshold=</span><span class="fu">log2</span>(minFC), </span>
<span id="lst-set-deg-12"><a href="#lst-set-deg-12" aria-hidden="true" tabindex="-1"></a>          <span class="at">contrast=</span><span class="fu">c</span>(<span class="st">"Group"</span>,<span class="fu">paste0</span>(my.GA,<span class="fu">c</span>(<span class="st">"Case"</span>,<span class="st">"Control"</span>))),</span>
<span id="lst-set-deg-13"><a href="#lst-set-deg-13" aria-hidden="true" tabindex="-1"></a>          <span class="at">parallel=</span><span class="cn">TRUE</span>)  </span>
<span id="lst-set-deg-14"><a href="#lst-set-deg-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lfcShrink</span>(dds.f2, </span>
<span id="lst-set-deg-15"><a href="#lst-set-deg-15" aria-hidden="true" tabindex="-1"></a>            <span class="at">res=</span>my.res,</span>
<span id="lst-set-deg-16"><a href="#lst-set-deg-16" aria-hidden="true" tabindex="-1"></a>            <span class="at">lfcThreshold=</span><span class="fu">log2</span>(minFC), <span class="co"># not applicable for 'asher' </span></span>
<span id="lst-set-deg-17"><a href="#lst-set-deg-17" aria-hidden="true" tabindex="-1"></a>            <span class="at">type=</span><span class="st">"ashr"</span>,</span>
<span id="lst-set-deg-18"><a href="#lst-set-deg-18" aria-hidden="true" tabindex="-1"></a>            <span class="at">parallel=</span><span class="cn">TRUE</span>)  </span>
<span id="lst-set-deg-19"><a href="#lst-set-deg-19" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="lst-set-deg-20"><a href="#lst-set-deg-20" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(li.resLFC)<span class="ot">&lt;-</span><span class="fu">names</span>(li.GA)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="lst-set-deg-21"><a href="#lst-set-deg-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-set-deg-22"><a href="#lst-set-deg-22" aria-hidden="true" tabindex="-1"></a>dl.resLFC<span class="ot">&lt;-</span><span class="fu">lapply</span>(li.resLFC, <span class="cf">function</span>(i)</span>
<span id="lst-set-deg-23"><a href="#lst-set-deg-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>(<span class="st">`</span><span class="at">gene_id</span><span class="st">`</span><span class="ot">=</span><span class="fu">rownames</span>(i), </span>
<span id="lst-set-deg-24"><a href="#lst-set-deg-24" aria-hidden="true" tabindex="-1"></a>              <span class="fu">as.data.frame</span>(i))[<span class="fu">order</span>(pvalue)][,<span class="st">`</span><span class="at">:=</span><span class="st">`</span>(<span class="st">"BH"</span><span class="ot">=</span><span class="fu">p.adjust</span>(pvalue,<span class="st">"BH"</span>),</span>
<span id="lst-set-deg-25"><a href="#lst-set-deg-25" aria-hidden="true" tabindex="-1"></a>                                              <span class="st">"BY"</span><span class="ot">=</span><span class="fu">p.adjust</span>(pvalue,<span class="st">"BY"</span>),</span>
<span id="lst-set-deg-26"><a href="#lst-set-deg-26" aria-hidden="true" tabindex="-1"></a>                                              <span class="st">"bf"</span><span class="ot">=</span><span class="fu">p.adjust</span>(pvalue,<span class="st">"bonferroni"</span>))]</span>
<span id="lst-set-deg-27"><a href="#lst-set-deg-27" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="lst-set-deg-28"><a href="#lst-set-deg-28" aria-hidden="true" tabindex="-1"></a><span class="fu">fwrite</span>(dl.resLFC[[<span class="st">"28wk"</span>]], <span class="at">file=</span><span class="fu">paste0</span>(<span class="st">"static/R/result/DEG.DSeq2.28wk."</span>,my.type,<span class="st">"."</span>,my.salmon.index,<span class="st">".csv"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</figure>
</div>
</div>
</section>
<section id="sec-mtd-deg-edger" class="level3" data-number="2.2.2">
<h3 data-number="2.2.2" class="anchored" data-anchor-id="sec-mtd-deg-edger"><span class="header-section-number">2.2.2</span> <code>edgeR</code></h3>
<p>For <a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a> (v3.32.1) analysis, we used <code>makeDGEList</code> function of <code>tximeta</code> Bioconductor package to convert the data object of the 15,150 genes across the 221 samples as mentioned above. The gene-level count matrix was normalised by using <code>calcNormFactors</code> function of <code>edgeR</code> with <code>TMM</code> method (trimmed mean of M values) and a quasi-likelihood negative binomial generalised log-linear model (i.e.&nbsp;<code>glmQLFit</code>) was applied to account for the batch number, the fetal sex and the gestation information in the design matrix of edgeR. For a statistical test, we used <code>glmTreat</code> of edgeR with at least 20% fold-change (i.e.&nbsp;lfc=log2(1.2)).</p>
<div class="cell">
<div id="lst-set-edger" class="r cell-code listing quarto-float" data-code-summary="Code to find DEGs by `edgeR`">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-set-edger-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.8: Code to find DEGs by <code>edgeR</code>
</figcaption>
<div aria-describedby="lst-set-edger-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sourceCode cell-code" id="lst-set-edger" data-code-summary="Code to find DEGs by `edgeR`"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="lst-set-edger-1"><a href="#lst-set-edger-1" aria-hidden="true" tabindex="-1"></a>gse2 <span class="ot">&lt;-</span>gse[<span class="fu">rownames</span>(dds.f2),]</span>
<span id="lst-set-edger-2"><a href="#lst-set-edger-2" aria-hidden="true" tabindex="-1"></a>d2<span class="ot">&lt;-</span>tximeta<span class="sc">::</span><span class="fu">makeDGEList</span>(gse2)</span>
<span id="lst-set-edger-3"><a href="#lst-set-edger-3" aria-hidden="true" tabindex="-1"></a>d2<span class="sc">$</span>samples<span class="ot">&lt;-</span><span class="fu">cbind</span>(d2<span class="sc">$</span>samples,<span class="fu">colData</span>(gse2))</span>
<span id="lst-set-edger-4"><a href="#lst-set-edger-4" aria-hidden="true" tabindex="-1"></a>d2<span class="sc">$</span>samples<span class="sc">$</span>group<span class="ot">&lt;-</span><span class="fu">factor</span>(<span class="fu">paste0</span>(d2<span class="sc">$</span>samples<span class="sc">$</span>GA,d2<span class="sc">$</span>samples<span class="sc">$</span>Condition)) <span class="co"># add 'group'</span></span>
<span id="lst-set-edger-5"><a href="#lst-set-edger-5" aria-hidden="true" tabindex="-1"></a>d2<span class="sc">$</span>samples<span class="sc">$</span>GA<span class="ot">&lt;-</span><span class="fu">droplevels</span>(d2<span class="sc">$</span>samples<span class="sc">$</span>GA) <span class="co"># 36wk removed</span></span>
<span id="lst-set-edger-6"><a href="#lst-set-edger-6" aria-hidden="true" tabindex="-1"></a>d2<span class="sc">$</span>samples<span class="sc">$</span>Batch<span class="ot">&lt;-</span><span class="fu">droplevels</span>(d2<span class="sc">$</span>samples<span class="sc">$</span>Batch) <span class="co"># some batches removed</span></span>
<span id="lst-set-edger-7"><a href="#lst-set-edger-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-set-edger-8"><a href="#lst-set-edger-8" aria-hidden="true" tabindex="-1"></a><span class="co"># additional filter</span></span>
<span id="lst-set-edger-9"><a href="#lst-set-edger-9" aria-hidden="true" tabindex="-1"></a><span class="co">#keep &lt;- filterByExpr(d2)</span></span>
<span id="lst-set-edger-10"><a href="#lst-set-edger-10" aria-hidden="true" tabindex="-1"></a><span class="co">#keep %&gt;% table # FALSE:13, TRUE:15137</span></span>
<span id="lst-set-edger-11"><a href="#lst-set-edger-11" aria-hidden="true" tabindex="-1"></a><span class="co">#d2&lt;-d2[keep,]</span></span>
<span id="lst-set-edger-12"><a href="#lst-set-edger-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-set-edger-13"><a href="#lst-set-edger-13" aria-hidden="true" tabindex="-1"></a><span class="co"># TMM normalisation (default). It adds `norm.factors` d2$samples</span></span>
<span id="lst-set-edger-14"><a href="#lst-set-edger-14" aria-hidden="true" tabindex="-1"></a><span class="co"># NB, we have `offsets`, which take precedence over lib.size and norm.factors</span></span>
<span id="lst-set-edger-15"><a href="#lst-set-edger-15" aria-hidden="true" tabindex="-1"></a>d2<span class="ot">&lt;-</span><span class="fu">calcNormFactors</span>(d2,<span class="at">method=</span><span class="st">"TMM"</span>) </span>
<span id="lst-set-edger-16"><a href="#lst-set-edger-16" aria-hidden="true" tabindex="-1"></a>                                    </span>
<span id="lst-set-edger-17"><a href="#lst-set-edger-17" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(d2, <span class="at">col=</span><span class="fu">as.numeric</span>(d2<span class="sc">$</span>samples<span class="sc">$</span>GA))</span>
<span id="lst-set-edger-18"><a href="#lst-set-edger-18" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(d2, <span class="at">col=</span><span class="fu">as.numeric</span>(d2<span class="sc">$</span>samples<span class="sc">$</span>Condition))</span>
<span id="lst-set-edger-19"><a href="#lst-set-edger-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-set-edger-20"><a href="#lst-set-edger-20" aria-hidden="true" tabindex="-1"></a><span class="co"># design</span></span>
<span id="lst-set-edger-21"><a href="#lst-set-edger-21" aria-hidden="true" tabindex="-1"></a>my.design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> <span class="dv">0</span> <span class="sc">+</span> group <span class="sc">+</span> Sex <span class="sc">+</span> Batch , <span class="at">data=</span>d2<span class="sc">$</span>samples)      <span class="co"># isa 'matrix'</span></span>
<span id="lst-set-edger-22"><a href="#lst-set-edger-22" aria-hidden="true" tabindex="-1"></a>(my.contrasts <span class="ot">&lt;-</span> <span class="fu">makeContrasts</span>(<span class="st">`</span><span class="at">12wk</span><span class="st">`</span><span class="ot">=</span>group12wkCase<span class="sc">-</span>group12wkControl, </span>
<span id="lst-set-edger-23"><a href="#lst-set-edger-23" aria-hidden="true" tabindex="-1"></a>                              <span class="st">`</span><span class="at">20wk</span><span class="st">`</span><span class="ot">=</span>group20wkCase<span class="sc">-</span>group20wkControl, </span>
<span id="lst-set-edger-24"><a href="#lst-set-edger-24" aria-hidden="true" tabindex="-1"></a>                              <span class="st">`</span><span class="at">28wk</span><span class="st">`</span><span class="ot">=</span>group28wkCase<span class="sc">-</span>group28wkControl, </span>
<span id="lst-set-edger-25"><a href="#lst-set-edger-25" aria-hidden="true" tabindex="-1"></a>                              <span class="at">levels=</span>my.design)</span>
<span id="lst-set-edger-26"><a href="#lst-set-edger-26" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="lst-set-edger-27"><a href="#lst-set-edger-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-set-edger-28"><a href="#lst-set-edger-28" aria-hidden="true" tabindex="-1"></a><span class="co"># dispersion</span></span>
<span id="lst-set-edger-29"><a href="#lst-set-edger-29" aria-hidden="true" tabindex="-1"></a><span class="co">#dp2 = estimateDisp(d2, design=my.design, verbose=T)</span></span>
<span id="lst-set-edger-30"><a href="#lst-set-edger-30" aria-hidden="true" tabindex="-1"></a>dp2 <span class="ot">=</span> <span class="fu">estimateGLMCommonDisp</span>(d2, <span class="at">design=</span>my.design, <span class="at">verbose=</span>T) </span>
<span id="lst-set-edger-31"><a href="#lst-set-edger-31" aria-hidden="true" tabindex="-1"></a>dp2 <span class="ot">=</span> <span class="fu">estimateGLMTrendedDisp</span>(dp2, <span class="at">design=</span>my.design, <span class="at">verbose=</span>T)</span>
<span id="lst-set-edger-32"><a href="#lst-set-edger-32" aria-hidden="true" tabindex="-1"></a>dp2 <span class="ot">=</span> <span class="fu">estimateGLMTagwiseDisp</span>(dp2, <span class="at">design=</span>my.design)</span>
<span id="lst-set-edger-33"><a href="#lst-set-edger-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-set-edger-34"><a href="#lst-set-edger-34" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(dp2)</span>
<span id="lst-set-edger-35"><a href="#lst-set-edger-35" aria-hidden="true" tabindex="-1"></a><span class="fu">plotBCV</span>(dp2)</span>
<span id="lst-set-edger-36"><a href="#lst-set-edger-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-set-edger-37"><a href="#lst-set-edger-37" aria-hidden="true" tabindex="-1"></a><span class="co"># fit</span></span>
<span id="lst-set-edger-38"><a href="#lst-set-edger-38" aria-hidden="true" tabindex="-1"></a><span class="co">#f = glmFit(dp2, design=my.design) # </span></span>
<span id="lst-set-edger-39"><a href="#lst-set-edger-39" aria-hidden="true" tabindex="-1"></a>f <span class="ot">=</span> <span class="fu">glmQLFit</span>(dp2, <span class="at">design=</span>my.design) <span class="co"># QL(Quasi-like) pipeline</span></span>
<span id="lst-set-edger-40"><a href="#lst-set-edger-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-set-edger-41"><a href="#lst-set-edger-41" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMD</span>(f)</span>
<span id="lst-set-edger-42"><a href="#lst-set-edger-42" aria-hidden="true" tabindex="-1"></a><span class="fu">plotQLDisp</span>(f)</span>
<span id="lst-set-edger-43"><a href="#lst-set-edger-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-set-edger-44"><a href="#lst-set-edger-44" aria-hidden="true" tabindex="-1"></a><span class="co"># get the edgeR results</span></span>
<span id="lst-set-edger-45"><a href="#lst-set-edger-45" aria-hidden="true" tabindex="-1"></a>li.res.edgeR<span class="ot">&lt;-</span><span class="fu">lapply</span>(<span class="fu">names</span>(li.GA)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], <span class="cf">function</span>(i){</span>
<span id="lst-set-edger-46"><a href="#lst-set-edger-46" aria-hidden="true" tabindex="-1"></a>  te <span class="ot">&lt;-</span> <span class="fu">glmTreat</span>(f, <span class="at">contrast=</span>my.contrasts[,i], <span class="at">lfc=</span><span class="fu">log2</span>(minFC))</span>
<span id="lst-set-edger-47"><a href="#lst-set-edger-47" aria-hidden="true" tabindex="-1"></a>  <span class="fu">topTags</span>(te, <span class="at">n=</span><span class="fu">nrow</span>(te))    <span class="co"># default sort by pvalue</span></span>
<span id="lst-set-edger-48"><a href="#lst-set-edger-48" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="lst-set-edger-49"><a href="#lst-set-edger-49" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(li.res.edgeR) <span class="ot">&lt;-</span><span class="fu">names</span>(li.GA)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>] </span>
<span id="lst-set-edger-50"><a href="#lst-set-edger-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-set-edger-51"><a href="#lst-set-edger-51" aria-hidden="true" tabindex="-1"></a>dl.res.edgeR<span class="ot">&lt;-</span><span class="fu">lapply</span>(li.res.edgeR, <span class="cf">function</span>(i)</span>
<span id="lst-set-edger-52"><a href="#lst-set-edger-52" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>(<span class="st">`</span><span class="at">gene_id</span><span class="st">`</span><span class="ot">=</span><span class="fu">rownames</span>(i),i<span class="sc">$</span>table)[<span class="fu">order</span>(PValue)][,<span class="st">`</span><span class="at">:=</span><span class="st">`</span>(<span class="st">"BH"</span><span class="ot">=</span><span class="fu">p.adjust</span>(PValue,<span class="st">"BH"</span>),</span>
<span id="lst-set-edger-53"><a href="#lst-set-edger-53" aria-hidden="true" tabindex="-1"></a>                                                                  <span class="st">"BY"</span><span class="ot">=</span><span class="fu">p.adjust</span>(PValue,<span class="st">"BY"</span>),</span>
<span id="lst-set-edger-54"><a href="#lst-set-edger-54" aria-hidden="true" tabindex="-1"></a>                                                                  <span class="st">"bf"</span><span class="ot">=</span><span class="fu">p.adjust</span>(PValue,<span class="st">"bonferroni"</span>))]</span>
<span id="lst-set-edger-55"><a href="#lst-set-edger-55" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="lst-set-edger-56"><a href="#lst-set-edger-56" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(dl.res.edgeR, <span class="cf">function</span>(i) i[FDR<span class="sc">&lt;</span><span class="fl">0.05</span>] <span class="sc">%&gt;%</span> nrow)</span>
<span id="lst-set-edger-57"><a href="#lst-set-edger-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-set-edger-58"><a href="#lst-set-edger-58" aria-hidden="true" tabindex="-1"></a><span class="fu">fwrite</span>(dl.res.edgeR[[<span class="st">"28wk"</span>]], <span class="at">file=</span><span class="fu">paste0</span>(<span class="st">"static/R/result/DEG.edgeR.28wk."</span>,my.type,<span class="st">"."</span>,my.salmon.index,<span class="st">".csv"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</figure>
</div>
</div>
</section>
<section id="sec-zscore-mat" class="level3" data-number="2.2.3">
<h3 data-number="2.2.3" class="anchored" data-anchor-id="sec-zscore-mat"><span class="header-section-number">2.2.3</span> The z-score matrix</h3>
<p>The gene-level count matrix was converted as the unit of CPM (Count Per Million), in log2-scale, via the “cpm” function of edgeR and it was further transformed into a matrix of the z-score using the mean and standard deviation of logCPM from the control samples of each corresponding gestational age group.</p>
<div class="cell">
<div id="lst-cnt-cpm" class="r cell-code listing quarto-float" data-code-summary="Code to make the z-score matrix from the discovery dataset">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-cnt-cpm-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.9: Code to make the z-score matrix from the discovery dataset
</figcaption>
<div aria-describedby="lst-cnt-cpm-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sourceCode cell-code" id="lst-cnt-cpm" data-code-summary="Code to make the z-score matrix from the discovery dataset"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="lst-cnt-cpm-1"><a href="#lst-cnt-cpm-1" aria-hidden="true" tabindex="-1"></a><span class="co"># CPM based on edgeR TMM  (NB, d isa "DGEList")</span></span>
<span id="lst-cnt-cpm-2"><a href="#lst-cnt-cpm-2" aria-hidden="true" tabindex="-1"></a>dt.logcpm2<span class="ot">&lt;-</span><span class="fu">merge</span>(</span>
<span id="lst-cnt-cpm-3"><a href="#lst-cnt-cpm-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>(<span class="st">`</span><span class="at">geneName</span><span class="st">`</span><span class="ot">=</span><span class="fu">rownames</span>(d2),<span class="fu">cpm</span>(d2,<span class="at">log=</span>T)) <span class="sc">%&gt;%</span> </span>
<span id="lst-cnt-cpm-4"><a href="#lst-cnt-cpm-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">melt.data.table</span>(<span class="at">id.vars=</span><span class="fu">c</span>(<span class="st">"geneName"</span>),<span class="at">variable.name=</span><span class="st">"SampleID"</span>,<span class="at">value.name=</span><span class="st">"logCPM"</span>),</span>
<span id="lst-cnt-cpm-5"><a href="#lst-cnt-cpm-5" aria-hidden="true" tabindex="-1"></a>  dt.samples[Type<span class="sc">==</span>my.type,.(SampleID,GA,Condition,<span class="at">Group=</span><span class="fu">paste</span>(GA,Condition,<span class="at">sep=</span><span class="st">"-"</span>))]</span>
<span id="lst-cnt-cpm-6"><a href="#lst-cnt-cpm-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="lst-cnt-cpm-7"><a href="#lst-cnt-cpm-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-cnt-cpm-8"><a href="#lst-cnt-cpm-8" aria-hidden="true" tabindex="-1"></a>dt.cpmZ<span class="ot">=</span><span class="fu">merge</span>(dt.logcpm2,</span>
<span id="lst-cnt-cpm-9"><a href="#lst-cnt-cpm-9" aria-hidden="true" tabindex="-1"></a>              dt.logcpm2[Condition<span class="sc">==</span><span class="st">"Control"</span>,.(<span class="at">Mean=</span><span class="fu">mean</span>(logCPM),<span class="at">SD=</span><span class="fu">sd</span>(logCPM)),.(GA,geneName)]</span>
<span id="lst-cnt-cpm-10"><a href="#lst-cnt-cpm-10" aria-hidden="true" tabindex="-1"></a>      ,<span class="at">by=</span><span class="fu">c</span>(<span class="st">"GA"</span>,<span class="st">"geneName"</span>)</span>
<span id="lst-cnt-cpm-11"><a href="#lst-cnt-cpm-11" aria-hidden="true" tabindex="-1"></a>      )[,.(Group,GA,Condition,SampleID,geneName,logCPM,<span class="at">logCPMZ=</span>(logCPM<span class="sc">-</span>Mean)<span class="sc">/</span>SD)]</span>
<span id="lst-cnt-cpm-12"><a href="#lst-cnt-cpm-12" aria-hidden="true" tabindex="-1"></a><span class="co">#save(dt.cpmZ,file=my.cpmZ.RData) # cache to use it later</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</figure>
</div>
</div>
</section>
<section id="sec-logreg" class="level3" data-number="2.2.4">
<h3 data-number="2.2.4" class="anchored" data-anchor-id="sec-logreg"><span class="header-section-number">2.2.4</span> Univariate logistic regression</h3>
<p>Using the binary outcomes, (i.e.&nbsp;case and control status) as dependant variables and the z-scores as independent variables, we applied a generalised linear model for each of the 15,150 genes using the <code>glm</code> function of R <code>stat</code> package (v4.0.3). Both the <a href="https://en.wikipedia.org/wiki/Akaike_information_criterion">Akaike information criterion (AIC)</a> and the <a href="https://en.wikipedia.org/wiki/Bayesian_information_criterion">Bayesian information criterion (BIC)</a> were obtained from the corresponding univariable model and the area under the <a href="https://en.wikipedia.org/wiki/Receiver_operating_characteristic">ROC curve</a> (AUC) was calculated using the <a href="https://cran.r-project.org/web/packages/pROC/"><code>pROC</code></a> package (v1.17.0.1). The p-values were calculated against the null hypothesis that is the odds ratio is equal to 1 and they were adjusted for multiple comparisons using <a href="https://en.wikipedia.org/wiki/False_discovery_rate#BH_procedure">Benjamini and Hochberg method</a>. The distribution of the p-values was tested against a uniform distribution using one-sample <a href="https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test">Kolmogorov-Smirnov test</a>, which is further explained in <a href="main_figure.html#sec-fig2a" class="quarto-xref"><span>Section 6.2.1</span></a>.</p>
<div class="cell">
<div id="lst-logreg" class="r cell-code listing quarto-float" data-code-summary="Code for univariate logisitic regression">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-logreg-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.10: Code for univariate logisitic regression
</figcaption>
<div aria-describedby="lst-logreg-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sourceCode cell-code" id="lst-logreg" data-code-summary="Code for univariate logisitic regression"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="lst-logreg-1"><a href="#lst-logreg-1" aria-hidden="true" tabindex="-1"></a>this.mat<span class="ot">&lt;-</span>dt.cpmZ[,.(SampleID,geneName,logCPMZ,<span class="at">y=</span><span class="fu">ifelse</span>(Condition<span class="sc">==</span><span class="st">"Case"</span>,<span class="dv">1</span>,<span class="dv">0</span>))] <span class="sc">%&gt;%</span> </span>
<span id="lst-logreg-2"><a href="#lst-logreg-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dcast.data.table</span>(SampleID<span class="sc">+</span>y<span class="sc">~</span>geneName,<span class="at">value.var=</span><span class="st">"logCPMZ"</span>) <span class="sc">%&gt;%</span> </span>
<span id="lst-logreg-3"><a href="#lst-logreg-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>(<span class="at">rownames=</span><span class="st">"SampleID"</span>) <span class="co"># logCPMZ</span></span>
<span id="lst-logreg-4"><a href="#lst-logreg-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-logreg-5"><a href="#lst-logreg-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(my.GA <span class="cf">in</span> <span class="fu">names</span>(li.GA)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]){ <span class="co"># for each 12wk, 20wk and 28wkGA</span></span>
<span id="lst-logreg-6"><a href="#lst-logreg-6" aria-hidden="true" tabindex="-1"></a>  this.samples<span class="ot">&lt;-</span><span class="fu">colnames</span>(dds.f2)[<span class="fu">colData</span>(dds.f2)<span class="sc">$</span>GA<span class="sc">==</span>my.GA]</span>
<span id="lst-logreg-7"><a href="#lst-logreg-7" aria-hidden="true" tabindex="-1"></a>  foo<span class="ot">&lt;-</span><span class="fu">list</span>() <span class="co"># per GA</span></span>
<span id="lst-logreg-8"><a href="#lst-logreg-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(my.ID <span class="cf">in</span> <span class="fu">rownames</span>(dds.f2)){</span>
<span id="lst-logreg-9"><a href="#lst-logreg-9" aria-hidden="true" tabindex="-1"></a>      <span class="co">#message(paste(my.GA,my.ID))</span></span>
<span id="lst-logreg-10"><a href="#lst-logreg-10" aria-hidden="true" tabindex="-1"></a>      df.mat<span class="ot">&lt;-</span>this.mat[this.samples,<span class="fu">c</span>(<span class="st">"y"</span>,my.ID)] <span class="sc">%&gt;%</span> as.data.frame</span>
<span id="lst-logreg-11"><a href="#lst-logreg-11" aria-hidden="true" tabindex="-1"></a>      my.model<span class="ot">&lt;-</span><span class="fu">glm</span>(y <span class="sc">~</span>., <span class="at">data=</span>df.mat, <span class="at">family=</span><span class="st">"binomial"</span>)</span>
<span id="lst-logreg-12"><a href="#lst-logreg-12" aria-hidden="true" tabindex="-1"></a>      <span class="co">#oddsratio::or_glm(data=df.mat, model=my.model, incr=list(OID20239=1))</span></span>
<span id="lst-logreg-13"><a href="#lst-logreg-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-logreg-14"><a href="#lst-logreg-14" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Log Odds Ratio &amp; p-values &amp; CI</span></span>
<span id="lst-logreg-15"><a href="#lst-logreg-15" aria-hidden="true" tabindex="-1"></a>      <span class="co">#my.model %&gt;% summary</span></span>
<span id="lst-logreg-16"><a href="#lst-logreg-16" aria-hidden="true" tabindex="-1"></a>      <span class="co">#coef(summary(my.model)) # isa 'matrix'</span></span>
<span id="lst-logreg-17"><a href="#lst-logreg-17" aria-hidden="true" tabindex="-1"></a>      foo1<span class="ot">&lt;-</span><span class="fu">as.data.frame</span>(<span class="fu">coef</span>(<span class="fu">summary</span>(my.model)))</span>
<span id="lst-logreg-18"><a href="#lst-logreg-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-logreg-19"><a href="#lst-logreg-19" aria-hidden="true" tabindex="-1"></a>      <span class="co">#cbind(coef(my.model), confint(my.model)) # Log Odds Ratio &amp; CI (95%)</span></span>
<span id="lst-logreg-20"><a href="#lst-logreg-20" aria-hidden="true" tabindex="-1"></a>      <span class="co">#exp(cbind(coef(my.model), confint(my.model))) # Odds Ratio &amp; CI (95%)</span></span>
<span id="lst-logreg-21"><a href="#lst-logreg-21" aria-hidden="true" tabindex="-1"></a>      foo2<span class="ot">&lt;-</span><span class="fu">as.data.frame</span>(<span class="fu">exp</span>(<span class="fu">cbind</span>(<span class="fu">coef</span>(my.model), <span class="fu">confint</span>(my.model)))) <span class="co"># Odds Ratio &amp; CI (95%)</span></span>
<span id="lst-logreg-22"><a href="#lst-logreg-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-logreg-23"><a href="#lst-logreg-23" aria-hidden="true" tabindex="-1"></a>      foo3<span class="ot">&lt;-</span><span class="fu">cbind</span>(my.ID,<span class="fu">cbind</span>(foo1,foo2)[<span class="dv">2</span>,]) <span class="sc">%&gt;%</span> data.table </span>
<span id="lst-logreg-24"><a href="#lst-logreg-24" aria-hidden="true" tabindex="-1"></a>      <span class="fu">colnames</span>(foo3)<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"gene_id"</span>,<span class="st">"log_odds"</span>,<span class="st">"se"</span>,<span class="st">"zval"</span>,<span class="st">"pval"</span>,<span class="st">"odds"</span>,<span class="st">"odds_lo"</span>,<span class="st">"odds_hi"</span>)</span>
<span id="lst-logreg-25"><a href="#lst-logreg-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-logreg-26"><a href="#lst-logreg-26" aria-hidden="true" tabindex="-1"></a>      <span class="co"># ROC &amp; AUC </span></span>
<span id="lst-logreg-27"><a href="#lst-logreg-27" aria-hidden="true" tabindex="-1"></a>      <span class="co">#predict(my.model,type=c("response"))  # predicted probability</span></span>
<span id="lst-logreg-28"><a href="#lst-logreg-28" aria-hidden="true" tabindex="-1"></a>      my.prob<span class="ot">&lt;-</span><span class="fu">fitted</span>(my.model) <span class="co"># same as above</span></span>
<span id="lst-logreg-29"><a href="#lst-logreg-29" aria-hidden="true" tabindex="-1"></a>      <span class="co"># no probability in case of NA in the matrix</span></span>
<span id="lst-logreg-30"><a href="#lst-logreg-30" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span>(<span class="fu">nrow</span>(df.mat)<span class="sc">!=</span><span class="fu">length</span>(my.prob)){</span>
<span id="lst-logreg-31"><a href="#lst-logreg-31" aria-hidden="true" tabindex="-1"></a>          dt.prob<span class="ot">&lt;-</span><span class="fu">rbind</span>(</span>
<span id="lst-logreg-32"><a href="#lst-logreg-32" aria-hidden="true" tabindex="-1"></a>          <span class="fu">data.table</span>(<span class="st">`</span><span class="at">index</span><span class="st">`</span><span class="ot">=</span><span class="fu">as.numeric</span>(<span class="fu">names</span>(my.prob)),<span class="st">`</span><span class="at">prob</span><span class="st">`</span><span class="ot">=</span>my.prob),</span>
<span id="lst-logreg-33"><a href="#lst-logreg-33" aria-hidden="true" tabindex="-1"></a>          <span class="fu">data.table</span>(<span class="st">`</span><span class="at">index</span><span class="st">`</span><span class="ot">=</span><span class="fu">as.numeric</span>(df.mat[,my.ID]<span class="sc">%&gt;%</span> is.na <span class="sc">%&gt;%</span> which), <span class="st">`</span><span class="at">prob</span><span class="st">`</span><span class="ot">=</span><span class="cn">NA</span>)</span>
<span id="lst-logreg-34"><a href="#lst-logreg-34" aria-hidden="true" tabindex="-1"></a>          )[<span class="fu">order</span>(index)]</span>
<span id="lst-logreg-35"><a href="#lst-logreg-35" aria-hidden="true" tabindex="-1"></a>          my.prob<span class="ot">&lt;-</span>dt.prob<span class="sc">$</span>prob</span>
<span id="lst-logreg-36"><a href="#lst-logreg-36" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="lst-logreg-37"><a href="#lst-logreg-37" aria-hidden="true" tabindex="-1"></a>      df.mat<span class="sc">$</span>prob<span class="ot">&lt;-</span>my.prob</span>
<span id="lst-logreg-38"><a href="#lst-logreg-38" aria-hidden="true" tabindex="-1"></a>      my.roc <span class="ot">&lt;-</span> pROC<span class="sc">::</span><span class="fu">roc</span>(y <span class="sc">~</span> prob, <span class="at">data =</span> df.mat, <span class="at">quiet=</span>T, <span class="at">ci=</span><span class="cn">TRUE</span>)</span>
<span id="lst-logreg-39"><a href="#lst-logreg-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-logreg-40"><a href="#lst-logreg-40" aria-hidden="true" tabindex="-1"></a>      foo[[my.ID]] <span class="ot">&lt;-</span> </span>
<span id="lst-logreg-41"><a href="#lst-logreg-41" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cbind</span>(</span>
<span id="lst-logreg-42"><a href="#lst-logreg-42" aria-hidden="true" tabindex="-1"></a>            foo3,</span>
<span id="lst-logreg-43"><a href="#lst-logreg-43" aria-hidden="true" tabindex="-1"></a>            <span class="fu">data.table</span>(<span class="st">`</span><span class="at">auc</span><span class="st">`</span><span class="ot">=</span>my.roc<span class="sc">$</span>ci[<span class="dv">2</span>], </span>
<span id="lst-logreg-44"><a href="#lst-logreg-44" aria-hidden="true" tabindex="-1"></a>                        <span class="st">`</span><span class="at">auc_lo</span><span class="st">`</span><span class="ot">=</span>my.roc<span class="sc">$</span>ci[<span class="dv">1</span>], </span>
<span id="lst-logreg-45"><a href="#lst-logreg-45" aria-hidden="true" tabindex="-1"></a>                        <span class="st">`</span><span class="at">auc_hi</span><span class="st">`</span><span class="ot">=</span>my.roc<span class="sc">$</span>ci[<span class="dv">3</span>],</span>
<span id="lst-logreg-46"><a href="#lst-logreg-46" aria-hidden="true" tabindex="-1"></a>                        <span class="st">`</span><span class="at">AIC</span><span class="st">`</span><span class="ot">=</span><span class="fu">aic</span>(my.model),</span>
<span id="lst-logreg-47"><a href="#lst-logreg-47" aria-hidden="true" tabindex="-1"></a>                        <span class="st">`</span><span class="at">BIC</span><span class="st">`</span><span class="ot">=</span><span class="fu">bic</span>(my.model) ) </span>
<span id="lst-logreg-48"><a href="#lst-logreg-48" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="lst-logreg-49"><a href="#lst-logreg-49" aria-hidden="true" tabindex="-1"></a>  } <span class="co"># end of for   </span></span>
<span id="lst-logreg-50"><a href="#lst-logreg-50" aria-hidden="true" tabindex="-1"></a>  dl.logregZ[[my.GA]]<span class="ot">&lt;-</span><span class="fu">rbindlist</span>(foo)</span>
<span id="lst-logreg-51"><a href="#lst-logreg-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-logreg-52"><a href="#lst-logreg-52" aria-hidden="true" tabindex="-1"></a>  <span class="co"># apply p.adjust</span></span>
<span id="lst-logreg-53"><a href="#lst-logreg-53" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> adjust.methods){</span>
<span id="lst-logreg-54"><a href="#lst-logreg-54" aria-hidden="true" tabindex="-1"></a>      dl.logregZ[[my.GA]][,<span class="fu">paste0</span>(<span class="st">"padj."</span>,i)<span class="sc">:</span><span class="er">=</span><span class="fu">p.adjust</span>(pval,i)]</span>
<span id="lst-logreg-55"><a href="#lst-logreg-55" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="lst-logreg-56"><a href="#lst-logreg-56" aria-hidden="true" tabindex="-1"></a>} <span class="co"># end of for my.GA</span></span>
<span id="lst-logreg-57"><a href="#lst-logreg-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-logreg-58"><a href="#lst-logreg-58" aria-hidden="true" tabindex="-1"></a><span class="co"># cache to use it later</span></span>
<span id="lst-logreg-59"><a href="#lst-logreg-59" aria-hidden="true" tabindex="-1"></a><span class="co">#save(dl.logregZ, file=paste0("RData/dl.logregZ.",my.type,".",my.salmon.index,".RData"))</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</figure>
</div>
</div>
</section>
</section>
<section id="sec-core-deg" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="sec-core-deg"><span class="header-section-number">2.3</span> Selection of core differentially expressed genes</h2>
<p>To select a subset of genes that best explains the outcome of samples from the 28wkGA group of the discovery cohort, we used the following four criteria: 1) the p-values from DESeq2, 2) the p-values from edgeR, 3) AIC, and 4) AUC from univariable logistic regressions. Then we selected the top 1% genes for each category (i.e.&nbsp;151 genes of the lowest p-values from DESeq2 and edgeR, 151 genes having the lowest AIC, and 151 genes having the highest AUC) and constructed a Venn diagram using ggvenn (v0.1.0) R package (52). We selected a total of 17 genes satisfying all the four criteria (i.e.&nbsp;the intersection) and constructed a gene expression matrix (i.e.&nbsp;the 17 genes across the samples from the 28wkGA group of the discovery cohort) using the z-score.</p>
<div class="cell">
<div id="lst-set-core-deg" class="r cell-code listing quarto-float" data-code-summary="Code to find the core DEGs">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-set-core-deg-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.11: Code to find the core DEGs
</figcaption>
<div aria-describedby="lst-set-core-deg-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sourceCode cell-code" id="lst-set-core-deg" data-code-summary="Code to find the core DEGs"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="lst-set-core-deg-1"><a href="#lst-set-core-deg-1" aria-hidden="true" tabindex="-1"></a>venn.top1pctZ<span class="ot">&lt;-</span><span class="fu">list</span>(<span class="st">`</span><span class="at">DESeq2</span><span class="st">`</span><span class="ot">=</span>dl.resLFC[[<span class="st">"28wk"</span>]][<span class="fu">order</span>(pvalue)][<span class="dv">1</span><span class="sc">:</span><span class="dv">151</span>]<span class="sc">$</span>gene_id,</span>
<span id="lst-set-core-deg-2"><a href="#lst-set-core-deg-2" aria-hidden="true" tabindex="-1"></a>                  <span class="st">`</span><span class="at">edgeR</span><span class="st">`</span><span class="ot">=</span>dl.res.edgeR[[<span class="st">"28wk"</span>]][<span class="fu">order</span>(PValue)][<span class="dv">1</span><span class="sc">:</span><span class="dv">151</span>]<span class="sc">$</span>gene_id,</span>
<span id="lst-set-core-deg-3"><a href="#lst-set-core-deg-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">`</span><span class="at">AUC</span><span class="st">`</span><span class="ot">=</span>dl.logregZ[[<span class="st">"28wk"</span>]][<span class="fu">order</span>(<span class="sc">-</span>auc)][<span class="dv">1</span><span class="sc">:</span><span class="dv">151</span>]<span class="sc">$</span>gene_id,</span>
<span id="lst-set-core-deg-4"><a href="#lst-set-core-deg-4" aria-hidden="true" tabindex="-1"></a>                <span class="st">`</span><span class="at">AIC</span><span class="st">`</span><span class="ot">=</span>dl.logregZ[[<span class="st">"28wk"</span>]][<span class="fu">order</span>(AIC)][<span class="dv">1</span><span class="sc">:</span><span class="dv">151</span>]<span class="sc">$</span>gene_id</span>
<span id="lst-set-core-deg-5"><a href="#lst-set-core-deg-5" aria-hidden="true" tabindex="-1"></a>                )</span>
<span id="lst-set-core-deg-6"><a href="#lst-set-core-deg-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-set-core-deg-7"><a href="#lst-set-core-deg-7" aria-hidden="true" tabindex="-1"></a><span class="co"># the union of all</span></span>
<span id="lst-set-core-deg-8"><a href="#lst-set-core-deg-8" aria-hidden="true" tabindex="-1"></a>dt.venn.top1pctZ<span class="ot">&lt;-</span><span class="fu">lapply</span>(<span class="fu">names</span>(venn.top1pctZ), <span class="cf">function</span>(i) <span class="fu">data.table</span>(i,venn.top1pctZ[[i]])) <span class="sc">%&gt;%</span> </span>
<span id="lst-set-core-deg-9"><a href="#lst-set-core-deg-9" aria-hidden="true" tabindex="-1"></a>  rbindlist <span class="sc">%&gt;%</span> </span>
<span id="lst-set-core-deg-10"><a href="#lst-set-core-deg-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dcast.data.table</span>(V2<span class="sc">~</span>i, <span class="at">fun=</span>length)</span>
<span id="lst-set-core-deg-11"><a href="#lst-set-core-deg-11" aria-hidden="true" tabindex="-1"></a><span class="fu">setnames</span>(dt.venn.top1pctZ,<span class="st">"V2"</span>,<span class="st">"Gene"</span>)</span>
<span id="lst-set-core-deg-12"><a href="#lst-set-core-deg-12" aria-hidden="true" tabindex="-1"></a>top1pct.genesZ<span class="ot">&lt;-</span>dt.venn.top1pctZ[AIC<span class="sc">==</span><span class="dv">1</span> <span class="sc">&amp;</span> AUC<span class="sc">==</span><span class="dv">1</span> <span class="sc">&amp;</span> DESeq2<span class="sc">==</span><span class="dv">1</span> <span class="sc">&amp;</span> edgeR<span class="sc">==</span><span class="dv">1</span>]<span class="sc">$</span>Gene</span>
<span id="lst-set-core-deg-13"><a href="#lst-set-core-deg-13" aria-hidden="true" tabindex="-1"></a>core17 <span class="ot">&lt;-</span> top1pct.genesZ <span class="co"># used in 5-fold CV</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</figure>
</div>
</div>
<p>The Venn diagram for the 17 core DEGs is shown in <a href="main_figure.html#sec-fig2c" class="quarto-xref"><span>Section 6.2.3</span></a>.</p>


<!-- -->

</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/ObsGynaeCam\.github\.io\/cell-free-rna-2024");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./sample.html" class="pagination-link" aria-label="Sample Information">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Sample Information</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./cross_validation.html" class="pagination-link" aria-label="Cross Validation">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Cross Validation</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb1" data-shortcodes="false"><pre class="sourceCode numberSource markdown number-lines code-with-copy"><code class="sourceCode markdown"><span id="cb1-1"><a href="#cb1-1"></a><span class="fu"># RNA-seq data processing {#sec-mtd-rna-seq}</span></span>
<span id="cb1-2"><a href="#cb1-2"></a></span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="fu">## Transcript quantification {#sec-mtd-rna-seq-quant}</span></span>
<span id="cb1-4"><a href="#cb1-4"></a></span>
<span id="cb1-5"><a href="#cb1-5"></a>For an RNA-seq quantification method applied in this study, we chose the most reliable approach based on the performance of predicting fetal sex by measuring the extent of chromosome Y encoded transcripts (see @sec-si-quant-method for detail). Based on this benchmark result, we chose <span class="co">[</span><span class="ot">`Salmon`</span><span class="co">](https://combine-lab.github.io/salmon/)</span> (v1.5.2) in mapping-mode to process our RNA-seq datasets. </span>
<span id="cb1-6"><a href="#cb1-6"></a></span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="fu">### Salmon index {#sec-mtd-salmon-index}</span></span>
<span id="cb1-8"><a href="#cb1-8"></a></span>
<span id="cb1-9"><a href="#cb1-9"></a>``<span class="in">`{#lst-salmon-index .bash lst-cap="A pseudo-code to run `</span>salmon index`" filename="static/shell/salmon.decoy.sh"}</span>
<span id="cb1-10"><a href="#cb1-10"></a>{{&lt; include static/shell/salmon.decoy.sh &gt;}}</span>
<span id="cb1-11"><a href="#cb1-11"></a><span class="in">```</span></span>
<span id="cb1-12"><a href="#cb1-12"></a></span>
<span id="cb1-13"><a href="#cb1-13"></a><span class="in">### Salmon quant {#sec-mtd-salmon-quant}</span></span>
<span id="cb1-14"><a href="#cb1-14"></a></span>
<span id="cb1-15"><a href="#cb1-15"></a><span class="in">```{#lst-salmon-quant .bash lst-cap="A pseudo-code to run `salmon quant` in mapping mode" filename="static/shell/salmon-quant.sh"}</span></span>
<span id="cb1-16"><a href="#cb1-16"></a><span class="in">{{&lt; include static/shell/salmon-quant.sh &gt;}}</span></span>
<span id="cb1-17"><a href="#cb1-17"></a><span class="in">```</span></span>
<span id="cb1-18"><a href="#cb1-18"></a></span>
<span id="cb1-19"><a href="#cb1-19"></a><span class="fu">## Differentially expressed gene analysis {#sec-mtd-deg}</span></span>
<span id="cb1-20"><a href="#cb1-20"></a></span>
<span id="cb1-21"><a href="#cb1-21"></a>The differentially expressed gene analysis was conducted for each gestational epoch (12wk, 20wk, 28wk, and 36wk) separately, except for the 36wkGA gestation samples of the pre-term dataset which has only one PE case. </span>
<span id="cb1-22"><a href="#cb1-22"></a></span>
<span id="cb1-23"><a href="#cb1-23"></a>Firstly, the following parameters were defined:</span>
<span id="cb1-24"><a href="#cb1-24"></a><span class="in">```{#lst-set-config .r lst-cap="Set the parameters" filename="static/R/config/cfRNA-2024.R"}</span></span>
<span id="cb1-25"><a href="#cb1-25"></a><span class="in">{{&lt; include static/R/config/cfRNA-2024.R &gt;}}</span></span>
<span id="cb1-26"><a href="#cb1-26"></a><span class="in">```</span></span>
<span id="cb1-27"><a href="#cb1-27"></a></span>
<span id="cb1-28"><a href="#cb1-28"></a>The transcript-level read count matrices (e.g. quant.sf files) were imported using <span class="co">[</span><span class="ot">`tximeta`</span><span class="co">](https://bioconductor.org/packages/release/bioc/html/tximeta.html)</span> (v1.8.5) Bioconductor package and merged at the gene-level. </span>
<span id="cb1-29"><a href="#cb1-29"></a></span>
<span id="cb1-30"><a href="#cb1-30"></a><span class="in">```{r prep-quant-sf}</span></span>
<span id="cb1-31"><a href="#cb1-31"></a><span class="in">#| label: lst-prep-quant-sf</span></span>
<span id="cb1-32"><a href="#cb1-32"></a><span class="in">#| lst-label: lst-prep-quant-sf</span></span>
<span id="cb1-33"><a href="#cb1-33"></a><span class="in">#| eval: false</span></span>
<span id="cb1-34"><a href="#cb1-34"></a><span class="in">#| code-summary: Code to set the `Salmon` quant files of the pre-term dataset (i.e. Discovery dataset)</span></span>
<span id="cb1-35"><a href="#cb1-35"></a><span class="in">#| lst-cap: Code to set the `Salmon` quant files of the pre-term dataset (i.e. Discovery dataset)</span></span>
<span id="cb1-36"><a href="#cb1-36"></a></span>
<span id="cb1-37"><a href="#cb1-37"></a><span class="in">dt.foo&lt;-data.table(`files`=system(paste0("ls ", "~/results/",my.slx,"/",my.salmon,"/",my.salmon.index,"/*/quant.sf"), intern=T))</span></span>
<span id="cb1-38"><a href="#cb1-38"></a><span class="in">dt.foo[,names:=tstrsplit(files,"/",keep=8)]</span></span>
<span id="cb1-39"><a href="#cb1-39"></a><span class="in">dt.colDataAll&lt;-merge(dt.foo,dt.samples,by.x="names",by.y="SampleID") # n=755</span></span>
<span id="cb1-40"><a href="#cb1-40"></a></span>
<span id="cb1-41"><a href="#cb1-41"></a><span class="in"># pre-term</span></span>
<span id="cb1-42"><a href="#cb1-42"></a><span class="in">dt.colData&lt;-dt.colDataAll[Type==my.type] # n=279</span></span>
<span id="cb1-43"><a href="#cb1-43"></a><span class="in">#dt.colData[grepl("-b$",names)] # CX (CX-b): both failed; HQ (HQ-b): only HQ-b passed QC according to illumina</span></span>
<span id="cb1-44"><a href="#cb1-44"></a><span class="in">dt.colData&lt;-dt.colData[!names %in% c("GS-59-CX-b","GS-179-HQ")] # remove these two samples (n=277)</span></span>
<span id="cb1-45"><a href="#cb1-45"></a><span class="in">li.GA&lt;-split(dt.colData, dt.colData$GA)</span></span>
<span id="cb1-46"><a href="#cb1-46"></a></span>
<span id="cb1-47"><a href="#cb1-47"></a><span class="in">#tx2gene</span></span>
<span id="cb1-48"><a href="#cb1-48"></a><span class="in">dt.tx2gene&lt;-fread("~/results/RNA-Seq/Placentome/gffcompare/POPS-2022/POPS-Placenta-Transcriptome/POPS-2022.GRCh38.88.Novel.Known.Freq.0.1.TPM.0.1.tr.reconstruction.tx2gene.txt", header=F,col.names=c("transcript_id","gene_id"))</span></span>
<span id="cb1-49"><a href="#cb1-49"></a><span class="in">```</span></span>
<span id="cb1-50"><a href="#cb1-50"></a></span>
<span id="cb1-51"><a href="#cb1-51"></a><span class="in">```{r prep-dds}</span></span>
<span id="cb1-52"><a href="#cb1-52"></a><span class="in">#| label: lst-prep-dds</span></span>
<span id="cb1-53"><a href="#cb1-53"></a><span class="in">#| lst-label: lst-prep-dds</span></span>
<span id="cb1-54"><a href="#cb1-54"></a><span class="in">#| eval: false</span></span>
<span id="cb1-55"><a href="#cb1-55"></a><span class="in">#| code-summary: "Code to make `dds` (DESeq Data Set) object"</span></span>
<span id="cb1-56"><a href="#cb1-56"></a><span class="in">#| lst-cap: "Code to make `dds` (DESeq Data Set) object"</span></span>
<span id="cb1-57"><a href="#cb1-57"></a></span>
<span id="cb1-58"><a href="#cb1-58"></a><span class="in">library(DESeq2)</span></span>
<span id="cb1-59"><a href="#cb1-59"></a><span class="in">library(edgeR)</span></span>
<span id="cb1-60"><a href="#cb1-60"></a><span class="in">library("BiocParallel")</span></span>
<span id="cb1-61"><a href="#cb1-61"></a><span class="in">register(MulticoreParam(12))</span></span>
<span id="cb1-62"><a href="#cb1-62"></a></span>
<span id="cb1-63"><a href="#cb1-63"></a><span class="in">#tximeta via tx2gene</span></span>
<span id="cb1-64"><a href="#cb1-64"></a><span class="in"># 36wk dataset excluded from the beginnin</span></span>
<span id="cb1-65"><a href="#cb1-65"></a><span class="in">gse&lt;-tximeta::tximeta(dt.colData[GA!="36wk"],</span></span>
<span id="cb1-66"><a href="#cb1-66"></a><span class="in">                      skipMeta=T,</span></span>
<span id="cb1-67"><a href="#cb1-67"></a><span class="in">                      tx2gene=dt.tx2gene[,.(transcript_id,gene_id)],txOut=F)</span></span>
<span id="cb1-68"><a href="#cb1-68"></a></span>
<span id="cb1-69"><a href="#cb1-69"></a><span class="in"># set up `dds` at gene-level</span></span>
<span id="cb1-70"><a href="#cb1-70"></a><span class="in">my.design &lt;- formula(~ Batch + GA + Sex + Condition)    # isa 'formula'</span></span>
<span id="cb1-71"><a href="#cb1-71"></a><span class="in">dds &lt;- DESeqDataSet(se=gse, design=my.design)</span></span>
<span id="cb1-72"><a href="#cb1-72"></a></span>
<span id="cb1-73"><a href="#cb1-73"></a><span class="in">dds$Group&lt;-factor(paste0(dds$GA,dds$Condition)) # add 'Group'</span></span>
<span id="cb1-74"><a href="#cb1-74"></a><span class="in">design(dds) &lt;- formula(~ Batch + Sex + Group)     # isa 'formula'</span></span>
<span id="cb1-75"><a href="#cb1-75"></a></span>
<span id="cb1-76"><a href="#cb1-76"></a><span class="in">dds&lt;- DESeq(dds, parallel=TRUE) # isa 'DESeqDataSet'</span></span>
<span id="cb1-77"><a href="#cb1-77"></a><span class="in">```</span></span>
<span id="cb1-78"><a href="#cb1-78"></a></span>
<span id="cb1-79"><a href="#cb1-79"></a><span class="fu">### `DESeq2` {#sec-mtd-deg-deseq2}</span></span>
<span id="cb1-80"><a href="#cb1-80"></a></span>
<span id="cb1-81"><a href="#cb1-81"></a>We only considered genes found in ≥10% of samples (i.e. ≥ 23) having ≥10 reads, and discarded genes detected as dispersion outliers by <span class="co">[</span><span class="ot">`DESeq2`</span><span class="co">](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)</span> (v1.30.0).</span>
<span id="cb1-82"><a href="#cb1-82"></a></span>
<span id="cb1-83"><a href="#cb1-83"></a><span class="in">```{r set-dds1}</span></span>
<span id="cb1-84"><a href="#cb1-84"></a><span class="in">#| label: lst-set-dds</span></span>
<span id="cb1-85"><a href="#cb1-85"></a><span class="in">#| lst-label: lst-set-dds</span></span>
<span id="cb1-86"><a href="#cb1-86"></a><span class="in">#| eval: false</span></span>
<span id="cb1-87"><a href="#cb1-87"></a><span class="in">#| code-summary: Code to set up `dds` object</span></span>
<span id="cb1-88"><a href="#cb1-88"></a><span class="in">#| lst-cap: Code to set up `dds` object</span></span>
<span id="cb1-89"><a href="#cb1-89"></a><span class="in"># at leat 10 reads for at leat 10  % of the samples </span></span>
<span id="cb1-90"><a href="#cb1-90"></a><span class="in">keep &lt;- rowSums(counts(dds) &gt;= minRead) &gt;= ncol(dds)*minFreq </span></span>
<span id="cb1-91"><a href="#cb1-91"></a><span class="in">dds.f&lt;- DESeq(dds[keep,], parallel=TRUE) # isa 'DESeqDataSet'</span></span>
<span id="cb1-92"><a href="#cb1-92"></a><span class="in">dim(dds.f) # 15380 x 221</span></span>
<span id="cb1-93"><a href="#cb1-93"></a><span class="in">```</span></span>
<span id="cb1-94"><a href="#cb1-94"></a></span>
<span id="cb1-95"><a href="#cb1-95"></a>A total of 15,150 genes and 221 samples were used to find differentially expressed genes by adjusting the batch number, the fetal sex, and the gestion of the samples in the design matrix of DESeq2. The p-values were calculated from the null hypothesis that the fold changes were less than or equal to 20% (i.e. lfcThreshold=log2(1.2)) in case and control groups. </span>
<span id="cb1-96"><a href="#cb1-96"></a></span>
<span id="cb1-97"><a href="#cb1-97"></a><span class="in">```{r set-deg1}</span></span>
<span id="cb1-98"><a href="#cb1-98"></a><span class="in">#| label: lst-set-deg</span></span>
<span id="cb1-99"><a href="#cb1-99"></a><span class="in">#| lst-label: lst-set-deg</span></span>
<span id="cb1-100"><a href="#cb1-100"></a><span class="in">#| eval: false</span></span>
<span id="cb1-101"><a href="#cb1-101"></a><span class="in">#| code-summary: Code to find DEGs by DESeq</span></span>
<span id="cb1-102"><a href="#cb1-102"></a><span class="in">#| lst-cap: Code to find DEGs by DESeq</span></span>
<span id="cb1-103"><a href="#cb1-103"></a><span class="in"># remove dispOutlier genes</span></span>
<span id="cb1-104"><a href="#cb1-104"></a><span class="in">keep2&lt;-!is.na(rowData(dds.f)[,"dispOutlier"]) &amp; !rowData(dds.f)[,"dispOutlier"]</span></span>
<span id="cb1-105"><a href="#cb1-105"></a><span class="in">dds.f2&lt;- DESeq(dds.f[keep2], parallel=T)</span></span>
<span id="cb1-106"><a href="#cb1-106"></a></span>
<span id="cb1-107"><a href="#cb1-107"></a><span class="in">dim(dds.f2) # 15150 x 221</span></span>
<span id="cb1-108"><a href="#cb1-108"></a></span>
<span id="cb1-109"><a href="#cb1-109"></a><span class="in"># apply shink separately for 12wk, 20wk, and 28wk (li.GA[1:3])</span></span>
<span id="cb1-110"><a href="#cb1-110"></a><span class="in">li.resLFC&lt;-lapply(names(li.GA)[1:3],function(my.GA){ </span></span>
<span id="cb1-111"><a href="#cb1-111"></a><span class="in">  my.res&lt;-results(dds.f2, </span></span>
<span id="cb1-112"><a href="#cb1-112"></a><span class="in">          independentFiltering=FALSE, #by default independant filtering at FDR (alpha) 0.1 (10%) </span></span>
<span id="cb1-113"><a href="#cb1-113"></a><span class="in">          lfcThreshold=log2(minFC), </span></span>
<span id="cb1-114"><a href="#cb1-114"></a><span class="in">          contrast=c("Group",paste0(my.GA,c("Case","Control"))),</span></span>
<span id="cb1-115"><a href="#cb1-115"></a><span class="in">          parallel=TRUE)  </span></span>
<span id="cb1-116"><a href="#cb1-116"></a><span class="in">  lfcShrink(dds.f2, </span></span>
<span id="cb1-117"><a href="#cb1-117"></a><span class="in">            res=my.res,</span></span>
<span id="cb1-118"><a href="#cb1-118"></a><span class="in">            lfcThreshold=log2(minFC), # not applicable for 'asher' </span></span>
<span id="cb1-119"><a href="#cb1-119"></a><span class="in">            type="ashr",</span></span>
<span id="cb1-120"><a href="#cb1-120"></a><span class="in">            parallel=TRUE)  </span></span>
<span id="cb1-121"><a href="#cb1-121"></a><span class="in">})</span></span>
<span id="cb1-122"><a href="#cb1-122"></a><span class="in">names(li.resLFC)&lt;-names(li.GA)[1:3]</span></span>
<span id="cb1-123"><a href="#cb1-123"></a></span>
<span id="cb1-124"><a href="#cb1-124"></a><span class="in">dl.resLFC&lt;-lapply(li.resLFC, function(i)</span></span>
<span id="cb1-125"><a href="#cb1-125"></a><span class="in">  data.table(`gene_id`=rownames(i), </span></span>
<span id="cb1-126"><a href="#cb1-126"></a><span class="in">              as.data.frame(i))[order(pvalue)][,`:=`("BH"=p.adjust(pvalue,"BH"),</span></span>
<span id="cb1-127"><a href="#cb1-127"></a><span class="in">                                              "BY"=p.adjust(pvalue,"BY"),</span></span>
<span id="cb1-128"><a href="#cb1-128"></a><span class="in">                                              "bf"=p.adjust(pvalue,"bonferroni"))]</span></span>
<span id="cb1-129"><a href="#cb1-129"></a><span class="in">)</span></span>
<span id="cb1-130"><a href="#cb1-130"></a><span class="in">fwrite(dl.resLFC[["28wk"]], file=paste0("static/R/result/DEG.DSeq2.28wk.",my.type,".",my.salmon.index,".csv"))</span></span>
<span id="cb1-131"><a href="#cb1-131"></a><span class="in">```</span></span>
<span id="cb1-132"><a href="#cb1-132"></a></span>
<span id="cb1-133"><a href="#cb1-133"></a><span class="fu">### `edgeR` {#sec-mtd-deg-edger}</span></span>
<span id="cb1-134"><a href="#cb1-134"></a></span>
<span id="cb1-135"><a href="#cb1-135"></a>For <span class="co">[</span><span class="ot">edgeR</span><span class="co">](https://bioconductor.org/packages/release/bioc/html/edgeR.html)</span> (v3.32.1) analysis, we used <span class="in">`makeDGEList`</span> function of <span class="in">`tximeta`</span> Bioconductor package to convert the data object of the 15,150 genes across the 221 samples as mentioned above. The gene-level count matrix was normalised by using <span class="in">`calcNormFactors`</span> function of <span class="in">`edgeR`</span> with <span class="in">`TMM`</span> method (trimmed mean of M values) and a quasi-likelihood negative binomial generalised log-linear model (i.e. <span class="in">`glmQLFit`</span>) was applied to account for the batch number, the fetal sex and the gestation information in the design matrix of edgeR. For a statistical test, we used <span class="in">`glmTreat`</span> of edgeR with at least 20% fold-change (i.e. lfc=log2(1.2)).</span>
<span id="cb1-136"><a href="#cb1-136"></a></span>
<span id="cb1-137"><a href="#cb1-137"></a><span class="in">```{r set-edger1}</span></span>
<span id="cb1-138"><a href="#cb1-138"></a><span class="in">#| label: lst-set-edger</span></span>
<span id="cb1-139"><a href="#cb1-139"></a><span class="in">#| lst-label: lst-set-edger</span></span>
<span id="cb1-140"><a href="#cb1-140"></a><span class="in">#| eval: false</span></span>
<span id="cb1-141"><a href="#cb1-141"></a><span class="in">#| code-summary: "Code to find DEGs by `edgeR`"</span></span>
<span id="cb1-142"><a href="#cb1-142"></a><span class="in">#| lst-cap: "Code to find DEGs by `edgeR`"</span></span>
<span id="cb1-143"><a href="#cb1-143"></a><span class="in">gse2 &lt;-gse[rownames(dds.f2),]</span></span>
<span id="cb1-144"><a href="#cb1-144"></a><span class="in">d2&lt;-tximeta::makeDGEList(gse2)</span></span>
<span id="cb1-145"><a href="#cb1-145"></a><span class="in">d2$samples&lt;-cbind(d2$samples,colData(gse2))</span></span>
<span id="cb1-146"><a href="#cb1-146"></a><span class="in">d2$samples$group&lt;-factor(paste0(d2$samples$GA,d2$samples$Condition)) # add 'group'</span></span>
<span id="cb1-147"><a href="#cb1-147"></a><span class="in">d2$samples$GA&lt;-droplevels(d2$samples$GA) # 36wk removed</span></span>
<span id="cb1-148"><a href="#cb1-148"></a><span class="in">d2$samples$Batch&lt;-droplevels(d2$samples$Batch) # some batches removed</span></span>
<span id="cb1-149"><a href="#cb1-149"></a></span>
<span id="cb1-150"><a href="#cb1-150"></a><span class="in"># additional filter</span></span>
<span id="cb1-151"><a href="#cb1-151"></a><span class="in">#keep &lt;- filterByExpr(d2)</span></span>
<span id="cb1-152"><a href="#cb1-152"></a><span class="in">#keep %&gt;% table # FALSE:13, TRUE:15137</span></span>
<span id="cb1-153"><a href="#cb1-153"></a><span class="in">#d2&lt;-d2[keep,]</span></span>
<span id="cb1-154"><a href="#cb1-154"></a></span>
<span id="cb1-155"><a href="#cb1-155"></a><span class="in"># TMM normalisation (default). It adds `norm.factors` d2$samples</span></span>
<span id="cb1-156"><a href="#cb1-156"></a><span class="in"># NB, we have `offsets`, which take precedence over lib.size and norm.factors</span></span>
<span id="cb1-157"><a href="#cb1-157"></a><span class="in">d2&lt;-calcNormFactors(d2,method="TMM") </span></span>
<span id="cb1-158"><a href="#cb1-158"></a><span class="in">                                    </span></span>
<span id="cb1-159"><a href="#cb1-159"></a><span class="in">plotMDS(d2, col=as.numeric(d2$samples$GA))</span></span>
<span id="cb1-160"><a href="#cb1-160"></a><span class="in">plotMDS(d2, col=as.numeric(d2$samples$Condition))</span></span>
<span id="cb1-161"><a href="#cb1-161"></a></span>
<span id="cb1-162"><a href="#cb1-162"></a><span class="in"># design</span></span>
<span id="cb1-163"><a href="#cb1-163"></a><span class="in">my.design &lt;- model.matrix(~ 0 + group + Sex + Batch , data=d2$samples)      # isa 'matrix'</span></span>
<span id="cb1-164"><a href="#cb1-164"></a><span class="in">(my.contrasts &lt;- makeContrasts(`12wk`=group12wkCase-group12wkControl, </span></span>
<span id="cb1-165"><a href="#cb1-165"></a><span class="in">                              `20wk`=group20wkCase-group20wkControl, </span></span>
<span id="cb1-166"><a href="#cb1-166"></a><span class="in">                              `28wk`=group28wkCase-group28wkControl, </span></span>
<span id="cb1-167"><a href="#cb1-167"></a><span class="in">                              levels=my.design)</span></span>
<span id="cb1-168"><a href="#cb1-168"></a><span class="in">)</span></span>
<span id="cb1-169"><a href="#cb1-169"></a></span>
<span id="cb1-170"><a href="#cb1-170"></a><span class="in"># dispersion</span></span>
<span id="cb1-171"><a href="#cb1-171"></a><span class="in">#dp2 = estimateDisp(d2, design=my.design, verbose=T)</span></span>
<span id="cb1-172"><a href="#cb1-172"></a><span class="in">dp2 = estimateGLMCommonDisp(d2, design=my.design, verbose=T) </span></span>
<span id="cb1-173"><a href="#cb1-173"></a><span class="in">dp2 = estimateGLMTrendedDisp(dp2, design=my.design, verbose=T)</span></span>
<span id="cb1-174"><a href="#cb1-174"></a><span class="in">dp2 = estimateGLMTagwiseDisp(dp2, design=my.design)</span></span>
<span id="cb1-175"><a href="#cb1-175"></a></span>
<span id="cb1-176"><a href="#cb1-176"></a><span class="in">plotMDS(dp2)</span></span>
<span id="cb1-177"><a href="#cb1-177"></a><span class="in">plotBCV(dp2)</span></span>
<span id="cb1-178"><a href="#cb1-178"></a></span>
<span id="cb1-179"><a href="#cb1-179"></a><span class="in"># fit</span></span>
<span id="cb1-180"><a href="#cb1-180"></a><span class="in">#f = glmFit(dp2, design=my.design) # </span></span>
<span id="cb1-181"><a href="#cb1-181"></a><span class="in">f = glmQLFit(dp2, design=my.design) # QL(Quasi-like) pipeline</span></span>
<span id="cb1-182"><a href="#cb1-182"></a></span>
<span id="cb1-183"><a href="#cb1-183"></a><span class="in">plotMD(f)</span></span>
<span id="cb1-184"><a href="#cb1-184"></a><span class="in">plotQLDisp(f)</span></span>
<span id="cb1-185"><a href="#cb1-185"></a></span>
<span id="cb1-186"><a href="#cb1-186"></a><span class="in"># get the edgeR results</span></span>
<span id="cb1-187"><a href="#cb1-187"></a><span class="in">li.res.edgeR&lt;-lapply(names(li.GA)[1:3], function(i){</span></span>
<span id="cb1-188"><a href="#cb1-188"></a><span class="in">  te &lt;- glmTreat(f, contrast=my.contrasts[,i], lfc=log2(minFC))</span></span>
<span id="cb1-189"><a href="#cb1-189"></a><span class="in">  topTags(te, n=nrow(te))    # default sort by pvalue</span></span>
<span id="cb1-190"><a href="#cb1-190"></a><span class="in">})</span></span>
<span id="cb1-191"><a href="#cb1-191"></a><span class="in">names(li.res.edgeR) &lt;-names(li.GA)[1:3] </span></span>
<span id="cb1-192"><a href="#cb1-192"></a></span>
<span id="cb1-193"><a href="#cb1-193"></a><span class="in">dl.res.edgeR&lt;-lapply(li.res.edgeR, function(i)</span></span>
<span id="cb1-194"><a href="#cb1-194"></a><span class="in">  data.table(`gene_id`=rownames(i),i$table)[order(PValue)][,`:=`("BH"=p.adjust(PValue,"BH"),</span></span>
<span id="cb1-195"><a href="#cb1-195"></a><span class="in">                                                                  "BY"=p.adjust(PValue,"BY"),</span></span>
<span id="cb1-196"><a href="#cb1-196"></a><span class="in">                                                                  "bf"=p.adjust(PValue,"bonferroni"))]</span></span>
<span id="cb1-197"><a href="#cb1-197"></a><span class="in">)</span></span>
<span id="cb1-198"><a href="#cb1-198"></a><span class="in">lapply(dl.res.edgeR, function(i) i[FDR&lt;0.05] %&gt;% nrow)</span></span>
<span id="cb1-199"><a href="#cb1-199"></a></span>
<span id="cb1-200"><a href="#cb1-200"></a><span class="in">fwrite(dl.res.edgeR[["28wk"]], file=paste0("static/R/result/DEG.edgeR.28wk.",my.type,".",my.salmon.index,".csv"))</span></span>
<span id="cb1-201"><a href="#cb1-201"></a><span class="in">```</span></span>
<span id="cb1-202"><a href="#cb1-202"></a></span>
<span id="cb1-203"><a href="#cb1-203"></a><span class="fu">### The z-score matrix {#sec-zscore-mat}</span></span>
<span id="cb1-204"><a href="#cb1-204"></a></span>
<span id="cb1-205"><a href="#cb1-205"></a>The gene-level count matrix was converted as the unit of CPM (Count Per Million), in log2-scale, via the “cpm” function of edgeR and it was further transformed into a matrix of the z-score using the mean and standard deviation of logCPM from the control samples of each corresponding gestational age group.</span>
<span id="cb1-206"><a href="#cb1-206"></a></span>
<span id="cb1-207"><a href="#cb1-207"></a><span class="in">```{r cnt-cpm}</span></span>
<span id="cb1-208"><a href="#cb1-208"></a><span class="in">#| label: lst-cnt-cpm</span></span>
<span id="cb1-209"><a href="#cb1-209"></a><span class="in">#| lst-label: lst-cnt-cpm</span></span>
<span id="cb1-210"><a href="#cb1-210"></a><span class="in">#| eval: false</span></span>
<span id="cb1-211"><a href="#cb1-211"></a><span class="in">#| code-summary: Code to make the z-score matrix from the discovery dataset</span></span>
<span id="cb1-212"><a href="#cb1-212"></a><span class="in">#| lst-cap: Code to make the z-score matrix from the discovery dataset</span></span>
<span id="cb1-213"><a href="#cb1-213"></a><span class="in"># CPM based on edgeR TMM  (NB, d isa "DGEList")</span></span>
<span id="cb1-214"><a href="#cb1-214"></a><span class="in">dt.logcpm2&lt;-merge(</span></span>
<span id="cb1-215"><a href="#cb1-215"></a><span class="in">  data.table(`geneName`=rownames(d2),cpm(d2,log=T)) %&gt;% </span></span>
<span id="cb1-216"><a href="#cb1-216"></a><span class="in">    melt.data.table(id.vars=c("geneName"),variable.name="SampleID",value.name="logCPM"),</span></span>
<span id="cb1-217"><a href="#cb1-217"></a><span class="in">  dt.samples[Type==my.type,.(SampleID,GA,Condition,Group=paste(GA,Condition,sep="-"))]</span></span>
<span id="cb1-218"><a href="#cb1-218"></a><span class="in">  )</span></span>
<span id="cb1-219"><a href="#cb1-219"></a></span>
<span id="cb1-220"><a href="#cb1-220"></a><span class="in">dt.cpmZ=merge(dt.logcpm2,</span></span>
<span id="cb1-221"><a href="#cb1-221"></a><span class="in">              dt.logcpm2[Condition=="Control",.(Mean=mean(logCPM),SD=sd(logCPM)),.(GA,geneName)]</span></span>
<span id="cb1-222"><a href="#cb1-222"></a><span class="in">      ,by=c("GA","geneName")</span></span>
<span id="cb1-223"><a href="#cb1-223"></a><span class="in">      )[,.(Group,GA,Condition,SampleID,geneName,logCPM,logCPMZ=(logCPM-Mean)/SD)]</span></span>
<span id="cb1-224"><a href="#cb1-224"></a><span class="in">#save(dt.cpmZ,file=my.cpmZ.RData) # cache to use it later</span></span>
<span id="cb1-225"><a href="#cb1-225"></a><span class="in">```</span></span>
<span id="cb1-226"><a href="#cb1-226"></a></span>
<span id="cb1-227"><a href="#cb1-227"></a><span class="fu">### Univariate logistic regression {#sec-logreg}</span></span>
<span id="cb1-228"><a href="#cb1-228"></a></span>
<span id="cb1-229"><a href="#cb1-229"></a>Using the binary outcomes, (i.e. case and control status) as dependant variables and the z-scores as independent variables, we applied a generalised linear model for each of the 15,150 genes using the <span class="in">`glm`</span> function of R <span class="in">`stat`</span> package (v4.0.3). Both the <span class="co">[</span><span class="ot">Akaike information criterion (AIC)</span><span class="co">](https://en.wikipedia.org/wiki/Akaike_information_criterion)</span> and the <span class="co">[</span><span class="ot">Bayesian information criterion (BIC)</span><span class="co">](https://en.wikipedia.org/wiki/Bayesian_information_criterion)</span> were obtained from the corresponding univariable model and the area under the <span class="co">[</span><span class="ot">ROC curve</span><span class="co">](https://en.wikipedia.org/wiki/Receiver_operating_characteristic)</span> (AUC) was calculated using the <span class="co">[</span><span class="ot">`pROC`</span><span class="co">](https://cran.r-project.org/web/packages/pROC/)</span> package (v1.17.0.1). The p-values were calculated against the null hypothesis that is the odds ratio is equal to 1 and they were adjusted for multiple comparisons using <span class="co">[</span><span class="ot">Benjamini and Hochberg method</span><span class="co">](https://en.wikipedia.org/wiki/False_discovery_rate#BH_procedure)</span>. The distribution of the p-values was tested against a uniform distribution using one-sample <span class="co">[</span><span class="ot">Kolmogorov-Smirnov test</span><span class="co">](https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test)</span>, which is further explained in @sec-fig2a.</span>
<span id="cb1-230"><a href="#cb1-230"></a></span>
<span id="cb1-231"><a href="#cb1-231"></a><span class="in">```{r logreg}</span></span>
<span id="cb1-232"><a href="#cb1-232"></a><span class="in">#| label: lst-logreg</span></span>
<span id="cb1-233"><a href="#cb1-233"></a><span class="in">#| lst-label: lst-logreg</span></span>
<span id="cb1-234"><a href="#cb1-234"></a><span class="in">#| eval: false</span></span>
<span id="cb1-235"><a href="#cb1-235"></a><span class="in">#| code-summary: Code for univariate logisitic regression</span></span>
<span id="cb1-236"><a href="#cb1-236"></a><span class="in">#| lst-cap: Code for univariate logisitic regression</span></span>
<span id="cb1-237"><a href="#cb1-237"></a><span class="in">this.mat&lt;-dt.cpmZ[,.(SampleID,geneName,logCPMZ,y=ifelse(Condition=="Case",1,0))] %&gt;% </span></span>
<span id="cb1-238"><a href="#cb1-238"></a><span class="in">  dcast.data.table(SampleID+y~geneName,value.var="logCPMZ") %&gt;% </span></span>
<span id="cb1-239"><a href="#cb1-239"></a><span class="in">  as.matrix(rownames="SampleID") # logCPMZ</span></span>
<span id="cb1-240"><a href="#cb1-240"></a></span>
<span id="cb1-241"><a href="#cb1-241"></a><span class="in">for(my.GA in names(li.GA)[1:3]){ # for each 12wk, 20wk and 28wkGA</span></span>
<span id="cb1-242"><a href="#cb1-242"></a><span class="in">  this.samples&lt;-colnames(dds.f2)[colData(dds.f2)$GA==my.GA]</span></span>
<span id="cb1-243"><a href="#cb1-243"></a><span class="in">  foo&lt;-list() # per GA</span></span>
<span id="cb1-244"><a href="#cb1-244"></a><span class="in">  for(my.ID in rownames(dds.f2)){</span></span>
<span id="cb1-245"><a href="#cb1-245"></a><span class="in">      #message(paste(my.GA,my.ID))</span></span>
<span id="cb1-246"><a href="#cb1-246"></a><span class="in">      df.mat&lt;-this.mat[this.samples,c("y",my.ID)] %&gt;% as.data.frame</span></span>
<span id="cb1-247"><a href="#cb1-247"></a><span class="in">      my.model&lt;-glm(y ~., data=df.mat, family="binomial")</span></span>
<span id="cb1-248"><a href="#cb1-248"></a><span class="in">      #oddsratio::or_glm(data=df.mat, model=my.model, incr=list(OID20239=1))</span></span>
<span id="cb1-249"><a href="#cb1-249"></a></span>
<span id="cb1-250"><a href="#cb1-250"></a><span class="in">      # Log Odds Ratio &amp; p-values &amp; CI</span></span>
<span id="cb1-251"><a href="#cb1-251"></a><span class="in">      #my.model %&gt;% summary</span></span>
<span id="cb1-252"><a href="#cb1-252"></a><span class="in">      #coef(summary(my.model)) # isa 'matrix'</span></span>
<span id="cb1-253"><a href="#cb1-253"></a><span class="in">      foo1&lt;-as.data.frame(coef(summary(my.model)))</span></span>
<span id="cb1-254"><a href="#cb1-254"></a></span>
<span id="cb1-255"><a href="#cb1-255"></a><span class="in">      #cbind(coef(my.model), confint(my.model)) # Log Odds Ratio &amp; CI (95%)</span></span>
<span id="cb1-256"><a href="#cb1-256"></a><span class="in">      #exp(cbind(coef(my.model), confint(my.model))) # Odds Ratio &amp; CI (95%)</span></span>
<span id="cb1-257"><a href="#cb1-257"></a><span class="in">      foo2&lt;-as.data.frame(exp(cbind(coef(my.model), confint(my.model)))) # Odds Ratio &amp; CI (95%)</span></span>
<span id="cb1-258"><a href="#cb1-258"></a></span>
<span id="cb1-259"><a href="#cb1-259"></a><span class="in">      foo3&lt;-cbind(my.ID,cbind(foo1,foo2)[2,]) %&gt;% data.table </span></span>
<span id="cb1-260"><a href="#cb1-260"></a><span class="in">      colnames(foo3)&lt;-c("gene_id","log_odds","se","zval","pval","odds","odds_lo","odds_hi")</span></span>
<span id="cb1-261"><a href="#cb1-261"></a></span>
<span id="cb1-262"><a href="#cb1-262"></a><span class="in">      # ROC &amp; AUC </span></span>
<span id="cb1-263"><a href="#cb1-263"></a><span class="in">      #predict(my.model,type=c("response"))  # predicted probability</span></span>
<span id="cb1-264"><a href="#cb1-264"></a><span class="in">      my.prob&lt;-fitted(my.model) # same as above</span></span>
<span id="cb1-265"><a href="#cb1-265"></a><span class="in">      # no probability in case of NA in the matrix</span></span>
<span id="cb1-266"><a href="#cb1-266"></a><span class="in">      if(nrow(df.mat)!=length(my.prob)){</span></span>
<span id="cb1-267"><a href="#cb1-267"></a><span class="in">          dt.prob&lt;-rbind(</span></span>
<span id="cb1-268"><a href="#cb1-268"></a><span class="in">          data.table(`index`=as.numeric(names(my.prob)),`prob`=my.prob),</span></span>
<span id="cb1-269"><a href="#cb1-269"></a><span class="in">          data.table(`index`=as.numeric(df.mat[,my.ID]%&gt;% is.na %&gt;% which), `prob`=NA)</span></span>
<span id="cb1-270"><a href="#cb1-270"></a><span class="in">          )[order(index)]</span></span>
<span id="cb1-271"><a href="#cb1-271"></a><span class="in">          my.prob&lt;-dt.prob$prob</span></span>
<span id="cb1-272"><a href="#cb1-272"></a><span class="in">      }</span></span>
<span id="cb1-273"><a href="#cb1-273"></a><span class="in">      df.mat$prob&lt;-my.prob</span></span>
<span id="cb1-274"><a href="#cb1-274"></a><span class="in">      my.roc &lt;- pROC::roc(y ~ prob, data = df.mat, quiet=T, ci=TRUE)</span></span>
<span id="cb1-275"><a href="#cb1-275"></a></span>
<span id="cb1-276"><a href="#cb1-276"></a><span class="in">      foo[[my.ID]] &lt;- </span></span>
<span id="cb1-277"><a href="#cb1-277"></a><span class="in">      cbind(</span></span>
<span id="cb1-278"><a href="#cb1-278"></a><span class="in">            foo3,</span></span>
<span id="cb1-279"><a href="#cb1-279"></a><span class="in">            data.table(`auc`=my.roc$ci[2], </span></span>
<span id="cb1-280"><a href="#cb1-280"></a><span class="in">                        `auc_lo`=my.roc$ci[1], </span></span>
<span id="cb1-281"><a href="#cb1-281"></a><span class="in">                        `auc_hi`=my.roc$ci[3],</span></span>
<span id="cb1-282"><a href="#cb1-282"></a><span class="in">                        `AIC`=aic(my.model),</span></span>
<span id="cb1-283"><a href="#cb1-283"></a><span class="in">                        `BIC`=bic(my.model) ) </span></span>
<span id="cb1-284"><a href="#cb1-284"></a><span class="in">      )</span></span>
<span id="cb1-285"><a href="#cb1-285"></a><span class="in">  } # end of for   </span></span>
<span id="cb1-286"><a href="#cb1-286"></a><span class="in">  dl.logregZ[[my.GA]]&lt;-rbindlist(foo)</span></span>
<span id="cb1-287"><a href="#cb1-287"></a></span>
<span id="cb1-288"><a href="#cb1-288"></a><span class="in">  # apply p.adjust</span></span>
<span id="cb1-289"><a href="#cb1-289"></a><span class="in">  for(i in adjust.methods){</span></span>
<span id="cb1-290"><a href="#cb1-290"></a><span class="in">      dl.logregZ[[my.GA]][,paste0("padj.",i):=p.adjust(pval,i)]</span></span>
<span id="cb1-291"><a href="#cb1-291"></a><span class="in">  }</span></span>
<span id="cb1-292"><a href="#cb1-292"></a><span class="in">} # end of for my.GA</span></span>
<span id="cb1-293"><a href="#cb1-293"></a></span>
<span id="cb1-294"><a href="#cb1-294"></a><span class="in"># cache to use it later</span></span>
<span id="cb1-295"><a href="#cb1-295"></a><span class="in">#save(dl.logregZ, file=paste0("RData/dl.logregZ.",my.type,".",my.salmon.index,".RData"))</span></span>
<span id="cb1-296"><a href="#cb1-296"></a><span class="in">```</span></span>
<span id="cb1-297"><a href="#cb1-297"></a></span>
<span id="cb1-298"><a href="#cb1-298"></a><span class="fu">## Selection of core differentially expressed genes {#sec-core-deg}</span></span>
<span id="cb1-299"><a href="#cb1-299"></a></span>
<span id="cb1-300"><a href="#cb1-300"></a>To select a subset of genes that best explains the outcome of samples from the 28wkGA group of the discovery cohort, we used the following four criteria: 1) the p-values from DESeq2, 2) the p-values from edgeR, 3) AIC, and 4) AUC from univariable logistic regressions. Then we selected the top 1% genes for each category (i.e. 151 genes of the lowest p-values from DESeq2 and edgeR, 151 genes having the lowest AIC, and 151 genes having the highest AUC) and constructed a Venn diagram using ggvenn (v0.1.0) R package (52). We selected a total of 17 genes satisfying all the four criteria (i.e. the intersection) and constructed a gene expression matrix (i.e. the 17 genes across the samples from the 28wkGA group of the discovery cohort) using the z-score.</span>
<span id="cb1-301"><a href="#cb1-301"></a></span>
<span id="cb1-302"><a href="#cb1-302"></a><span class="in">```{r set-core-deg}</span></span>
<span id="cb1-303"><a href="#cb1-303"></a><span class="in">#| label: lst-set-core-deg</span></span>
<span id="cb1-304"><a href="#cb1-304"></a><span class="in">#| lst-label: lst-set-core-deg</span></span>
<span id="cb1-305"><a href="#cb1-305"></a><span class="in">#| eval: false</span></span>
<span id="cb1-306"><a href="#cb1-306"></a><span class="in">#| code-summary: Code to find the core DEGs</span></span>
<span id="cb1-307"><a href="#cb1-307"></a><span class="in">#| lst-cap: Code to find the core DEGs</span></span>
<span id="cb1-308"><a href="#cb1-308"></a><span class="in">venn.top1pctZ&lt;-list(`DESeq2`=dl.resLFC[["28wk"]][order(pvalue)][1:151]$gene_id,</span></span>
<span id="cb1-309"><a href="#cb1-309"></a><span class="in">                  `edgeR`=dl.res.edgeR[["28wk"]][order(PValue)][1:151]$gene_id,</span></span>
<span id="cb1-310"><a href="#cb1-310"></a><span class="in">                `AUC`=dl.logregZ[["28wk"]][order(-auc)][1:151]$gene_id,</span></span>
<span id="cb1-311"><a href="#cb1-311"></a><span class="in">                `AIC`=dl.logregZ[["28wk"]][order(AIC)][1:151]$gene_id</span></span>
<span id="cb1-312"><a href="#cb1-312"></a><span class="in">                )</span></span>
<span id="cb1-313"><a href="#cb1-313"></a></span>
<span id="cb1-314"><a href="#cb1-314"></a><span class="in"># the union of all</span></span>
<span id="cb1-315"><a href="#cb1-315"></a><span class="in">dt.venn.top1pctZ&lt;-lapply(names(venn.top1pctZ), function(i) data.table(i,venn.top1pctZ[[i]])) %&gt;% </span></span>
<span id="cb1-316"><a href="#cb1-316"></a><span class="in">  rbindlist %&gt;% </span></span>
<span id="cb1-317"><a href="#cb1-317"></a><span class="in">  dcast.data.table(V2~i, fun=length)</span></span>
<span id="cb1-318"><a href="#cb1-318"></a><span class="in">setnames(dt.venn.top1pctZ,"V2","Gene")</span></span>
<span id="cb1-319"><a href="#cb1-319"></a><span class="in">top1pct.genesZ&lt;-dt.venn.top1pctZ[AIC==1 &amp; AUC==1 &amp; DESeq2==1 &amp; edgeR==1]$Gene</span></span>
<span id="cb1-320"><a href="#cb1-320"></a><span class="in">core17 &lt;- top1pct.genesZ # used in 5-fold CV</span></span>
<span id="cb1-321"><a href="#cb1-321"></a><span class="in">```</span></span>
<span id="cb1-322"><a href="#cb1-322"></a></span>
<span id="cb1-323"><a href="#cb1-323"></a>The Venn diagram for the 17 core DEGs is shown in @sec-fig2c.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>Cookbook for the analysis of cell-free RNA, © Sung Gong (2024)</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/obsgynaecam/cell-free-rna-2024/edit/main/quant.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/obsgynaecam/cell-free-rna-2024/blob/main/quant.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li><li><a href="https://github.com/obsgynaecam/cell-free-rna-2024/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
<p>This book was built with <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>




</body></html>