<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.551">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>cfRNA POP Study - 4&nbsp; Selection of RNA-seq quantification method</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./SI_trimming.html" rel="next">
<link href="./cross_validation.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./SI_select_quant_method.html">Supplementary Text</a></li><li class="breadcrumb-item"><a href="./SI_select_quant_method.html"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Selection of RNA-seq quantification method</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">cfRNA POP Study</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/obsgynaecam/cell-free-rna-2024" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Methods</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./sample.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Sample Information</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./quant.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">RNA-seq data processing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cross_validation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Cross Validation</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Supplementary Text</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./SI_select_quant_method.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Selection of RNA-seq quantification method</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./SI_trimming.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Assessing the effect of trimming the adaptor sequences</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./figure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figures</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./main_figure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Main Figures</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./suppl_figure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Supplementary Text Figures</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#sec-si-rand" id="toc-sec-si-rand" class="nav-link active" data-scroll-target="#sec-si-rand"><span class="header-section-number">4.1</span> Randomly select 100 samples and downsample reads</a></li>
  <li><a href="#sec-si-quant-salmon" id="toc-sec-si-quant-salmon" class="nav-link" data-scroll-target="#sec-si-quant-salmon"><span class="header-section-number">4.2</span> Import downsampled <code>Salmon</code> result</a></li>
  <li><a href="#number-of-samples-by-ga-and-sex-among-the-100-randomly-selected-samples" id="toc-number-of-samples-by-ga-and-sex-among-the-100-randomly-selected-samples" class="nav-link" data-scroll-target="#number-of-samples-by-ga-and-sex-among-the-100-randomly-selected-samples"><span class="header-section-number">4.3</span> Number of samples by GA and sex among the 100 randomly selected samples</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/obsgynaecam/cell-free-rna-2024/edit/main/SI_select_quant_method.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/obsgynaecam/cell-free-rna-2024/blob/main/SI_select_quant_method.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li><li><a href="https://github.com/obsgynaecam/cell-free-rna-2024/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./SI_select_quant_method.html">Supplementary Text</a></li><li class="breadcrumb-item"><a href="./SI_select_quant_method.html"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Selection of RNA-seq quantification method</span></a></li></ol></nav>
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-si-quant-method" class="quarto-section-identifier"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Selection of RNA-seq quantification method</span></span></h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>We assessed several RNA-seq data processing pipelines and selected the most accurate approach based on the performance of predicting fetal sex by measuring the extent of chromosome Y (chrY) gene expression level. This task, however, is rather complicated because 1) there are multiple copies of chrY gene families, such as <em>TSPY</em> which contains ~30 copies, and 2) most chrY genes have close homologues in chromosome X – some chrY genes show ~99% identity with their corresponding X-linked homologues.</p>
<p>The study from Godfrey et al.&nbsp;(1) showed that kallisto (2) performed better than the two most used conventional methods (TopHat2 (3)+ featureCounts (4) and TopHat2 (3) + Cufflinks(5)) in terms of quantifying the expression level of male specific chrY genes precisely. Recently, Srivastava et al.&nbsp;(6) showed improved performance of their new methodology, so called selective alignment (SA) implemented in the salmon software package. This outperformed various existing methods, including kallisto, in both simulated and real RNA-seq datasets. Therefore, we tested the performance of salmon (v1.5.2) (7) with the following two approaches: 1) the mapping-based mode, and 2) the alignment-based mode. For the mapping-based mode, we applied so-called selective alignment option and employed “decoy” sequences – this approach was named as “Salmon (SA mode)”. For the alignment-based mode, we used HISAT2 (v2.2.1) (8), a successor of TopHat2 (3), followed by salmon in the alignment mode – this approach was named as “HiSat2+Salmon”. Firstly, we selected 100 RNA-seq samples randomly and down-sampled the original sequencing reads to 1, 5 and 10 million per sample (Supplementary Text Table 1) using <a href="https://github.com/lh3/seqtk">seqtk</a> (v1.2-r101-dirty) (9). We ran the two pipelines for each set of down-sampled reads, then measured the read counts for protein-coding chrY genes – there were 42 such chrY genes based on the Ensembl annotation v88 (10). If there was at least one read quantified in any of the 42 chrY genes, such a sample was simply predicted as male; otherwise, female. Then we tabulated confusion matrices by the numbers of true-positives (i.e.&nbsp;predicted as male and the true fetal sex being male; TP), true-negatives (i.e.&nbsp;predicted as female as the true fetal sex being female; TN), false-positives (i.e.&nbsp;predicted as male but the true fetal sex being female; FP), and false-negative (i.e.&nbsp;predicted as female but the true fetal sex being male; FN). We calculated the precision (also known as, positive predictive value), recall (also known as, true positive rate or sensitivity), and the F1 score as the followings:</p>
<p><span class="math display">\[Precision = TP / (TP +FP)\]</span></p>
<p><span class="math display">\[Recall=TP⁄((TP+FN))\]</span></p>
<p><span class="math display">\[F_1  score=2  (Precision×Recall)/(Precision+Recall)\]</span></p>
<p>Finally, we chose the pipeline based on the higher value of F1 score which is in favour of a smaller number of FP and FN.</p>
<p>For the 42 chrY genes, we found that a smaller number of reads was quantified from the female samples when “Salmon (SA mode)” was used compared to “HiSat2+Salmon” – this pattern was more noticeable from the samples of advanced gestation (<a href="suppl_figure.html#sec-si-figure-a" class="quarto-xref"><span>Section 7.1</span></a>). Therefore, the number of FP and FN were smaller for “Salmon (SA mode)” (Supplementary Text Table 2), and the F1 score was higher for “Salmon (SA mode)” than “HiSat2+Salmon” (<a href="suppl_figure.html#sec-si-figure-b" class="quarto-xref"><span>Section 7.2</span></a> and Supplementary Text Table 2). Based on the 10M down-sampled RNA-seq samples, there were 24 and 16 chrY genes quantified as having at least one read from the 100 randomly selected samples, by “HiSat2+Salmon” and “Salmon (SA mode)”, respectively (<a href="suppl_figure.html#sec-si-figure-c" class="quarto-xref"><span>Section 7.3</span></a>).</p>
<section id="sec-si-rand" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="sec-si-rand"><span class="header-section-number">4.1</span> Randomly select 100 samples and downsample reads</h2>
<div id="lst-si-downsample" class="bash listing quarto-float">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-si-downsample-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;4.1: downsampling command using <code>bash</code> and <a href="https://github.com/lh3/seqtk"><code>seqtk</code></a>
</figcaption>
<div aria-describedby="lst-si-downsample-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sourceCode" id="lst-si-downsample"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="lst-si-downsample-1"><a href="#lst-si-downsample-1" aria-hidden="true" tabindex="-1"></a><span class="co">###########################</span></span>
<span id="lst-si-downsample-2"><a href="#lst-si-downsample-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Random sampling for 100 #</span></span>
<span id="lst-si-downsample-3"><a href="#lst-si-downsample-3" aria-hidden="true" tabindex="-1"></a><span class="co">###########################</span></span>
<span id="lst-si-downsample-4"><a href="#lst-si-downsample-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">ls</span> ~/data/fastq/SLX-ILM-Plasma2021/<span class="pp">*</span>r_1.fq.gz<span class="kw">`;</span> <span class="cf">do</span> <span class="va">BARCODE</span><span class="op">=</span><span class="kw">`</span><span class="bu">echo</span> <span class="va">$i</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d</span>/ <span class="at">-f7</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d.</span> <span class="at">-f2</span><span class="kw">`;</span> <span class="bu">echo</span> <span class="va">$BARCODE</span><span class="kw">;</span><span class="cf">done</span> <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-v</span> PPC <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="at">-c</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'$1==4{print $0}'</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-R</span> <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 100 <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'{print $2}'</span> <span class="op">&gt;</span> ~/results/SLX-ILM-Plasma2021.Homo_sapiens.v1/random.sample.100.txt</span>
<span id="lst-si-downsample-5"><a href="#lst-si-downsample-5" aria-hidden="true" tabindex="-1"></a><span class="fu">join</span> <span class="at">-t</span> <span class="st">","</span> <span class="op">&lt;(</span><span class="fu">sort</span> ~/results/SLX-ILM-Plasma2021.Homo_sapiens.v1/random.sample.100.txt<span class="op">)</span> <span class="op">&lt;(</span><span class="fu">sort</span> <span class="at">-t</span> <span class="st">","</span> <span class="at">-k1,1</span> ~/results/SLX-ILM-Plasma2021.Homo_sapiens.v1/FASTQ.list.r1.txt<span class="op">)</span> <span class="op">&gt;</span> ~/results/SLX-ILM-Plasma2021.Homo_sapiens.v1/random.sample.100.FASTQ.list.r1.txt</span>
<span id="lst-si-downsample-6"><a href="#lst-si-downsample-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-si-downsample-7"><a href="#lst-si-downsample-7" aria-hidden="true" tabindex="-1"></a><span class="co">######################################</span></span>
<span id="lst-si-downsample-8"><a href="#lst-si-downsample-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Downsampling to 1M for 100 samples #</span></span>
<span id="lst-si-downsample-9"><a href="#lst-si-downsample-9" aria-hidden="true" tabindex="-1"></a><span class="co">######################################</span></span>
<span id="lst-si-downsample-10"><a href="#lst-si-downsample-10" aria-hidden="true" tabindex="-1"></a> <span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">awk</span> <span class="at">-F</span><span class="st">","</span> <span class="st">'{print $2}'</span> ~/results/SLX-ILM-Plasma2021.Homo_sapiens.v1/random.sample.100.FASTQ.list.r1.txt<span class="kw">`;</span> <span class="cf">do</span> <span class="va">FNAME</span><span class="op">=</span>SLX-ILM-Plasma2021-ds1M.<span class="kw">`</span><span class="bu">echo</span> <span class="va">$i</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d</span>/ <span class="at">-f7</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d.</span> <span class="at">-f2,3,4,5,6,7</span><span class="kw">`;</span> <span class="bu">echo</span> <span class="va">$i</span><span class="kw">;</span> <span class="bu">time</span> seqtk sample <span class="at">-s11</span> <span class="va">$i</span> 250000 <span class="kw">|</span> <span class="fu">gzip</span> <span class="at">-9</span> <span class="op">&gt;</span>  ~/data/fastq/SLX-ILM-Plasma2021-ds1M/<span class="va">$FNAME</span><span class="kw">;</span> <span class="cf">done</span></span>
<span id="lst-si-downsample-11"><a href="#lst-si-downsample-11" aria-hidden="true" tabindex="-1"></a> <span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">awk</span> <span class="at">-F</span><span class="st">","</span> <span class="st">'{print $2}'</span> ~/results/SLX-ILM-Plasma2021.Homo_sapiens.v1/random.sample.100.FASTQ.list.r2.txt<span class="kw">`;</span> <span class="cf">do</span> <span class="va">FNAME</span><span class="op">=</span>SLX-ILM-Plasma2021-ds1M.<span class="kw">`</span><span class="bu">echo</span> <span class="va">$i</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d</span>/ <span class="at">-f7</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d.</span> <span class="at">-f2,3,4,5,6,7</span><span class="kw">`;</span> <span class="bu">echo</span> <span class="va">$i</span><span class="kw">;</span> <span class="bu">time</span> seqtk sample <span class="at">-s11</span> <span class="va">$i</span> 250000 <span class="kw">|</span> <span class="fu">gzip</span> <span class="at">-9</span> <span class="op">&gt;</span>  ~/data/fastq/SLX-ILM-Plasma2021-ds1M/<span class="va">$FNAME</span><span class="kw">;</span> <span class="cf">done</span></span>
<span id="lst-si-downsample-12"><a href="#lst-si-downsample-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-si-downsample-13"><a href="#lst-si-downsample-13" aria-hidden="true" tabindex="-1"></a><span class="co">######################################</span></span>
<span id="lst-si-downsample-14"><a href="#lst-si-downsample-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Downsampling to 5M for 100 samples #</span></span>
<span id="lst-si-downsample-15"><a href="#lst-si-downsample-15" aria-hidden="true" tabindex="-1"></a><span class="co">######################################</span></span>
<span id="lst-si-downsample-16"><a href="#lst-si-downsample-16" aria-hidden="true" tabindex="-1"></a> <span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">awk</span> <span class="at">-F</span><span class="st">","</span> <span class="st">'{print $2}'</span> ~/results/SLX-ILM-Plasma2021.Homo_sapiens.v1/random.sample.100.FASTQ.list.r1.txt<span class="kw">`;</span> <span class="cf">do</span> <span class="va">FNAME</span><span class="op">=</span>SLX-ILM-Plasma2021-ds5M.<span class="kw">`</span><span class="bu">echo</span> <span class="va">$i</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d</span>/ <span class="at">-f7</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d.</span> <span class="at">-f2,3,4,5,6,7</span><span class="kw">`;</span> <span class="bu">echo</span> <span class="va">$i</span><span class="kw">;</span> <span class="bu">time</span> seqtk sample <span class="at">-s11</span> <span class="va">$i</span> 1250000 <span class="kw">|</span> <span class="fu">gzip</span> <span class="at">-9</span> <span class="op">&gt;</span>  ~/data/fastq/SLX-ILM-Plasma2021-ds5M/<span class="va">$FNAME</span><span class="kw">;</span> <span class="cf">done</span></span>
<span id="lst-si-downsample-17"><a href="#lst-si-downsample-17" aria-hidden="true" tabindex="-1"></a> <span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">awk</span> <span class="at">-F</span><span class="st">","</span> <span class="st">'{print $2}'</span> ~/results/SLX-ILM-Plasma2021.Homo_sapiens.v1/random.sample.100.FASTQ.list.r2.txt<span class="kw">`;</span> <span class="cf">do</span> <span class="va">FNAME</span><span class="op">=</span>SLX-ILM-Plasma2021-ds5M.<span class="kw">`</span><span class="bu">echo</span> <span class="va">$i</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d</span>/ <span class="at">-f7</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d.</span> <span class="at">-f2,3,4,5,6,7</span><span class="kw">`;</span> <span class="bu">echo</span> <span class="va">$i</span><span class="kw">;</span> <span class="bu">time</span> seqtk sample <span class="at">-s11</span> <span class="va">$i</span> 1250000 <span class="kw">|</span> <span class="fu">gzip</span> <span class="at">-9</span> <span class="op">&gt;</span>  ~/data/fastq/SLX-ILM-Plasma2021-ds5M/<span class="va">$FNAME</span><span class="kw">;</span> <span class="cf">done</span></span>
<span id="lst-si-downsample-18"><a href="#lst-si-downsample-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-si-downsample-19"><a href="#lst-si-downsample-19" aria-hidden="true" tabindex="-1"></a><span class="co">#######################################</span></span>
<span id="lst-si-downsample-20"><a href="#lst-si-downsample-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Downsampling to 10M for 100 samples #</span></span>
<span id="lst-si-downsample-21"><a href="#lst-si-downsample-21" aria-hidden="true" tabindex="-1"></a><span class="co">###################]###################</span></span>
<span id="lst-si-downsample-22"><a href="#lst-si-downsample-22" aria-hidden="true" tabindex="-1"></a> <span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">awk</span> <span class="at">-F</span><span class="st">","</span> <span class="st">'{print $2}'</span> ~/results/SLX-ILM-Plasma2021.Homo_sapiens.v1/random.sample.100.FASTQ.list.r1.txt<span class="kw">`;</span> <span class="cf">do</span> <span class="va">FNAME</span><span class="op">=</span>SLX-ILM-Plasma2021-ds10M.<span class="kw">`</span><span class="bu">echo</span> <span class="va">$i</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d</span>/ <span class="at">-f7</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d.</span> <span class="at">-f2,3,4,5,6,7</span><span class="kw">`;</span> <span class="bu">echo</span> <span class="va">$i</span><span class="kw">;</span> <span class="bu">time</span> seqtk sample <span class="at">-s11</span> <span class="va">$i</span> 2500000 <span class="kw">|</span> <span class="fu">gzip</span> <span class="at">-9</span> <span class="op">&gt;</span>  ~/data/fastq/SLX-ILM-Plasma2021-ds10M/<span class="va">$FNAME</span><span class="kw">;</span> <span class="cf">done</span></span>
<span id="lst-si-downsample-23"><a href="#lst-si-downsample-23" aria-hidden="true" tabindex="-1"></a> <span class="cf">for</span> i <span class="kw">in</span> <span class="kw">`</span><span class="fu">awk</span> <span class="at">-F</span><span class="st">","</span> <span class="st">'{print $2}'</span> ~/results/SLX-ILM-Plasma2021.Homo_sapiens.v1/random.sample.100.FASTQ.list.r2.txt<span class="kw">`;</span> <span class="cf">do</span> <span class="va">FNAME</span><span class="op">=</span>SLX-ILM-Plasma2021-ds10M.<span class="kw">`</span><span class="bu">echo</span> <span class="va">$i</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d</span>/ <span class="at">-f7</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d.</span> <span class="at">-f2,3,4,5,6,7</span><span class="kw">`;</span> <span class="bu">echo</span> <span class="va">$i</span><span class="kw">;</span> <span class="bu">time</span> seqtk sample <span class="at">-s11</span> <span class="va">$i</span> 2500000 <span class="kw">|</span> <span class="fu">gzip</span> <span class="at">-9</span> <span class="op">&gt;</span>  ~/data/fastq/SLX-ILM-Plasma2021-ds10M/<span class="va">$FNAME</span><span class="kw">;</span> <span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</figure>
</div>
</section>
<section id="sec-si-quant-salmon" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="sec-si-quant-salmon"><span class="header-section-number">4.2</span> Import downsampled <code>Salmon</code> result</h2>
<p>The code below internally calls the <code>prep_salmon</code> function defined in the <code>_libs</code> folder.</p>
<div class="cell">
<details class="code-fold">
<summary>Code to parse downsampled <code>Salmon</code> results</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>dl.chrY.TPM<span class="ot">&lt;-</span><span class="fu">list</span>()</span>
<span id="cb1-2"><a href="#cb1-2"></a>all.ds<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"1M"</span>,<span class="st">"5M"</span>,<span class="st">"10M"</span>)</span>
<span id="cb1-3"><a href="#cb1-3"></a>dl.chrY.TPM<span class="ot">&lt;-</span><span class="fu">lapply</span>(all.ds, <span class="cf">function</span>(my.ds){</span>
<span id="cb1-4"><a href="#cb1-4"></a>  <span class="fu">message</span>(my.ds)</span>
<span id="cb1-5"><a href="#cb1-5"></a>  my.RData<span class="ot">&lt;-</span><span class="fu">paste0</span>(<span class="st">"RData/li.TPM."</span>,my.ds,<span class="st">".RData"</span>)</span>
<span id="cb1-6"><a href="#cb1-6"></a>  <span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(my.RData)){ <span class="co"># read the local output of Salmon </span></span>
<span id="cb1-7"><a href="#cb1-7"></a>    my.slx<span class="ot">&lt;-</span><span class="fu">ifelse</span>(my.ds<span class="sc">==</span><span class="st">"1M"</span>,<span class="fu">paste0</span>(<span class="st">"SLX-ILM-Plasma2021-ds"</span>,my.ds,<span class="st">".Homo_sapiens.v2"</span>),<span class="fu">paste0</span>(<span class="st">"SLX-ILM-Plasma2021-ds"</span>,my.ds,<span class="st">".Homo_sapiens.v1"</span>))</span>
<span id="cb1-8"><a href="#cb1-8"></a>    li.TPM<span class="ot">&lt;-</span><span class="fu">prep_ds_salmon</span>(my.slx) <span class="co"># set li.TPM. See `_libs/local.R` how `li.TPM` was set up</span></span>
<span id="cb1-9"><a href="#cb1-9"></a>    <span class="fu">save</span>(li.TPM, <span class="at">file=</span><span class="fu">paste0</span>(<span class="st">"RData/li.TPM."</span>,my.ds,<span class="st">".RData"</span>))</span>
<span id="cb1-10"><a href="#cb1-10"></a>    <span class="fu">message</span>(<span class="st">"li.TPM saved"</span>)</span>
<span id="cb1-11"><a href="#cb1-11"></a>  }</span>
<span id="cb1-12"><a href="#cb1-12"></a>  <span class="fu">load</span>(my.RData) <span class="sc">%&gt;%</span> system.time</span>
<span id="cb1-13"><a href="#cb1-13"></a></span>
<span id="cb1-14"><a href="#cb1-14"></a>  dt.chrY.TPM<span class="ot">&lt;-</span><span class="fu">rbind</span>(</span>
<span id="cb1-15"><a href="#cb1-15"></a>    <span class="fu">data.table</span>(</span>
<span id="cb1-16"><a href="#cb1-16"></a>      <span class="st">`</span><span class="at">Type</span><span class="st">`</span><span class="ot">=</span><span class="st">"Salmon (SA mode)"</span>,</span>
<span id="cb1-17"><a href="#cb1-17"></a>      <span class="st">`</span><span class="at">SampleID</span><span class="st">`</span><span class="ot">=</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> names,</span>
<span id="cb1-18"><a href="#cb1-18"></a>      <span class="st">`</span><span class="at">Counts</span><span class="st">`</span><span class="ot">=</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums, </span>
<span id="cb1-19"><a href="#cb1-19"></a>      <span class="st">`</span><span class="at">TPM</span><span class="st">`</span><span class="ot">=</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"TPM"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums </span>
<span id="cb1-20"><a href="#cb1-20"></a>      ),</span>
<span id="cb1-21"><a href="#cb1-21"></a>    <span class="fu">data.table</span>(</span>
<span id="cb1-22"><a href="#cb1-22"></a>      <span class="st">`</span><span class="at">Type</span><span class="st">`</span><span class="ot">=</span><span class="st">"HiSat2+Salmon"</span>,</span>
<span id="cb1-23"><a href="#cb1-23"></a>      <span class="st">`</span><span class="at">SampleID</span><span class="st">`</span><span class="ot">=</span>li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> names,</span>
<span id="cb1-24"><a href="#cb1-24"></a>      <span class="st">`</span><span class="at">Counts</span><span class="st">`</span><span class="ot">=</span>li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums,</span>
<span id="cb1-25"><a href="#cb1-25"></a>      <span class="st">`</span><span class="at">TPM</span><span class="st">`</span><span class="ot">=</span>li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"TPM"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums </span>
<span id="cb1-26"><a href="#cb1-26"></a>      )</span>
<span id="cb1-27"><a href="#cb1-27"></a>  )</span>
<span id="cb1-28"><a href="#cb1-28"></a>  <span class="co"># add Sex and GA</span></span>
<span id="cb1-29"><a href="#cb1-29"></a>  <span class="fu">merge</span>(dt.chrY.TPM, dt.samples[,.(SampleID,GA,Sex,pn_female)])</span>
<span id="cb1-30"><a href="#cb1-30"></a>})</span>
<span id="cb1-31"><a href="#cb1-31"></a><span class="fu">names</span>(dl.chrY.TPM)<span class="ot">&lt;-</span>all.ds</span>
<span id="cb1-32"><a href="#cb1-32"></a></span>
<span id="cb1-33"><a href="#cb1-33"></a>dl.stat<span class="ot">&lt;-</span><span class="fu">lapply</span>(dl.chrY.TPM, <span class="cf">function</span>(i) i[,.(.N,<span class="at">SumCount=</span><span class="fu">sum</span>(Counts),<span class="at">SumTPM=</span><span class="fu">sum</span>(TPM),<span class="at">MeanCount=</span><span class="fu">mean</span>(Counts),<span class="at">MeanTPM=</span><span class="fu">mean</span>(TPM)),.(GA,Sex,Type)][<span class="fu">order</span>(GA,Sex)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="number-of-samples-by-ga-and-sex-among-the-100-randomly-selected-samples" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="number-of-samples-by-ga-and-sex-among-the-100-randomly-selected-samples"><span class="header-section-number">4.3</span> Number of samples by GA and sex among the 100 randomly selected samples</h2>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="co">#! label: ds100</span></span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb2-3"><a href="#cb2-3"></a>dt.samples<span class="ot">&lt;-</span><span class="fu">fread</span>(<span class="st">"static/R/data/dt.samples.csv"</span>) <span class="co"># n=755</span></span>
<span id="cb2-4"><a href="#cb2-4"></a>rand<span class="fl">.100</span><span class="ot">&lt;-</span><span class="fu">fread</span>(<span class="st">"static/R/data/random.sample.100.txt"</span>, <span class="at">header=</span>F) <span class="co"># n=100</span></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="fu">xtabs</span>(<span class="sc">~</span>GA<span class="sc">+</span>Sex, <span class="at">data=</span>dt.samples[SampleID <span class="sc">%in%</span> rand<span class="fl">.100</span><span class="sc">$</span>V1]) <span class="sc">%&gt;%</span> addmargins</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>      Sex
GA     Female Male Sum
  12wk     17    5  22
  20wk     18    7  25
  28wk     19   15  34
  36wk     11    8  19
  Sum      65   35 100</code></pre>
</div>
</div>


<!-- -->

</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/ObsGynaeCam\.github\.io\/cell-free-rna-2024");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./cross_validation.html" class="pagination-link" aria-label="Cross Validation">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Cross Validation</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./SI_trimming.html" class="pagination-link" aria-label="Assessing the effect of trimming the adaptor sequences">
        <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Assessing the effect of trimming the adaptor sequences</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb4" data-shortcodes="false"><pre class="sourceCode numberSource markdown number-lines code-with-copy"><code class="sourceCode markdown"><span id="cb4-1"><a href="#cb4-1"></a><span class="fu"># Selection of RNA-seq quantification method {#sec-si-quant-method}</span></span>
<span id="cb4-2"><a href="#cb4-2"></a></span>
<span id="cb4-3"><a href="#cb4-3"></a>We assessed several RNA-seq data processing pipelines and selected the most accurate approach based on the performance of predicting fetal sex by measuring the extent of chromosome Y (chrY) gene expression level. This task, however, is rather complicated because 1) there are multiple copies of chrY gene families, such as _TSPY_ which contains ~30 copies, and 2) most chrY genes have close homologues in chromosome X – some chrY genes show ~99% identity with their corresponding X-linked homologues.</span>
<span id="cb4-4"><a href="#cb4-4"></a></span>
<span id="cb4-5"><a href="#cb4-5"></a>The study from Godfrey et al. (1) showed that kallisto (2) performed better than the two most used conventional methods (TopHat2 (3)+ featureCounts (4) and TopHat2 (3) + Cufflinks(5)) in terms of quantifying the expression level of male specific chrY genes precisely. Recently, Srivastava et al. (6) showed improved performance of their new methodology, so called selective alignment (SA) implemented in the salmon software package. This outperformed various existing methods, including kallisto, in both simulated and real RNA-seq datasets. Therefore, we tested the performance of salmon (v1.5.2) (7) with the following two approaches: 1) the mapping-based mode, and 2) the alignment-based mode. For the mapping-based mode, we applied so-called selective alignment option and employed “decoy” sequences – this approach was named as “Salmon (SA mode)”. For the alignment-based mode, we used HISAT2 (v2.2.1) (8), a successor of TopHat2 (3), followed by salmon in the alignment mode – this approach was named as “HiSat2+Salmon”. Firstly, we selected 100 RNA-seq samples randomly and down-sampled the original sequencing reads to 1, 5 and 10 million per sample (Supplementary Text Table 1) using <span class="co">[</span><span class="ot">seqtk</span><span class="co">](https://github.com/lh3/seqtk)</span> (v1.2-r101-dirty) (9). We ran the two pipelines for each set of down-sampled reads, then measured the read counts for protein-coding chrY genes – there were 42 such chrY genes based on the Ensembl annotation v88 (10). If there was at least one read quantified in any of the 42 chrY genes, such a sample was simply predicted as male; otherwise, female. Then we tabulated confusion matrices by the numbers of true-positives (i.e. predicted as male and the true fetal sex being male; TP), true-negatives (i.e. predicted as female as the true fetal sex being female; TN), false-positives (i.e. predicted as male but the true fetal sex being female; FP), and false-negative (i.e. predicted as female but the true fetal sex being male; FN). We calculated the precision (also known as, positive predictive value), recall (also known as, true positive rate or sensitivity), and the F1 score as the followings:</span>
<span id="cb4-6"><a href="#cb4-6"></a></span>
<span id="cb4-7"><a href="#cb4-7"></a>$$Precision = TP / (TP +FP)$$</span>
<span id="cb4-8"><a href="#cb4-8"></a></span>
<span id="cb4-9"><a href="#cb4-9"></a>$$Recall=TP⁄((TP+FN))$$</span>
<span id="cb4-10"><a href="#cb4-10"></a></span>
<span id="cb4-11"><a href="#cb4-11"></a>$$F_1  score=2  (Precision×Recall)/(Precision+Recall)$$</span>
<span id="cb4-12"><a href="#cb4-12"></a></span>
<span id="cb4-13"><a href="#cb4-13"></a>Finally, we chose the pipeline based on the higher value of F1 score which is in favour of a smaller number of FP and FN.</span>
<span id="cb4-14"><a href="#cb4-14"></a></span>
<span id="cb4-15"><a href="#cb4-15"></a>For the 42 chrY genes, we found that a smaller number of reads was quantified from the female samples when “Salmon (SA mode)” was used compared to “HiSat2+Salmon” – this pattern was more noticeable from the samples of advanced gestation (@sec-si-figure-a). Therefore, the number of FP and FN were smaller for “Salmon (SA mode)” (Supplementary Text Table 2), and the F1 score was higher for “Salmon (SA mode)” than “HiSat2+Salmon” (@sec-si-figure-b and Supplementary Text Table 2). Based on the 10M down-sampled RNA-seq samples, there were 24 and 16 chrY genes quantified as having at least one read from the 100 randomly selected samples, by “HiSat2+Salmon” and “Salmon (SA mode)”, respectively (@sec-si-figure-c).</span>
<span id="cb4-16"><a href="#cb4-16"></a></span>
<span id="cb4-17"><a href="#cb4-17"></a><span class="fu">## Randomly select 100 samples and downsample reads {#sec-si-rand}</span></span>
<span id="cb4-18"><a href="#cb4-18"></a></span>
<span id="cb4-19"><a href="#cb4-19"></a>``<span class="in">`{#lst-si-downsample .bash lst-cap="downsampling command using `</span>bash<span class="in">` and [`</span>seqtk`](https://github.com/lh3/seqtk)"}</span>
<span id="cb4-20"><a href="#cb4-20"></a>{{&lt; include static/shell/downsample.sh &gt;}}</span>
<span id="cb4-21"><a href="#cb4-21"></a><span class="in">```</span></span>
<span id="cb4-22"><a href="#cb4-22"></a></span>
<span id="cb4-23"><a href="#cb4-23"></a><span class="in">## Import downsampled `Salmon` result {#sec-si-quant-salmon}</span></span>
<span id="cb4-24"><a href="#cb4-24"></a><span class="in">The code below internally calls the `prep_salmon` function defined in the `_libs` folder.</span></span>
<span id="cb4-25"><a href="#cb4-25"></a></span>
<span id="cb4-26"><a href="#cb4-26"></a><span class="in">```{r load-ds-salmon}</span></span>
<span id="cb4-27"><a href="#cb4-27"></a><span class="in">#| label: load-ds-salmon</span></span>
<span id="cb4-28"><a href="#cb4-28"></a><span class="in">#| eval: false</span></span>
<span id="cb4-29"><a href="#cb4-29"></a><span class="in">#| code-summary: "Code to parse downsampled `Salmon` results"</span></span>
<span id="cb4-30"><a href="#cb4-30"></a><span class="in">dl.chrY.TPM&lt;-list()</span></span>
<span id="cb4-31"><a href="#cb4-31"></a><span class="in">all.ds&lt;-c("1M","5M","10M")</span></span>
<span id="cb4-32"><a href="#cb4-32"></a><span class="in">dl.chrY.TPM&lt;-lapply(all.ds, function(my.ds){</span></span>
<span id="cb4-33"><a href="#cb4-33"></a><span class="in">  message(my.ds)</span></span>
<span id="cb4-34"><a href="#cb4-34"></a><span class="in">  my.RData&lt;-paste0("RData/li.TPM.",my.ds,".RData")</span></span>
<span id="cb4-35"><a href="#cb4-35"></a><span class="in">  if(!file.exists(my.RData)){ # read the local output of Salmon </span></span>
<span id="cb4-36"><a href="#cb4-36"></a><span class="in">    my.slx&lt;-ifelse(my.ds=="1M",paste0("SLX-ILM-Plasma2021-ds",my.ds,".Homo_sapiens.v2"),paste0("SLX-ILM-Plasma2021-ds",my.ds,".Homo_sapiens.v1"))</span></span>
<span id="cb4-37"><a href="#cb4-37"></a><span class="in">    li.TPM&lt;-prep_ds_salmon(my.slx) # set li.TPM. See `_libs/local.R` how `li.TPM` was set up</span></span>
<span id="cb4-38"><a href="#cb4-38"></a><span class="in">    save(li.TPM, file=paste0("RData/li.TPM.",my.ds,".RData"))</span></span>
<span id="cb4-39"><a href="#cb4-39"></a><span class="in">    message("li.TPM saved")</span></span>
<span id="cb4-40"><a href="#cb4-40"></a><span class="in">  }</span></span>
<span id="cb4-41"><a href="#cb4-41"></a><span class="in">  load(my.RData) %&gt;% system.time</span></span>
<span id="cb4-42"><a href="#cb4-42"></a></span>
<span id="cb4-43"><a href="#cb4-43"></a><span class="in">  dt.chrY.TPM&lt;-rbind(</span></span>
<span id="cb4-44"><a href="#cb4-44"></a><span class="in">    data.table(</span></span>
<span id="cb4-45"><a href="#cb4-45"></a><span class="in">      `Type`="Salmon (SA mode)",</span></span>
<span id="cb4-46"><a href="#cb4-46"></a><span class="in">      `SampleID`=li.TPM[["Salmon"]][["counts"]][this.gene,] %&gt;% colSums %&gt;% names,</span></span>
<span id="cb4-47"><a href="#cb4-47"></a><span class="in">      `Counts`=li.TPM[["Salmon"]][["counts"]][this.gene,] %&gt;% colSums, </span></span>
<span id="cb4-48"><a href="#cb4-48"></a><span class="in">      `TPM`=li.TPM[["Salmon"]][["TPM"]][this.gene,] %&gt;% colSums </span></span>
<span id="cb4-49"><a href="#cb4-49"></a><span class="in">      ),</span></span>
<span id="cb4-50"><a href="#cb4-50"></a><span class="in">    data.table(</span></span>
<span id="cb4-51"><a href="#cb4-51"></a><span class="in">      `Type`="HiSat2+Salmon",</span></span>
<span id="cb4-52"><a href="#cb4-52"></a><span class="in">      `SampleID`=li.TPM[["Salmon_aln"]][["counts"]][this.gene,] %&gt;% colSums %&gt;% names,</span></span>
<span id="cb4-53"><a href="#cb4-53"></a><span class="in">      `Counts`=li.TPM[["Salmon_aln"]][["counts"]][this.gene,] %&gt;% colSums,</span></span>
<span id="cb4-54"><a href="#cb4-54"></a><span class="in">      `TPM`=li.TPM[["Salmon_aln"]][["TPM"]][this.gene,] %&gt;% colSums </span></span>
<span id="cb4-55"><a href="#cb4-55"></a><span class="in">      )</span></span>
<span id="cb4-56"><a href="#cb4-56"></a><span class="in">  )</span></span>
<span id="cb4-57"><a href="#cb4-57"></a><span class="in">  # add Sex and GA</span></span>
<span id="cb4-58"><a href="#cb4-58"></a><span class="in">  merge(dt.chrY.TPM, dt.samples[,.(SampleID,GA,Sex,pn_female)])</span></span>
<span id="cb4-59"><a href="#cb4-59"></a><span class="in">})</span></span>
<span id="cb4-60"><a href="#cb4-60"></a><span class="in">names(dl.chrY.TPM)&lt;-all.ds</span></span>
<span id="cb4-61"><a href="#cb4-61"></a></span>
<span id="cb4-62"><a href="#cb4-62"></a><span class="in">dl.stat&lt;-lapply(dl.chrY.TPM, function(i) i[,.(.N,SumCount=sum(Counts),SumTPM=sum(TPM),MeanCount=mean(Counts),MeanTPM=mean(TPM)),.(GA,Sex,Type)][order(GA,Sex)])</span></span>
<span id="cb4-63"><a href="#cb4-63"></a><span class="in">```</span></span>
<span id="cb4-64"><a href="#cb4-64"></a></span>
<span id="cb4-65"><a href="#cb4-65"></a><span class="fu">## Number of samples by GA and sex among the 100 randomly selected samples</span></span>
<span id="cb4-66"><a href="#cb4-66"></a><span class="in">```{r ds100}</span></span>
<span id="cb4-67"><a href="#cb4-67"></a><span class="in">#! label: ds100</span></span>
<span id="cb4-68"><a href="#cb4-68"></a><span class="in">library(magrittr)</span></span>
<span id="cb4-69"><a href="#cb4-69"></a><span class="in">dt.samples&lt;-fread("static/R/data/dt.samples.csv") # n=755</span></span>
<span id="cb4-70"><a href="#cb4-70"></a><span class="in">rand.100&lt;-fread("static/R/data/random.sample.100.txt", header=F) # n=100</span></span>
<span id="cb4-71"><a href="#cb4-71"></a><span class="in">xtabs(~GA+Sex, data=dt.samples[SampleID %in% rand.100$V1]) %&gt;% addmargins</span></span>
<span id="cb4-72"><a href="#cb4-72"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>Cookbook for the analysis of cell-free RNA, © Sung Gong (2024)</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/obsgynaecam/cell-free-rna-2024/edit/main/SI_select_quant_method.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/obsgynaecam/cell-free-rna-2024/blob/main/SI_select_quant_method.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li><li><a href="https://github.com/obsgynaecam/cell-free-rna-2024/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
<p>This book was built with <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>




</body></html>