<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.31">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>9&nbsp; Supplementary Text Figures – cfRNA POP Study</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./main_figure.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-1d9754e6b7e1dc15a659fbcd6780067e.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark-ec59717222d4d18488216d07f8bb4c3b.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-1d9754e6b7e1dc15a659fbcd6780067e.css" rel="stylesheet" class="quarto-color-scheme-extra" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-9e5751c9c0e8fbb9471b0f0078eb07d4.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark-f7288d29b21af75f40a82d96691b0efd.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<link href="site_libs/bootstrap/bootstrap-9e5751c9c0e8fbb9471b0f0078eb07d4.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme-extra" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-0VJ9MW56MK"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-0VJ9MW56MK', { 'anonymize_ip': true});
</script>


<meta name="twitter:title" content="9&nbsp; Supplementary Text Figures – cfRNA POP Study">
<meta name="twitter:description" content="">
<meta name="twitter:image" content="https://ObsGynaeCam.github.io/cell-free-rna-2024/static/figure/cfRNA.SI.FigA.png">
<meta name="twitter:image-height" content="1006">
<meta name="twitter:image-width" content="1594">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating quarto-light"><script id="quarto-html-before-body" type="application/javascript">
    const toggleBodyColorMode = (bsSheetEl) => {
      const mode = bsSheetEl.getAttribute("data-mode");
      const bodyEl = window.document.querySelector("body");
      if (mode === "dark") {
        bodyEl.classList.add("quarto-dark");
        bodyEl.classList.remove("quarto-light");
      } else {
        bodyEl.classList.add("quarto-light");
        bodyEl.classList.remove("quarto-dark");
      }
    }
    const toggleBodyColorPrimary = () => {
      const bsSheetEl = window.document.querySelector("link#quarto-bootstrap:not([rel=disabled-stylesheet])");
      if (bsSheetEl) {
        toggleBodyColorMode(bsSheetEl);
      }
    }
    const setColorSchemeToggle = (alternate) => {
      const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
      for (let i=0; i < toggles.length; i++) {
        const toggle = toggles[i];
        if (toggle) {
          if (alternate) {
            toggle.classList.add("alternate");
          } else {
            toggle.classList.remove("alternate");
          }
        }
      }
    };
    const toggleColorMode = (alternate) => {
      // Switch the stylesheets
      const primaryStylesheets = window.document.querySelectorAll('link.quarto-color-scheme:not(.quarto-color-alternate)');
      const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
      manageTransitions('#quarto-margin-sidebar .nav-link', false);
      if (alternate) {
        // note: dark is layered on light, we don't disable primary!
        enableStylesheet(alternateStylesheets);
        for (const sheetNode of alternateStylesheets) {
          if (sheetNode.id === "quarto-bootstrap") {
            toggleBodyColorMode(sheetNode);
          }
        }
      } else {
        disableStylesheet(alternateStylesheets);
        enableStylesheet(primaryStylesheets)
        toggleBodyColorPrimary();
      }
      manageTransitions('#quarto-margin-sidebar .nav-link', true);
      // Switch the toggles
      setColorSchemeToggle(alternate)
      // Hack to workaround the fact that safari doesn't
      // properly recolor the scrollbar when toggling (#1455)
      if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
        manageTransitions("body", false);
        window.scrollTo(0, 1);
        setTimeout(() => {
          window.scrollTo(0, 0);
          manageTransitions("body", true);
        }, 40);
      }
    }
    const disableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        stylesheet.rel = 'disabled-stylesheet';
      }
    }
    const enableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        if(stylesheet.rel !== 'stylesheet') { // for Chrome, which will still FOUC without this check
          stylesheet.rel = 'stylesheet';
        }
      }
    }
    const manageTransitions = (selector, allowTransitions) => {
      const els = window.document.querySelectorAll(selector);
      for (let i=0; i < els.length; i++) {
        const el = els[i];
        if (allowTransitions) {
          el.classList.remove('notransition');
        } else {
          el.classList.add('notransition');
        }
      }
    }
    const isFileUrl = () => {
      return window.location.protocol === 'file:';
    }
    const hasAlternateSentinel = () => {
      let styleSentinel = getColorSchemeSentinel();
      if (styleSentinel !== null) {
        return styleSentinel === "alternate";
      } else {
        return false;
      }
    }
    const setStyleSentinel = (alternate) => {
      const value = alternate ? "alternate" : "default";
      if (!isFileUrl()) {
        window.localStorage.setItem("quarto-color-scheme", value);
      } else {
        localAlternateSentinel = value;
      }
    }
    const getColorSchemeSentinel = () => {
      if (!isFileUrl()) {
        const storageValue = window.localStorage.getItem("quarto-color-scheme");
        return storageValue != null ? storageValue : localAlternateSentinel;
      } else {
        return localAlternateSentinel;
      }
    }
    const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
      const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
      const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
      let newTheme = '';
      if(authorPrefersDark) {
        newTheme = isAlternate ? baseTheme : alternateTheme;
      } else {
        newTheme = isAlternate ? alternateTheme : baseTheme;
      }
      const changeGiscusTheme = () => {
        // From: https://github.com/giscus/giscus/issues/336
        const sendMessage = (message) => {
          const iframe = document.querySelector('iframe.giscus-frame');
          if (!iframe) return;
          iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
        }
        sendMessage({
          setConfig: {
            theme: newTheme
          }
        });
      }
      const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
      if (isGiscussLoaded) {
        changeGiscusTheme();
      }
    };
    const authorPrefersDark = false;
    const darkModeDefault = authorPrefersDark;
      document.querySelector('link#quarto-text-highlighting-styles.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
      document.querySelector('link#quarto-bootstrap.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
    let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
    // Dark / light mode switch
    window.quartoToggleColorScheme = () => {
      // Read the current dark / light value
      let toAlternate = !hasAlternateSentinel();
      toggleColorMode(toAlternate);
      setStyleSentinel(toAlternate);
      toggleGiscusIfUsed(toAlternate, darkModeDefault);
      window.dispatchEvent(new Event('resize'));
    };
    // Switch to dark mode if need be
    if (hasAlternateSentinel()) {
      toggleColorMode(true);
    } else {
      toggleColorMode(false);
    }
  </script>

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./figure.html">Figures</a></li><li class="breadcrumb-item"><a href="./suppl_figure.html"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Supplementary Text Figures</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">cfRNA POP Study</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/obsgynaecam/cell-free-rna-2024" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Methods</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./sample.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Sample Information</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./quant.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">RNA-seq Data Processing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cross_validation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Cross Validation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./change_by_gestational_age.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Longitudinal Analysis</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Supplementary Text</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./SI_select_quant_method.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Selection of RNA-seq quantification method</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./SI_trimming.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Assessing the effect of trimming the adaptor sequences</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./SI_combined_model.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">The combined model</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./figure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figures</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./main_figure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Main Figures</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./suppl_figure.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Supplementary Text Figures</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#sec-si-figure1" id="toc-sec-si-figure1" class="nav-link active" data-scroll-target="#sec-si-figure1"><span class="header-section-number">9.1</span> Supplementary Text Figure 1</a>
  <ul class="collapse">
  <li><a href="#sec-si-figure1-a" id="toc-sec-si-figure1-a" class="nav-link" data-scroll-target="#sec-si-figure1-a"><span class="header-section-number">9.1.1</span> Supplementary Text Figure 1A</a></li>
  <li><a href="#sec-si-figure-1b" id="toc-sec-si-figure-1b" class="nav-link" data-scroll-target="#sec-si-figure-1b"><span class="header-section-number">9.1.2</span> Supplementary Text Figure 1B</a></li>
  <li><a href="#sec-si-figure-1c" id="toc-sec-si-figure-1c" class="nav-link" data-scroll-target="#sec-si-figure-1c"><span class="header-section-number">9.1.3</span> Supplementary Text Figure 1C</a></li>
  <li><a href="#sec-si-figure-1d" id="toc-sec-si-figure-1d" class="nav-link" data-scroll-target="#sec-si-figure-1d"><span class="header-section-number">9.1.4</span> Supplementary Text Figure 1D</a></li>
  </ul></li>
  <li><a href="#sec-si-figure2" id="toc-sec-si-figure2" class="nav-link" data-scroll-target="#sec-si-figure2"><span class="header-section-number">9.2</span> Supplementary Text Figure 2</a></li>
  <li><a href="#sec-si-figure3" id="toc-sec-si-figure3" class="nav-link" data-scroll-target="#sec-si-figure3"><span class="header-section-number">9.3</span> Supplementary Text Figure 3</a></li>
  <li><a href="#sec-si-figure4-6" id="toc-sec-si-figure4-6" class="nav-link" data-scroll-target="#sec-si-figure4-6"><span class="header-section-number">9.4</span> Supplementary Text Figure 4, 5, and 6</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/obsgynaecam/cell-free-rna-2024/edit/main/suppl_figure.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/obsgynaecam/cell-free-rna-2024/blob/main/suppl_figure.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li><li><a href="https://github.com/obsgynaecam/cell-free-rna-2024/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">


<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./figure.html">Figures</a></li><li class="breadcrumb-item"><a href="./suppl_figure.html"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Supplementary Text Figures</span></a></li></ol></nav>
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-suppl-figure" class="quarto-section-identifier"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Supplementary Text Figures</span></span></h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>This chapter demonstrates the <code>R</code> codes that make supplementary text figures shown in the paper. Making the main figures are shown in <a href="main_figure.html" class="quarto-xref"><span>Chapter 8</span></a>.</p>
<section id="sec-si-figure1" class="level2" data-number="9.1">
<h2 data-number="9.1" class="anchored" data-anchor-id="sec-si-figure1"><span class="header-section-number">9.1</span> Supplementary Text Figure 1</h2>
<section id="sec-si-figure1-a" class="level3" data-number="9.1.1">
<h3 data-number="9.1.1" class="anchored" data-anchor-id="sec-si-figure1-a"><span class="header-section-number">9.1.1</span> Supplementary Text Figure 1A</h3>
<div id="fig-si-fig1a" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-si-fig1a-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="static/figure/cfRNA.SI.FigA.png" id="fig-si-fig1a" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-si-fig1a-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9.1
</figcaption>
</figure>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>  <span class="fu">lazyLoad</span>(<span class="st">"downsample.chrY.TPM_cache/beamer/load_ds_salmon_4b8dd727729fd0ae5deb423d69e7c16d"</span>) <span class="co"># dl.chrY.TPM and dl.stat</span></span>
<span id="cb1-2"><a href="#cb1-2"></a></span>
<span id="cb1-3"><a href="#cb1-3"></a>  dt.ds.chrY <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(dl.stat), <span class="cf">function</span>(DS) dl.stat[[DS]][,<span class="at">DS:=</span>DS]) <span class="sc">%&gt;%</span> rbindlist</span>
<span id="cb1-4"><a href="#cb1-4"></a>  dt.ds.chrY<span class="sc">$</span>DS<span class="ot">&lt;-</span><span class="fu">factor</span>(dt.ds.chrY<span class="sc">$</span>DS, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"1M"</span>,<span class="st">"5M"</span>,<span class="st">"10M"</span>))</span>
<span id="cb1-5"><a href="#cb1-5"></a>  dt.ds.chrY[,GA<span class="sc">:</span><span class="er">=</span><span class="fu">gsub</span>(<span class="st">" weeks"</span>,<span class="st">"wk"</span>,GA)]</span>
<span id="cb1-6"><a href="#cb1-6"></a>  <span class="fu">setnames</span>(dt.ds.chrY, <span class="st">"Type"</span>, <span class="st">"Method"</span>)</span>
<span id="cb1-7"><a href="#cb1-7"></a></span>
<span id="cb1-8"><a href="#cb1-8"></a>  p1.chrY<span class="ot">&lt;-</span><span class="fu">ggplot</span>(dt.ds.chrY, <span class="fu">aes</span>(Method,SumCount)) <span class="sc">+</span></span>
<span id="cb1-9"><a href="#cb1-9"></a>  <span class="fu">geom_rect</span>(<span class="at">data =</span> dt.ds.chrY[DS<span class="sc">==</span><span class="st">"5M"</span>],<span class="at">fill=</span><span class="st">"grey60"</span>,<span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">ymax =</span> <span class="cn">Inf</span>,<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb1-10"><a href="#cb1-10"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">fill=</span>Method), <span class="at">stat=</span><span class="st">"identity"</span>,<span class="at">width=</span>.<span class="dv">5</span>, <span class="at">alpha=</span>.<span class="dv">7</span>) <span class="sc">+</span></span>
<span id="cb1-11"><a href="#cb1-11"></a>  ggsci<span class="sc">::</span><span class="fu">scale_fill_jama</span>() <span class="sc">+</span></span>
<span id="cb1-12"><a href="#cb1-12"></a>  <span class="fu">facet_grid</span>(Sex<span class="sc">~</span>DS<span class="sc">+</span>GA,<span class="at">scales=</span><span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb1-13"><a href="#cb1-13"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">"Sum of read count on chrY genes"</span>) <span class="sc">+</span></span>
<span id="cb1-14"><a href="#cb1-14"></a>  <span class="fu">theme_Publication</span>() <span class="sc">+</span></span>
<span id="cb1-15"><a href="#cb1-15"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"top"</span>,</span>
<span id="cb1-16"><a href="#cb1-16"></a>        <span class="at">panel.border=</span><span class="fu">element_rect</span>(<span class="at">colour=</span><span class="st">"black"</span>),</span>
<span id="cb1-17"><a href="#cb1-17"></a>        <span class="at">axis.title.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb1-18"><a href="#cb1-18"></a>        <span class="at">axis.text.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb1-19"><a href="#cb1-19"></a>        <span class="at">axis.ticks.x=</span><span class="fu">element_blank</span>())</span>
<span id="cb1-20"><a href="#cb1-20"></a></span>
<span id="cb1-21"><a href="#cb1-21"></a>  <span class="co">#</span></span>
<span id="cb1-22"><a href="#cb1-22"></a>  <span class="fu">pdf</span>(<span class="at">file=</span><span class="st">"Figures/Suppl/cfRNA.Suppl.Text.Fig1A.pdf"</span>, <span class="at">width=</span><span class="dv">11</span>, <span class="at">height=</span><span class="dv">8</span>, <span class="at">title=</span><span class="st">"Abundance of chrY"</span>)</span>
<span id="cb1-23"><a href="#cb1-23"></a>  p1.chrY</span>
<span id="cb1-24"><a href="#cb1-24"></a>  <span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="sec-si-figure-1b" class="level3" data-number="9.1.2">
<h3 data-number="9.1.2" class="anchored" data-anchor-id="sec-si-figure-1b"><span class="header-section-number">9.1.2</span> Supplementary Text Figure 1B</h3>
<div id="fig-si-fig1b" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-si-fig1b-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="static/figure/cfRNA.SI.FigB.png" id="fig-si-fig1b" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-si-fig1b-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9.2
</figcaption>
</figure>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>  <span class="fu">lazyLoad</span>(<span class="st">"downsample.chrY.TPM_cache/beamer/load_ds_salmon_4b8dd727729fd0ae5deb423d69e7c16d"</span>) <span class="co"># dl.chrY.TPM and dl.stat</span></span>
<span id="cb2-2"><a href="#cb2-2"></a></span>
<span id="cb2-3"><a href="#cb2-3"></a>  <span class="fu">lapply</span>(dl.chrY.TPM, <span class="cf">function</span>(i) i[,.(<span class="fu">sum</span>(Counts),<span class="fu">sum</span>(TPM))])</span>
<span id="cb2-4"><a href="#cb2-4"></a></span>
<span id="cb2-5"><a href="#cb2-5"></a>  dl.chrY.TPM<span class="ot">&lt;-</span><span class="fu">lapply</span>(dl.chrY.TPM, <span class="cf">function</span>(DT) {</span>
<span id="cb2-6"><a href="#cb2-6"></a>                  <span class="co">#DT[Sex=="Male",Class:=ifelse(TPM&gt;0,"TP","FN")]</span></span>
<span id="cb2-7"><a href="#cb2-7"></a>                  <span class="co">#DT[Sex=="Female",Class:=ifelse(TPM==0,"TN","FP")] </span></span>
<span id="cb2-8"><a href="#cb2-8"></a>                  DT[,<span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span><span class="sc">:</span><span class="er">=</span><span class="fu">ifelse</span>(TPM<span class="sc">&gt;</span><span class="dv">0</span>, <span class="st">"Male"</span>,<span class="st">"Female"</span>),Type]</span>
<span id="cb2-9"><a href="#cb2-9"></a>                  DT<span class="sc">$</span>Sex<span class="ot">&lt;-</span><span class="fu">factor</span>(DT<span class="sc">$</span>Sex, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"Male"</span>,<span class="st">"Female"</span>))</span>
<span id="cb2-10"><a href="#cb2-10"></a>                  DT<span class="sc">$</span><span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span><span class="ot">&lt;-</span><span class="fu">factor</span>(DT<span class="sc">$</span><span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span>, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"Male"</span>,<span class="st">"Female"</span>))</span>
<span id="cb2-11"><a href="#cb2-11"></a>                  DT </span>
<span id="cb2-12"><a href="#cb2-12"></a>  }) </span>
<span id="cb2-13"><a href="#cb2-13"></a></span>
<span id="cb2-14"><a href="#cb2-14"></a>  <span class="do">#####################################</span></span>
<span id="cb2-15"><a href="#cb2-15"></a>  <span class="co"># Predictive performance by DS &amp; GA #</span></span>
<span id="cb2-16"><a href="#cb2-16"></a>  <span class="do">#####################################</span></span>
<span id="cb2-17"><a href="#cb2-17"></a>  dt.chrY.con.ga<span class="ot">&lt;-</span><span class="fu">lapply</span>(<span class="fu">names</span>(dl.chrY.TPM), <span class="cf">function</span>(DS){</span>
<span id="cb2-18"><a href="#cb2-18"></a>    <span class="fu">lapply</span>(<span class="fu">split</span>(dl.chrY.TPM[[DS]], dl.chrY.TPM[[DS]]<span class="sc">$</span>GA), <span class="cf">function</span>(DT){</span>
<span id="cb2-19"><a href="#cb2-19"></a>    my.GA<span class="ot">&lt;-</span>DT[,.N,GA]<span class="sc">$</span>GA</span>
<span id="cb2-20"><a href="#cb2-20"></a></span>
<span id="cb2-21"><a href="#cb2-21"></a>    cm1<span class="ot">&lt;-</span>caret<span class="sc">::</span><span class="fu">confusionMatrix</span>(DT[Type<span class="sc">==</span><span class="st">"HiSat2+Salmon"</span>][[<span class="st">"Predicted Sex"</span>]], <span class="at">reference=</span>DT[Type<span class="sc">==</span><span class="st">"HiSat2+Salmon"</span>][[<span class="st">"Sex"</span>]])</span>
<span id="cb2-22"><a href="#cb2-22"></a>    cm2<span class="ot">&lt;-</span>caret<span class="sc">::</span><span class="fu">confusionMatrix</span>(DT[Type<span class="sc">==</span><span class="st">"Salmon (SA mode)"</span>][[<span class="st">"Predicted Sex"</span>]], <span class="at">reference=</span>DT[Type<span class="sc">==</span><span class="st">"Salmon (SA mode)"</span>][[<span class="st">"Sex"</span>]])</span>
<span id="cb2-23"><a href="#cb2-23"></a></span>
<span id="cb2-24"><a href="#cb2-24"></a>    <span class="fu">rbind</span>(</span>
<span id="cb2-25"><a href="#cb2-25"></a>    <span class="fu">data.table</span>(<span class="at">DS=</span>DS, <span class="at">GA=</span>my.GA,<span class="at">Type=</span><span class="st">"HiSat2+Salmon"</span>, <span class="at">Measure=</span><span class="fu">names</span>(cm1<span class="sc">$</span>byClass), <span class="at">Value=</span>cm1<span class="sc">$</span>byClass),</span>
<span id="cb2-26"><a href="#cb2-26"></a>    <span class="fu">data.table</span>(<span class="at">DS=</span>DS, <span class="at">GA=</span>my.GA,<span class="at">Type=</span><span class="st">"Salmon (SA mode)"</span>, <span class="at">Measure=</span><span class="fu">names</span>(cm2<span class="sc">$</span>byClass), <span class="at">Value=</span>cm2<span class="sc">$</span>byClass)</span>
<span id="cb2-27"><a href="#cb2-27"></a>    )[Measure <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"F1"</span>,<span class="st">"Precision"</span>, <span class="st">"Recall"</span>)]</span>
<span id="cb2-28"><a href="#cb2-28"></a>    }) <span class="sc">%&gt;%</span> rbindlist </span>
<span id="cb2-29"><a href="#cb2-29"></a>  }) <span class="sc">%&gt;%</span> rbindlist </span>
<span id="cb2-30"><a href="#cb2-30"></a></span>
<span id="cb2-31"><a href="#cb2-31"></a>  dt.chrY.con.ga<span class="sc">$</span>DS<span class="ot">&lt;-</span><span class="fu">factor</span>(dt.chrY.con.ga<span class="sc">$</span>DS, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"1M"</span>,<span class="st">"5M"</span>,<span class="st">"10M"</span>))</span>
<span id="cb2-32"><a href="#cb2-32"></a>  dt.chrY.con.ga[,GA<span class="sc">:</span><span class="er">=</span><span class="fu">gsub</span>(<span class="st">" weeks"</span>,<span class="st">"wk"</span>,GA)]</span>
<span id="cb2-33"><a href="#cb2-33"></a>  <span class="fu">setnames</span>(dt.chrY.con.ga, <span class="st">"Type"</span>, <span class="st">"Method"</span>)</span>
<span id="cb2-34"><a href="#cb2-34"></a></span>
<span id="cb2-35"><a href="#cb2-35"></a>  p1.chrY.con<span class="ot">&lt;-</span><span class="fu">ggplot</span>(dt.chrY.con.ga, <span class="fu">aes</span>(Method,Value)) <span class="sc">+</span></span>
<span id="cb2-36"><a href="#cb2-36"></a>  <span class="fu">geom_rect</span>(<span class="at">data =</span> dt.chrY.con.ga[DS<span class="sc">==</span><span class="st">"5M"</span>],<span class="at">fill=</span><span class="st">"grey60"</span>,<span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">ymax =</span> <span class="cn">Inf</span>,<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb2-37"><a href="#cb2-37"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">fill=</span>Method), <span class="at">stat=</span><span class="st">"identity"</span>,<span class="at">width=</span>.<span class="dv">5</span>, <span class="at">alpha=</span>.<span class="dv">7</span>) <span class="sc">+</span></span>
<span id="cb2-38"><a href="#cb2-38"></a>  ggsci<span class="sc">::</span><span class="fu">scale_fill_jama</span>() <span class="sc">+</span></span>
<span id="cb2-39"><a href="#cb2-39"></a>  <span class="fu">facet_grid</span>(Measure<span class="sc">~</span>DS<span class="sc">+</span>GA,<span class="at">scales=</span><span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb2-40"><a href="#cb2-40"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">"Score"</span>) <span class="sc">+</span></span>
<span id="cb2-41"><a href="#cb2-41"></a>  <span class="fu">theme_Publication</span>() <span class="sc">+</span></span>
<span id="cb2-42"><a href="#cb2-42"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>,</span>
<span id="cb2-43"><a href="#cb2-43"></a>        <span class="at">panel.border=</span><span class="fu">element_rect</span>(<span class="at">colour=</span><span class="st">"black"</span>),</span>
<span id="cb2-44"><a href="#cb2-44"></a>        <span class="at">axis.title.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb2-45"><a href="#cb2-45"></a>        <span class="at">axis.text.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb2-46"><a href="#cb2-46"></a>        <span class="at">axis.ticks.x=</span><span class="fu">element_blank</span>())</span>
<span id="cb2-47"><a href="#cb2-47"></a></span>
<span id="cb2-48"><a href="#cb2-48"></a></span>
<span id="cb2-49"><a href="#cb2-49"></a>  <span class="co"># merge SI.Fig1A + SI.Fig1B</span></span>
<span id="cb2-50"><a href="#cb2-50"></a>  <span class="fu">pdf</span>(<span class="at">file=</span><span class="st">"Figures/Suppl/cfRNA.Suppl.Text.Fig1AB.pdf"</span>, <span class="at">width=</span><span class="dv">11</span>, <span class="at">height=</span><span class="dv">14</span>, <span class="at">title=</span><span class="st">"Abundance of chrY"</span>)</span>
<span id="cb2-51"><a href="#cb2-51"></a>  cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(p1.chrY, p1.chrY.con, <span class="at">labels=</span><span class="st">"AUTO"</span>, <span class="at">ncol=</span><span class="dv">1</span>, <span class="at">label_size=</span><span class="dv">27</span>, <span class="at">align=</span><span class="st">"v"</span>, <span class="at">axis=</span><span class="st">"l"</span>)</span>
<span id="cb2-52"><a href="#cb2-52"></a>  <span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="sec-si-figure-1c" class="level3" data-number="9.1.3">
<h3 data-number="9.1.3" class="anchored" data-anchor-id="sec-si-figure-1c"><span class="header-section-number">9.1.3</span> Supplementary Text Figure 1C</h3>
<div id="fig-si-fig1c" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-si-fig1c-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="static/figure/cfRNA.SI.FigC.png" id="fig-si-fig1c" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-si-fig1c-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9.3
</figcaption>
</figure>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>  <span class="co"># load the transcript definition used in Salmon (SA)</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>  <span class="fu">load</span>(<span class="st">"RData/gr.ensg.Homo_sapiens.GRCh38.88.cdna.all.ncrna.fa.gz.RData"</span>) <span class="sc">%&gt;%</span> system.time <span class="co"># isa GRanges</span></span>
<span id="cb3-3"><a href="#cb3-3"></a>  gr.ensg[<span class="fu">seqnames</span>(gr.ensg)<span class="sc">==</span><span class="st">"Y"</span> <span class="sc">&amp;</span> gr.ensg<span class="sc">$</span>gene_biotype<span class="sc">==</span><span class="st">"protein_coding"</span>,] <span class="co"># chrY protein-coding; n=42 </span></span>
<span id="cb3-4"><a href="#cb3-4"></a>  this.gene<span class="ot">&lt;-</span>(gr.ensg[<span class="fu">seqnames</span>(gr.ensg)<span class="sc">==</span><span class="st">"Y"</span> <span class="sc">&amp;</span> gr.ensg<span class="sc">$</span>gene_biotype<span class="sc">==</span><span class="st">"protein_coding"</span>,] <span class="sc">%&gt;%</span> names) <span class="co"># chrY protein-coding; n=42 (without the training version e.g. ENSG00000176679)</span></span>
<span id="cb3-5"><a href="#cb3-5"></a></span>
<span id="cb3-6"><a href="#cb3-6"></a>  <span class="co"># load the TPM based om 10M-ds dataset</span></span>
<span id="cb3-7"><a href="#cb3-7"></a>  <span class="fu">load</span>(<span class="st">"RData/li.TPM.10M.RData"</span>)</span>
<span id="cb3-8"><a href="#cb3-8"></a>  li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"TPM"</span>]] <span class="sc">%&gt;%</span> dim <span class="co"># 62803 genes x 100 samples</span></span>
<span id="cb3-9"><a href="#cb3-9"></a></span>
<span id="cb3-10"><a href="#cb3-10"></a>  <span class="do">## Heatamp of TPM using `HiSat2+Salmon` (based on 10M)</span></span>
<span id="cb3-11"><a href="#cb3-11"></a>  li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> rowSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="co"># count of 42 chrY genes</span></span>
<span id="cb3-12"><a href="#cb3-12"></a>  filter.g2<span class="ot">&lt;-</span>li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> rowSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="sc">&gt;</span><span class="dv">0</span> <span class="co"># filter for 42 chrY genes with TPM &gt;0</span></span>
<span id="cb3-13"><a href="#cb3-13"></a>  this.gene2<span class="ot">&lt;-</span>(li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> rowSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev)[filter.g2] <span class="sc">%&gt;%</span> names <span class="co"># n=24 genes</span></span>
<span id="cb3-14"><a href="#cb3-14"></a>  this.sample2<span class="ot">&lt;-</span>li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="sc">%&gt;%</span> names <span class="co"># n=100 samples sorted </span></span>
<span id="cb3-15"><a href="#cb3-15"></a>  filter.s2<span class="ot">&lt;-</span>li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene2,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="sc">&gt;</span><span class="dv">0</span> <span class="co"># filter for samples with TPM &gt;0 </span></span>
<span id="cb3-16"><a href="#cb3-16"></a>  predicted.male2<span class="ot">&lt;-</span>(li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev)[filter.s2] <span class="sc">%&gt;%</span> names <span class="co"># n=60 samples </span></span>
<span id="cb3-17"><a href="#cb3-17"></a></span>
<span id="cb3-18"><a href="#cb3-18"></a>  my.mat2<span class="ot">&lt;-</span><span class="fu">log2</span>(li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene2,this.sample2]<span class="sc">+</span><span class="dv">1</span>)</span>
<span id="cb3-19"><a href="#cb3-19"></a>  <span class="fu">rownames</span>(my.mat2)<span class="ot">&lt;-</span>gr.ensg[this.gene2]<span class="sc">$</span>gene_name</span>
<span id="cb3-20"><a href="#cb3-20"></a></span>
<span id="cb3-21"><a href="#cb3-21"></a>  col.anno<span class="ot">&lt;-</span>mat.samples[this.sample2,<span class="fu">c</span>(<span class="st">"GA"</span>,<span class="st">"Sex"</span>),drop<span class="ot">=</span>F] <span class="co"># isa `data.frame`</span></span>
<span id="cb3-22"><a href="#cb3-22"></a>  col.anno<span class="sc">$</span><span class="st">"Predicted Sex"</span><span class="ot">&lt;-</span><span class="fu">ifelse</span>(this.sample2 <span class="sc">%in%</span> predicted.male2,<span class="st">"Male"</span>,<span class="st">"Female"</span>)</span>
<span id="cb3-23"><a href="#cb3-23"></a>  col.anno<span class="ot">&lt;-</span>col.anno[,<span class="fu">c</span>(<span class="st">"Predicted Sex"</span>,<span class="st">"Sex"</span>,<span class="st">"GA"</span>)]</span>
<span id="cb3-24"><a href="#cb3-24"></a></span>
<span id="cb3-25"><a href="#cb3-25"></a>  GA.color<span class="ot">&lt;-</span>ggsci<span class="sc">::</span><span class="fu">pal_jama</span>(<span class="st">"default"</span>,<span class="at">alpha=</span>.<span class="dv">9</span>)(<span class="dv">4</span>)</span>
<span id="cb3-26"><a href="#cb3-26"></a>  <span class="fu">names</span>(GA.color)<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"12wk"</span>,<span class="st">"20wk"</span>,<span class="st">"28wk"</span>,<span class="st">"36wk"</span>)</span>
<span id="cb3-27"><a href="#cb3-27"></a>  my.color<span class="ot">=</span><span class="fu">list</span>(</span>
<span id="cb3-28"><a href="#cb3-28"></a>      <span class="st">`</span><span class="at">Sex</span><span class="st">`</span><span class="ot">=</span><span class="fu">c</span>(<span class="at">Female=</span><span class="st">"hotpink"</span>,<span class="at">Male=</span><span class="st">'skyblue'</span>),</span>
<span id="cb3-29"><a href="#cb3-29"></a>      <span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span><span class="ot">=</span><span class="fu">c</span>(<span class="at">Female=</span><span class="st">"hotpink3"</span>,<span class="at">Male=</span><span class="st">'skyblue3'</span>),</span>
<span id="cb3-30"><a href="#cb3-30"></a>      <span class="co">#`GA`=c("12wk"=cbPalette2[1],"20wk"=cbPalette2[2],"28wk"=cbPalette2[3],"36wk"=cbPalette2[4])</span></span>
<span id="cb3-31"><a href="#cb3-31"></a>      <span class="st">`</span><span class="at">GA</span><span class="st">`</span><span class="ot">=</span>GA.color</span>
<span id="cb3-32"><a href="#cb3-32"></a>  )</span>
<span id="cb3-33"><a href="#cb3-33"></a></span>
<span id="cb3-34"><a href="#cb3-34"></a>  cph1<span class="ot">&lt;-</span>ComplexHeatmap<span class="sc">::</span><span class="fu">pheatmap</span>(my.mat2,</span>
<span id="cb3-35"><a href="#cb3-35"></a>                                 <span class="at">annotation_col=</span>col.anno, </span>
<span id="cb3-36"><a href="#cb3-36"></a>                                 <span class="at">annotation_colors=</span>my.color, </span>
<span id="cb3-37"><a href="#cb3-37"></a>                                 <span class="at">cluster_cols=</span>F,</span>
<span id="cb3-38"><a href="#cb3-38"></a>                                 <span class="at">cluster_rows=</span>F,</span>
<span id="cb3-39"><a href="#cb3-39"></a>                                 <span class="at">show_colnames=</span>F,</span>
<span id="cb3-40"><a href="#cb3-40"></a>                                 <span class="at">border_color=</span><span class="cn">NA</span>,</span>
<span id="cb3-41"><a href="#cb3-41"></a>                                 <span class="at">fontsize=</span><span class="dv">12</span>,</span>
<span id="cb3-42"><a href="#cb3-42"></a>                                 <span class="at">name=</span><span class="st">"log2(count)"</span>,</span>
<span id="cb3-43"><a href="#cb3-43"></a>                                 <span class="at">row_title=</span><span class="st">"</span><span class="sc">\n</span><span class="st">HiSat2+Salmon"</span>,</span>
<span id="cb3-44"><a href="#cb3-44"></a>                                 <span class="at">row_title_gp =</span> <span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">20</span>),</span>
<span id="cb3-45"><a href="#cb3-45"></a>  )</span>
<span id="cb3-46"><a href="#cb3-46"></a></span>
<span id="cb3-47"><a href="#cb3-47"></a>  filter.g3<span class="ot">&lt;-</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> rowSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="sc">&gt;</span><span class="dv">0</span> <span class="co"># filter for 42 chrY genes with count &gt;0</span></span>
<span id="cb3-48"><a href="#cb3-48"></a>  this.gene3<span class="ot">&lt;-</span>(li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> rowSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev)[filter.g3] <span class="sc">%&gt;%</span> names <span class="co"># n=16 genes (out of 42) &gt;0 count </span></span>
<span id="cb3-49"><a href="#cb3-49"></a>  this.sample3<span class="ot">&lt;-</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="sc">%&gt;%</span> names <span class="co"># 100 sample names</span></span>
<span id="cb3-50"><a href="#cb3-50"></a>  filter.s3<span class="ot">&lt;-</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene3,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="sc">&gt;</span><span class="dv">0</span> <span class="co"># samples with count&gt;0 </span></span>
<span id="cb3-51"><a href="#cb3-51"></a>  <span class="co">#(li.TPM[["Salmon"]][["counts"]][this.gene,] %&gt;% colSums %&gt;% sort %&gt;% rev)[filter.s3] # n=38 samples with count&gt;0 (predicted as Male)</span></span>
<span id="cb3-52"><a href="#cb3-52"></a>  predicted.male3<span class="ot">&lt;-</span>(li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev)[filter.s3] <span class="sc">%&gt;%</span> names <span class="co"># n=38 samples with count&gt;0 (predicted as Male)</span></span>
<span id="cb3-53"><a href="#cb3-53"></a></span>
<span id="cb3-54"><a href="#cb3-54"></a>  my.mat3<span class="ot">&lt;-</span><span class="fu">log2</span>(li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene3,this.sample3]<span class="sc">+</span><span class="dv">1</span>)</span>
<span id="cb3-55"><a href="#cb3-55"></a>  <span class="fu">rownames</span>(my.mat3)<span class="ot">&lt;-</span>gr.ensg[this.gene3]<span class="sc">$</span>gene_name</span>
<span id="cb3-56"><a href="#cb3-56"></a></span>
<span id="cb3-57"><a href="#cb3-57"></a>  col.anno3<span class="ot">&lt;-</span>mat.samples[this.sample3,<span class="fu">c</span>(<span class="st">"GA"</span>,<span class="st">"Sex"</span>),drop<span class="ot">=</span>F]</span>
<span id="cb3-58"><a href="#cb3-58"></a>  col.anno3<span class="sc">$</span><span class="st">"Predicted Sex"</span><span class="ot">&lt;-</span><span class="fu">ifelse</span>(this.sample3 <span class="sc">%in%</span> predicted.male3,<span class="st">"Male"</span>,<span class="st">"Female"</span>)</span>
<span id="cb3-59"><a href="#cb3-59"></a>  col.anno3<span class="ot">&lt;-</span>col.anno3[,<span class="fu">c</span>(<span class="st">"Predicted Sex"</span>,<span class="st">"Sex"</span>,<span class="st">"GA"</span>)]</span>
<span id="cb3-60"><a href="#cb3-60"></a></span>
<span id="cb3-61"><a href="#cb3-61"></a>  <span class="co"># via ComplexHeatmap</span></span>
<span id="cb3-62"><a href="#cb3-62"></a>  cph3<span class="ot">&lt;-</span>ComplexHeatmap<span class="sc">::</span><span class="fu">pheatmap</span>(my.mat3,</span>
<span id="cb3-63"><a href="#cb3-63"></a>                                 <span class="at">annotation_col=</span>col.anno3, </span>
<span id="cb3-64"><a href="#cb3-64"></a>                                 <span class="at">annotation_colors=</span>my.color, </span>
<span id="cb3-65"><a href="#cb3-65"></a>                                 <span class="at">cluster_cols=</span>F,</span>
<span id="cb3-66"><a href="#cb3-66"></a>                                 <span class="at">cluster_rows=</span>F,</span>
<span id="cb3-67"><a href="#cb3-67"></a>                                 <span class="at">show_colnames=</span>F,</span>
<span id="cb3-68"><a href="#cb3-68"></a>                                 <span class="at">border_color=</span><span class="cn">NA</span>,</span>
<span id="cb3-69"><a href="#cb3-69"></a>                                 <span class="at">fontsize=</span><span class="dv">12</span>,</span>
<span id="cb3-70"><a href="#cb3-70"></a>                                 <span class="at">name=</span><span class="st">"log2(count)"</span>,</span>
<span id="cb3-71"><a href="#cb3-71"></a>                                 <span class="at">row_title=</span><span class="st">"</span><span class="sc">\n</span><span class="st">Salmon (SA mode)"</span>,</span>
<span id="cb3-72"><a href="#cb3-72"></a>                                 <span class="at">row_title_gp =</span> <span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">20</span>),</span>
<span id="cb3-73"><a href="#cb3-73"></a>                                 <span class="at">heatmap_legend_param =</span> <span class="fu">list</span>(</span>
<span id="cb3-74"><a href="#cb3-74"></a>                                                            <span class="at">title_gp=</span><span class="fu">gpar</span>(<span class="at">fontsize=</span><span class="dv">13</span>,<span class="at">fontface=</span><span class="st">"bold"</span>),</span>
<span id="cb3-75"><a href="#cb3-75"></a>                                                            <span class="at">labels_gp=</span><span class="fu">gpar</span>(<span class="at">fontsize=</span><span class="dv">11</span>),</span>
<span id="cb3-76"><a href="#cb3-76"></a>                            <span class="at">annotation_legend_param =</span> <span class="fu">list</span>(<span class="at">title_gp=</span><span class="fu">gpar</span>(<span class="at">fontsize=</span><span class="dv">13</span>,<span class="at">fontface=</span><span class="st">"bold"</span>),<span class="at">labels_gp =</span> <span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">11</span>))</span>
<span id="cb3-77"><a href="#cb3-77"></a>                                )</span>
<span id="cb3-78"><a href="#cb3-78"></a>  )</span>
<span id="cb3-79"><a href="#cb3-79"></a></span>
<span id="cb3-80"><a href="#cb3-80"></a></span>
<span id="cb3-81"><a href="#cb3-81"></a>  <span class="fu">pdf</span>(<span class="at">file=</span><span class="st">"Figures/Suppl/cfRNA.Suppl.Text.Fig1C.pdf"</span>, <span class="at">width=</span><span class="dv">11</span>, <span class="at">height=</span><span class="dv">14</span>, <span class="at">title=</span><span class="st">"Abundance of chrY"</span>)</span>
<span id="cb3-82"><a href="#cb3-82"></a>  <span class="fu">draw</span>(cph1 <span class="sc">%v%</span> cph3, <span class="at">main_heatmap=</span><span class="st">"log2(count)"</span>, <span class="at">ht_gap=</span><span class="fu">unit</span>(.<span class="dv">7</span>,<span class="st">"cm"</span>), <span class="at">auto_adjust=</span>F)</span>
<span id="cb3-83"><a href="#cb3-83"></a>  <span class="fu">grid.text</span>(<span class="at">label=</span><span class="st">"C"</span>,<span class="at">gp=</span><span class="fu">gpar</span>(<span class="at">fontsize=</span><span class="dv">27</span>,<span class="at">fontface=</span><span class="st">"bold"</span>),<span class="at">x=</span><span class="fl">0.02</span>,<span class="at">y=</span>.<span class="dv">99</span>,<span class="at">just=</span><span class="st">"top"</span>)</span>
<span id="cb3-84"><a href="#cb3-84"></a>  <span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="sec-si-figure-1d" class="level3" data-number="9.1.4">
<h3 data-number="9.1.4" class="anchored" data-anchor-id="sec-si-figure-1d"><span class="header-section-number">9.1.4</span> Supplementary Text Figure 1D</h3>
<div id="fig-si-fig1d" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-si-fig1d-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="static/figure/cfRNA.SI.FigD.png" id="fig-si-fig1d" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-si-fig1d-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9.4
</figcaption>
</figure>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>  dl.chrY.TPM <span class="sc">%&gt;%</span> names</span>
<span id="cb4-2"><a href="#cb4-2"></a></span>
<span id="cb4-3"><a href="#cb4-3"></a>  <span class="co"># load the TPM based om 10M-ds dataset</span></span>
<span id="cb4-4"><a href="#cb4-4"></a>  <span class="fu">load</span>(<span class="st">"RData/li.TPM.10M.trim.RData"</span>)</span>
<span id="cb4-5"><a href="#cb4-5"></a>  li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"TPM"</span>]] <span class="sc">%&gt;%</span> dim <span class="co"># 59354 genes x 100 samples</span></span>
<span id="cb4-6"><a href="#cb4-6"></a></span>
<span id="cb4-7"><a href="#cb4-7"></a>  dt.trim<span class="fl">.10</span>M<span class="ot">&lt;-</span><span class="fu">data.table</span>(</span>
<span id="cb4-8"><a href="#cb4-8"></a>      <span class="st">`</span><span class="at">Type</span><span class="st">`</span><span class="ot">=</span><span class="st">"Salmon (SA mode)"</span>,</span>
<span id="cb4-9"><a href="#cb4-9"></a>      <span class="st">`</span><span class="at">SampleID</span><span class="st">`</span><span class="ot">=</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> names,</span>
<span id="cb4-10"><a href="#cb4-10"></a>      <span class="st">`</span><span class="at">Counts</span><span class="st">`</span><span class="ot">=</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums, </span>
<span id="cb4-11"><a href="#cb4-11"></a>      <span class="st">`</span><span class="at">TPM</span><span class="st">`</span><span class="ot">=</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"TPM"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums </span>
<span id="cb4-12"><a href="#cb4-12"></a>      )</span>
<span id="cb4-13"><a href="#cb4-13"></a>  <span class="co"># add Sex and GA</span></span>
<span id="cb4-14"><a href="#cb4-14"></a>  dt.trim<span class="fl">.10</span>M<span class="ot">&lt;-</span><span class="fu">merge</span>(dt.trim<span class="fl">.10</span>M, dt.samples[,.(SampleID,GA,Sex,pn_female)])</span>
<span id="cb4-15"><a href="#cb4-15"></a></span>
<span id="cb4-16"><a href="#cb4-16"></a>  dt.trim<span class="fl">.10</span>M[,<span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span><span class="sc">:</span><span class="er">=</span><span class="fu">ifelse</span>(TPM<span class="sc">&gt;</span><span class="dv">0</span>, <span class="st">"Male"</span>,<span class="st">"Female"</span>),Type]</span>
<span id="cb4-17"><a href="#cb4-17"></a>  dt.trim<span class="fl">.10</span>M<span class="sc">$</span>Sex<span class="ot">&lt;-</span><span class="fu">factor</span>(dt.trim<span class="fl">.10</span>M<span class="sc">$</span>Sex, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"Male"</span>,<span class="st">"Female"</span>))</span>
<span id="cb4-18"><a href="#cb4-18"></a>  dt.trim<span class="fl">.10</span>M<span class="sc">$</span><span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span><span class="ot">&lt;-</span><span class="fu">factor</span>(dt.trim<span class="fl">.10</span>M<span class="sc">$</span><span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span>, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"Male"</span>,<span class="st">"Female"</span>))</span>
<span id="cb4-19"><a href="#cb4-19"></a></span>
<span id="cb4-20"><a href="#cb4-20"></a>  dt.chrY.trim<span class="ot">&lt;-</span><span class="fu">rbind</span>(</span>
<span id="cb4-21"><a href="#cb4-21"></a>        dl.chrY.TPM[[<span class="st">"10M"</span>]][Type<span class="sc">==</span><span class="st">"Salmon (SA mode)"</span>][,<span class="at">Trim:=</span><span class="st">"No"</span>],</span>
<span id="cb4-22"><a href="#cb4-22"></a>        dt.trim<span class="fl">.10</span>M[,<span class="at">Trim:=</span><span class="st">"Yes"</span>]</span>
<span id="cb4-23"><a href="#cb4-23"></a>        )</span>
<span id="cb4-24"><a href="#cb4-24"></a>  <span class="do">## Number of false positives: chrY signal detected from females </span></span>
<span id="cb4-25"><a href="#cb4-25"></a>  dt.chrY.trim[,.(Trim,Sex,<span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span>)] <span class="sc">%&gt;%</span> ftable</span>
<span id="cb4-26"><a href="#cb4-26"></a></span>
<span id="cb4-27"><a href="#cb4-27"></a>  <span class="do">##</span></span>
<span id="cb4-28"><a href="#cb4-28"></a>  <span class="do">## Read counts by trim or not</span></span>
<span id="cb4-29"><a href="#cb4-29"></a>  dt.bar<span class="ot">&lt;-</span><span class="fu">merge</span>(</span>
<span id="cb4-30"><a href="#cb4-30"></a>    dl.chrY.TPM[[<span class="st">"10M"</span>]][Type<span class="sc">==</span><span class="st">"Salmon (SA mode)"</span>,.(SampleID,<span class="at">GA=</span><span class="fu">gsub</span>(<span class="st">" weeks"</span>,<span class="st">"wk"</span>,GA),Sex,Counts,TPM)], <span class="co"># no-trim</span></span>
<span id="cb4-31"><a href="#cb4-31"></a>    dt.trim<span class="fl">.10</span>M[Type<span class="sc">==</span><span class="st">"Salmon (SA mode)"</span>,.(SampleID,GA,Sex,Counts,TPM)], <span class="co"># trim</span></span>
<span id="cb4-32"><a href="#cb4-32"></a>    <span class="at">by=</span><span class="fu">c</span>(<span class="st">"SampleID"</span>,<span class="st">"Sex"</span>,<span class="st">"GA"</span>)</span>
<span id="cb4-33"><a href="#cb4-33"></a>  )</span>
<span id="cb4-34"><a href="#cb4-34"></a>  <span class="fu">cor.test</span>(<span class="sc">~</span>Counts.x <span class="sc">+</span> Counts.y, dt.bar)</span>
<span id="cb4-35"><a href="#cb4-35"></a>  dt.bar[Counts.x<span class="sc">&gt;</span>Counts.y] <span class="co"># no-trim &gt; trim</span></span>
<span id="cb4-36"><a href="#cb4-36"></a>  dt.bar[Counts.x<span class="sc">&lt;</span>Counts.y] <span class="co"># no-trim &lt; trim</span></span>
<span id="cb4-37"><a href="#cb4-37"></a>  dt.bar[Sex<span class="sc">==</span><span class="st">"Female"</span> <span class="sc">&amp;</span> Counts.x<span class="sc">&lt;</span>Counts.y] <span class="co"># more counts for Trimmed Salmon</span></span>
<span id="cb4-38"><a href="#cb4-38"></a></span>
<span id="cb4-39"><a href="#cb4-39"></a>  my.limit<span class="ot">&lt;-</span>dt.bar[,<span class="fu">max</span>(Counts.x,Counts.y)]  <span class="sc">%&gt;%</span> ceiling</span>
<span id="cb4-40"><a href="#cb4-40"></a>  p1.chrY.trim<span class="ot">&lt;-</span><span class="fu">ggplot</span>(dt.bar, <span class="fu">aes</span>(Counts.x<span class="sc">+</span><span class="dv">1</span>,Counts.y<span class="sc">+</span><span class="dv">1</span>)) <span class="sc">+</span> </span>
<span id="cb4-41"><a href="#cb4-41"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>dt.bar, <span class="at">size=</span><span class="dv">5</span>, <span class="at">position=</span><span class="fu">position_jitter</span>(<span class="at">width=</span><span class="fl">0.2</span>,<span class="at">height=</span><span class="fl">0.2</span>),<span class="at">shape=</span><span class="dv">21</span>) <span class="sc">+</span></span>
<span id="cb4-42"><a href="#cb4-42"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">trans =</span> scales<span class="sc">::</span><span class="fu">log2_trans</span>(),</span>
<span id="cb4-43"><a href="#cb4-43"></a>                     <span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">trans_breaks</span>(<span class="st">"log2"</span>,<span class="cf">function</span>(x) <span class="dv">2</span><span class="sc">^</span>x),</span>
<span id="cb4-44"><a href="#cb4-44"></a>                     <span class="at">labels =</span> scales<span class="sc">::</span><span class="fu">trans_format</span>(<span class="st">"log2"</span>, <span class="fu">math_format</span>(<span class="dv">2</span><span class="sc">^</span>.x)),</span>
<span id="cb4-45"><a href="#cb4-45"></a>                     <span class="at">limits=</span><span class="fu">c</span>(.<span class="dv">5</span>,my.limit<span class="sc">*</span><span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb4-46"><a href="#cb4-46"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">trans =</span> scales<span class="sc">::</span><span class="fu">log2_trans</span>(),</span>
<span id="cb4-47"><a href="#cb4-47"></a>                     <span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">trans_breaks</span>(<span class="st">"log2"</span>,<span class="cf">function</span>(x) <span class="dv">2</span><span class="sc">^</span>x),</span>
<span id="cb4-48"><a href="#cb4-48"></a>                     <span class="at">labels =</span> scales<span class="sc">::</span><span class="fu">trans_format</span>(<span class="st">"log2"</span>, <span class="fu">math_format</span>(<span class="dv">2</span><span class="sc">^</span>.x)),</span>
<span id="cb4-49"><a href="#cb4-49"></a>                     <span class="at">limits=</span><span class="fu">c</span>(.<span class="dv">5</span>,my.limit<span class="sc">*</span><span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb4-50"><a href="#cb4-50"></a>  <span class="co">#coord_fixed() + </span></span>
<span id="cb4-51"><a href="#cb4-51"></a>  <span class="fu">facet_grid</span>(Sex<span class="sc">~</span>GA,<span class="at">scales=</span><span class="st">"free_x"</span>) <span class="sc">+</span></span>
<span id="cb4-52"><a href="#cb4-52"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Sum of read count (No trimmed reads)"</span>, <span class="at">y=</span><span class="st">"Sum of read count (trimmed reads)"</span>) <span class="sc">+</span></span>
<span id="cb4-53"><a href="#cb4-53"></a>  <span class="fu">theme_Publication</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"black"</span>))</span>
<span id="cb4-54"><a href="#cb4-54"></a></span>
<span id="cb4-55"><a href="#cb4-55"></a>  <span class="fu">pdf</span>(<span class="at">file=</span><span class="st">"Figures/Suppl/cfRNA.Suppl.Text.Fig1D.v2.pdf"</span>, <span class="at">width=</span><span class="dv">11</span>, <span class="at">height=</span><span class="dv">7</span>, <span class="at">title=</span><span class="st">"Trim vs No trim"</span>)</span>
<span id="cb4-56"><a href="#cb4-56"></a>  cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(p1.chrY.trim, <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"D"</span>),<span class="at">label_size=</span><span class="dv">27</span>)</span>
<span id="cb4-57"><a href="#cb4-57"></a>  <span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="sec-si-figure2" class="level2" data-number="9.2">
<h2 data-number="9.2" class="anchored" data-anchor-id="sec-si-figure2"><span class="header-section-number">9.2</span> Supplementary Text Figure 2</h2>
<div id="fig-si-fig2" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-si-fig2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="static/figure/cfRNA.SI.Fig2.png" id="fig-si-fig2" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-si-fig2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9.5
</figcaption>
</figure>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="fu">load</span>(<span class="st">"RData/dl.enet.result.core17.RData"</span>) <span class="co"># Validation of ENet models from 28wk, discovery dataset (preterm)</span></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="fu">load</span>(<span class="st">"RData/dl.combined.result.core17.RData"</span>) <span class="co"># Validation of ENet models from the combined model (preterm+term)</span></span>
<span id="cb5-3"><a href="#cb5-3"></a></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="do">##</span></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="do">## AUC in All </span></span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="do">##</span></span>
<span id="cb5-7"><a href="#cb5-7"></a>ga<span class="ot">&lt;-</span><span class="fu">paste0</span>(<span class="fu">c</span>(<span class="dv">12</span>,<span class="dv">20</span>,<span class="dv">28</span>,<span class="dv">36</span>), <span class="st">"wk"</span>)</span>
<span id="cb5-8"><a href="#cb5-8"></a>term<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"(preterm)"</span>,<span class="st">"(term)"</span>)</span>
<span id="cb5-9"><a href="#cb5-9"></a>my.dataset<span class="ot">&lt;-</span><span class="fu">data.table</span>(<span class="fu">expand.grid</span>(ga,term))[,Var3<span class="sc">:</span><span class="er">=</span><span class="fu">paste0</span>(Var1,Var2)][Var3<span class="sc">!=</span><span class="st">"36wk(preterm)"</span>]<span class="sc">$</span>Var3</span>
<span id="cb5-10"><a href="#cb5-10"></a>my.dataset<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"Munchel"</span>,my.dataset)</span>
<span id="cb5-11"><a href="#cb5-11"></a></span>
<span id="cb5-12"><a href="#cb5-12"></a>dt.auc.long<span class="ot">&lt;-</span></span>
<span id="cb5-13"><a href="#cb5-13"></a><span class="fu">lapply</span>(my.dataset, <span class="cf">function</span>(this.dataset){</span>
<span id="cb5-14"><a href="#cb5-14"></a>  <span class="fu">rbind</span>(</span>
<span id="cb5-15"><a href="#cb5-15"></a>        (<span class="fu">lapply</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>, <span class="cf">function</span>(my.num) dl.combined.result[[my.num<span class="dv">-1</span>]][fold<span class="sc">==</span>this.dataset,.(<span class="at">Num=</span>my.num,<span class="at">Predictor=</span>predictor,<span class="at">AUC_test=</span>AUC_test<span class="sc">/</span><span class="dv">100</span>,<span class="at">AUC_test_lo=</span>AUC_test_lo<span class="sc">/</span><span class="dv">100</span>,<span class="at">AUC_test_hi=</span>AUC_test_hi<span class="sc">/</span><span class="dv">100</span>)]) <span class="sc">%&gt;%</span> rbindlist)[,<span class="at">Model:=</span><span class="st">"preterm+term (28wk)"</span>]</span>
<span id="cb5-16"><a href="#cb5-16"></a>        ,</span>
<span id="cb5-17"><a href="#cb5-17"></a>        (<span class="fu">lapply</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>, <span class="cf">function</span>(my.num) dl.enet.result[[my.num<span class="dv">-1</span>]][fold<span class="sc">==</span>this.dataset <span class="sc">&amp;</span> <span class="fu">grepl</span>(<span class="st">"ENet"</span>,methods),.(<span class="at">Num=</span>my.num,<span class="at">Predictor=</span>predictor,<span class="at">AUC_test=</span>AUC_test<span class="sc">/</span><span class="dv">100</span>,<span class="at">AUC_test_lo=</span>AUC_test_lo<span class="sc">/</span><span class="dv">100</span>,<span class="at">AUC_test_hi=</span>AUC_test_hi<span class="sc">/</span><span class="dv">100</span>)]) <span class="sc">%&gt;%</span> rbindlist)[,<span class="at">Model:=</span><span class="st">"preterm (28wk)"</span>]</span>
<span id="cb5-18"><a href="#cb5-18"></a>        </span>
<span id="cb5-19"><a href="#cb5-19"></a>  )[,Dataset<span class="sc">:</span><span class="er">=</span>this.dataset] </span>
<span id="cb5-20"><a href="#cb5-20"></a>}) <span class="sc">%&gt;%</span> rbindlist <span class="sc">%&gt;%</span> <span class="fu">setcolorder</span>(<span class="fu">c</span>(<span class="st">"Dataset"</span>,<span class="st">"Model"</span>))</span>
<span id="cb5-21"><a href="#cb5-21"></a></span>
<span id="cb5-22"><a href="#cb5-22"></a>dt.auc.long[Dataset<span class="sc">!=</span><span class="st">"Munchel"</span>,GA<span class="sc">:</span><span class="er">=</span><span class="fu">substr</span>(Dataset,<span class="dv">1</span>,<span class="dv">4</span>)]</span>
<span id="cb5-23"><a href="#cb5-23"></a>dt.auc.long[Dataset<span class="sc">!=</span><span class="st">"Munchel"</span>,Source<span class="sc">:</span><span class="er">=</span><span class="fu">substr</span>(Dataset,<span class="dv">6</span>,<span class="fu">nchar</span>(Dataset)<span class="sc">-</span><span class="dv">1</span>)][,Source<span class="sc">:</span><span class="er">=</span><span class="fu">ifelse</span>(Source<span class="sc">==</span><span class="st">"preterm"</span>,<span class="st">"Discovery (preterm)"</span>,<span class="st">"Validation (term)"</span>)]</span>
<span id="cb5-24"><a href="#cb5-24"></a>dt.auc.long[Dataset<span class="sc">==</span><span class="st">"Munchel"</span>,<span class="st">`</span><span class="at">:=</span><span class="st">`</span>(<span class="at">Source=</span><span class="st">"Munchel"</span>,<span class="at">GA=</span><span class="st">"36wk"</span>)]</span>
<span id="cb5-25"><a href="#cb5-25"></a>dt.auc.long<span class="sc">$</span>Model<span class="ot">&lt;-</span><span class="fu">factor</span>(dt.auc.long<span class="sc">$</span>Model, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"preterm (28wk)"</span>,<span class="st">"preterm+term (28wk)"</span>))</span>
<span id="cb5-26"><a href="#cb5-26"></a>dt.auc.long<span class="sc">$</span>Source<span class="ot">&lt;-</span><span class="fu">factor</span>(dt.auc.long<span class="sc">$</span>Source, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"Discovery (preterm)"</span>,<span class="st">"Validation (term)"</span>,<span class="st">"Munchel"</span>))</span>
<span id="cb5-27"><a href="#cb5-27"></a></span>
<span id="cb5-28"><a href="#cb5-28"></a>p.auc.combined.A<span class="ot">&lt;-</span><span class="fu">ggplot</span>(dt.auc.long[Source<span class="sc">!=</span><span class="st">"Munchel"</span>], <span class="fu">aes</span>(Num, AUC_test,<span class="at">group=</span>Model)) <span class="sc">+</span> </span>
<span id="cb5-29"><a href="#cb5-29"></a>  <span class="fu">geom_rect</span>(<span class="at">data =</span> dt.auc.long[Source<span class="sc">!=</span><span class="st">"Munchel"</span> <span class="sc">&amp;</span> GA<span class="sc">==</span><span class="st">"28wk"</span>],<span class="fu">aes</span>(<span class="at">fill =</span> GA),<span class="at">fill=</span><span class="st">"grey90"</span>,<span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">ymax =</span> <span class="cn">Inf</span>,<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb5-30"><a href="#cb5-30"></a>  <span class="fu">geom_pointrange</span>(<span class="fu">aes</span>(<span class="at">col=</span>Model,<span class="at">ymin=</span>AUC_test_lo, <span class="at">ymax=</span>AUC_test_hi),</span>
<span id="cb5-31"><a href="#cb5-31"></a>                  <span class="at">position=</span><span class="fu">position_dodge</span>(<span class="at">width=</span><span class="fl">0.4</span>),</span>
<span id="cb5-32"><a href="#cb5-32"></a>                  <span class="at">size=</span>.<span class="dv">8</span>,<span class="at">alpha=</span>.<span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb5-33"><a href="#cb5-33"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>),<span class="at">breaks=</span><span class="fu">c</span>(<span class="dv">0</span>,.<span class="dv">2</span>,.<span class="dv">4</span>,.<span class="dv">6</span>,.<span class="dv">8</span>,<span class="dv">1</span>), <span class="at">limit=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">1.05</span>)) <span class="sc">+</span></span>
<span id="cb5-34"><a href="#cb5-34"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks=</span><span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb5-35"><a href="#cb5-35"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">"AUC"</span>, <span class="at">x=</span><span class="st">"Number of predictor mRNA in the training model"</span>) <span class="sc">+</span></span>
<span id="cb5-36"><a href="#cb5-36"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"grey30"</span>,cbPalette2[<span class="dv">2</span>])) <span class="sc">+</span></span>
<span id="cb5-37"><a href="#cb5-37"></a>  <span class="fu">facet_grid</span>(Source<span class="sc">~</span>GA) <span class="sc">+</span> </span>
<span id="cb5-38"><a href="#cb5-38"></a>  <span class="fu">theme_Publication</span>() <span class="sc">+</span></span>
<span id="cb5-39"><a href="#cb5-39"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">""</span>, <span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"black"</span>))</span>
<span id="cb5-40"><a href="#cb5-40"></a></span>
<span id="cb5-41"><a href="#cb5-41"></a>p.auc.combined.B<span class="ot">&lt;-</span><span class="fu">ggplot</span>(dt.auc.long[Source<span class="sc">==</span><span class="st">"Munchel"</span>], <span class="fu">aes</span>(Num, AUC_test,<span class="at">group=</span>Model)) <span class="sc">+</span> </span>
<span id="cb5-42"><a href="#cb5-42"></a>  <span class="fu">geom_pointrange</span>(<span class="fu">aes</span>(<span class="at">col=</span>Model,<span class="at">ymin=</span>AUC_test_lo, <span class="at">ymax=</span>AUC_test_hi),</span>
<span id="cb5-43"><a href="#cb5-43"></a>                  <span class="at">position=</span><span class="fu">position_dodge</span>(<span class="at">width=</span><span class="fl">0.4</span>),</span>
<span id="cb5-44"><a href="#cb5-44"></a>                  <span class="at">size=</span>.<span class="dv">8</span>,<span class="at">alpha=</span>.<span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb5-45"><a href="#cb5-45"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>),<span class="at">breaks=</span><span class="fu">c</span>(<span class="dv">0</span>,.<span class="dv">2</span>,.<span class="dv">4</span>,.<span class="dv">6</span>,.<span class="dv">8</span>,<span class="dv">1</span>), <span class="at">limit=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">1.05</span>)) <span class="sc">+</span></span>
<span id="cb5-46"><a href="#cb5-46"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks=</span><span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb5-47"><a href="#cb5-47"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">"AUC"</span>, <span class="at">x=</span><span class="st">"Number of predictor mRNA in the training model                                     "</span>) <span class="sc">+</span></span>
<span id="cb5-48"><a href="#cb5-48"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"grey30"</span>,cbPalette2[<span class="dv">2</span>])) <span class="sc">+</span></span>
<span id="cb5-49"><a href="#cb5-49"></a>  <span class="fu">facet_grid</span>(Source<span class="sc">~</span>.) <span class="sc">+</span> </span>
<span id="cb5-50"><a href="#cb5-50"></a>  <span class="fu">theme_Publication</span>() <span class="sc">+</span></span>
<span id="cb5-51"><a href="#cb5-51"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"left"</span>, </span>
<span id="cb5-52"><a href="#cb5-52"></a>        <span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"black"</span>),</span>
<span id="cb5-53"><a href="#cb5-53"></a>        <span class="at">plot.margin=</span><span class="fu">margin</span>(<span class="dv">1</span>,<span class="sc">-</span><span class="dv">15</span>,<span class="dv">7</span>,<span class="dv">0</span>)</span>
<span id="cb5-54"><a href="#cb5-54"></a>  )</span>
<span id="cb5-55"><a href="#cb5-55"></a></span>
<span id="cb5-56"><a href="#cb5-56"></a>p.auc.bottom<span class="ot">&lt;-</span>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(<span class="cn">NULL</span>,p.auc.combined.B, <span class="at">nrow=</span><span class="dv">1</span>,<span class="at">labels=</span><span class="fu">c</span>(<span class="st">""</span>,<span class="st">"B"</span>),<span class="at">label_size=</span><span class="dv">27</span>,<span class="at">rel_widths=</span><span class="fu">c</span>(.<span class="dv">87</span>,<span class="dv">1</span>))</span>
<span id="cb5-57"><a href="#cb5-57"></a></span>
<span id="cb5-58"><a href="#cb5-58"></a><span class="fu">pdf</span>(<span class="at">file=</span><span class="st">"Figures/Suppl/cfRNA.Suppl.Data.Fig.combined.AUC.AB.pdf"</span>, <span class="at">width=</span><span class="dv">12</span>, <span class="at">height=</span><span class="dv">10</span>, <span class="at">title=</span><span class="st">"Suppl Fig: AUC from the combined model vs preterm model"</span>)</span>
<span id="cb5-59"><a href="#cb5-59"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(p.auc.combined.A, p.auc.bottom,<span class="at">ncol=</span><span class="dv">1</span>, <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"A"</span>,<span class="st">""</span>),<span class="at">label_size=</span><span class="dv">27</span>,<span class="at">rel_heights=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">1</span>), <span class="at">align=</span><span class="st">"v"</span>, <span class="at">axis=</span><span class="st">"lr"</span>)</span>
<span id="cb5-60"><a href="#cb5-60"></a><span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="sec-si-figure3" class="level2" data-number="9.3">
<h2 data-number="9.3" class="anchored" data-anchor-id="sec-si-figure3"><span class="header-section-number">9.3</span> Supplementary Text Figure 3</h2>
<div id="fig-si-fig3" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-si-fig3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="static/figure/cfRNA.SI.Fig3.png" id="fig-si-fig3" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-si-fig3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9.6
</figcaption>
</figure>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="fu">load</span>(<span class="st">"RData/dt.cpmZ.preterm.POPS-2022.GRCh38.88.RData"</span>) <span class="co"># dt.cpmZ (preterm)</span></span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="fu">load</span>(<span class="st">"RData/dt.cpmZ.term.POPS-2022.GRCh38.88.RData"</span>) <span class="co"># dt.cpmZ.term (term)</span></span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="fu">load</span>(<span class="st">"RData/dt.cpmZ.munchel.RData"</span>) <span class="co"># dt.cpmZ.munchel (Munchel)</span></span>
<span id="cb6-4"><a href="#cb6-4"></a></span>
<span id="cb6-5"><a href="#cb6-5"></a>dt.lep.pappa2<span class="ot">&lt;-</span><span class="fu">rbind</span>(</span>
<span id="cb6-6"><a href="#cb6-6"></a>                      (dt.cpmZ[geneName <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"LEP"</span>,<span class="st">"PAPPA2"</span>)] <span class="sc">%&gt;%</span> <span class="fu">dcast.data.table</span>(SampleID<span class="sc">+</span>GA<span class="sc">+</span>Condition<span class="sc">~</span>geneName, <span class="at">value.var=</span><span class="st">"logCPMZ"</span>))[,<span class="at">Source:=</span><span class="st">"Discovery"</span>],</span>
<span id="cb6-7"><a href="#cb6-7"></a>                      (dt.cpmZ.term[geneName <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"LEP"</span>,<span class="st">"PAPPA2"</span>)] <span class="sc">%&gt;%</span> <span class="fu">dcast.data.table</span>(SampleID<span class="sc">+</span>GA<span class="sc">+</span>Condition<span class="sc">~</span>geneName, <span class="at">value.var=</span><span class="st">"logCPMZ"</span>))[,<span class="at">Source:=</span><span class="st">"Validation"</span>],</span>
<span id="cb6-8"><a href="#cb6-8"></a>                      (dt.cpmZ.munchel[geneName <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"LEP"</span>,<span class="st">"PAPPA2"</span>)] <span class="sc">%&gt;%</span> <span class="fu">dcast.data.table</span>(SampleID<span class="sc">+</span>Condition<span class="sc">~</span>geneName, <span class="at">value.var=</span><span class="st">"logCPMZ"</span>))[,<span class="at">Source:=</span><span class="st">"Munchel"</span>]</span>
<span id="cb6-9"><a href="#cb6-9"></a>    ,<span class="at">fill=</span>T)</span>
<span id="cb6-10"><a href="#cb6-10"></a></span>
<span id="cb6-11"><a href="#cb6-11"></a>dt.lep.pappa2[,Dataset<span class="sc">:</span><span class="er">=</span><span class="fu">ifelse</span>(<span class="fu">is.na</span>(GA),Source,<span class="fu">paste0</span>(Source,<span class="st">"("</span>,GA,<span class="st">")"</span>))]</span>
<span id="cb6-12"><a href="#cb6-12"></a>dt.lep.pappa2[,.(<span class="at">Cor=</span><span class="fu">cor</span>(LEP,PAPPA2),<span class="at">Pval=</span><span class="fu">cor.test</span>(LEP,PAPPA2)<span class="sc">$</span>p.value),.(Source,GA)][<span class="fu">order</span>(Source,GA)] </span>
<span id="cb6-13"><a href="#cb6-13"></a>dt.lep.pappa2[Condition<span class="sc">==</span><span class="st">"Case"</span>,.(<span class="at">Cor=</span><span class="fu">cor</span>(LEP,PAPPA2),<span class="at">Pval=</span><span class="fu">cor.test</span>(LEP,PAPPA2)<span class="sc">$</span>p.value),.(Source,GA)][<span class="fu">order</span>(Source,GA)] </span>
<span id="cb6-14"><a href="#cb6-14"></a></span>
<span id="cb6-15"><a href="#cb6-15"></a>p1<span class="ot">&lt;-</span></span>
<span id="cb6-16"><a href="#cb6-16"></a><span class="fu">ggplot</span>(dt.lep.pappa2[Source<span class="sc">!=</span><span class="st">"Munchel"</span>], <span class="fu">aes</span>(LEP,PAPPA2)) <span class="sc">+</span></span>
<span id="cb6-17"><a href="#cb6-17"></a>  <span class="fu">geom_rect</span>(<span class="at">data =</span> dt.lep.pappa2[Source<span class="sc">==</span><span class="st">"Discovery"</span> <span class="sc">&amp;</span> GA<span class="sc">==</span><span class="st">"28wk"</span>],<span class="fu">aes</span>(<span class="at">fill =</span> GA),<span class="at">fill=</span><span class="st">"grey80"</span>,<span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">ymax =</span> <span class="cn">Inf</span>,<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb6-18"><a href="#cb6-18"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">col=</span>Condition),<span class="at">size=</span><span class="dv">3</span>,<span class="at">shape=</span><span class="dv">21</span>,<span class="at">stroke=</span><span class="fl">1.1</span>,<span class="at">alpha=</span>.<span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb6-19"><a href="#cb6-19"></a>  <span class="fu">facet_grid</span>(Source<span class="sc">~</span>GA) <span class="sc">+</span> </span>
<span id="cb6-20"><a href="#cb6-20"></a>  ggsci<span class="sc">::</span><span class="fu">scale_color_jama</span>() <span class="sc">+</span></span>
<span id="cb6-21"><a href="#cb6-21"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Z-score (LEP)"</span>,<span class="at">y=</span><span class="st">"Z-score (PAPPA2)"</span>) <span class="sc">+</span></span>
<span id="cb6-22"><a href="#cb6-22"></a>  <span class="fu">theme_Publication</span>() <span class="sc">+</span><span class="fu">theme</span> (<span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"black"</span>),<span class="at">legend.position=</span><span class="st">"top"</span>)</span>
<span id="cb6-23"><a href="#cb6-23"></a></span>
<span id="cb6-24"><a href="#cb6-24"></a>p2<span class="ot">&lt;-</span></span>
<span id="cb6-25"><a href="#cb6-25"></a><span class="fu">ggplot</span>(dt.lep.pappa2[Source<span class="sc">==</span><span class="st">"Munchel"</span>], <span class="fu">aes</span>(LEP,PAPPA2)) <span class="sc">+</span></span>
<span id="cb6-26"><a href="#cb6-26"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">col=</span>Condition),<span class="at">size=</span><span class="dv">3</span>,<span class="at">shape=</span><span class="dv">21</span>,<span class="at">stroke=</span><span class="fl">1.1</span>,<span class="at">alpha=</span>.<span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb6-27"><a href="#cb6-27"></a>  <span class="fu">facet_grid</span>(Source<span class="sc">~</span>.) <span class="sc">+</span> </span>
<span id="cb6-28"><a href="#cb6-28"></a>  ggsci<span class="sc">::</span><span class="fu">scale_color_jama</span>() <span class="sc">+</span></span>
<span id="cb6-29"><a href="#cb6-29"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">""</span>,<span class="at">y=</span><span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb6-30"><a href="#cb6-30"></a>  <span class="fu">theme_Publication</span>() <span class="sc">+</span><span class="fu">theme</span> (<span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"black"</span>),<span class="at">legend.position=</span><span class="st">""</span>)</span>
<span id="cb6-31"><a href="#cb6-31"></a></span>
<span id="cb6-32"><a href="#cb6-32"></a>p2.right<span class="ot">&lt;-</span>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(<span class="cn">NULL</span>,p2,<span class="at">labels=</span><span class="fu">c</span>(<span class="st">""</span>,<span class="st">"B"</span>),<span class="at">ncol=</span><span class="dv">1</span>,<span class="at">label_size=</span><span class="dv">27</span>,<span class="at">rel_heights=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb6-33"><a href="#cb6-33"></a>p.corr2<span class="ot">&lt;-</span>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(p1, p2.right, <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"A"</span>,<span class="st">""</span>),<span class="at">label_size=</span><span class="dv">27</span>,<span class="at">rel_widths=</span><span class="fu">c</span>(<span class="fl">2.8</span>,<span class="dv">1</span>))</span>
<span id="cb6-34"><a href="#cb6-34"></a></span>
<span id="cb6-35"><a href="#cb6-35"></a><span class="fu">cairo_pdf</span>(<span class="at">file=</span><span class="st">"Figures/Suppl/cfRNA.Suppl.LEP.PAPPA2.cor2.pdf"</span>, <span class="at">width=</span><span class="dv">12</span>,<span class="at">height=</span><span class="dv">6</span>, <span class="at">onefile=</span>T)</span>
<span id="cb6-36"><a href="#cb6-36"></a>p.corr2</span>
<span id="cb6-37"><a href="#cb6-37"></a><span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="sec-si-figure4-6" class="level2" data-number="9.4">
<h2 data-number="9.4" class="anchored" data-anchor-id="sec-si-figure4-6"><span class="header-section-number">9.4</span> Supplementary Text Figure 4, 5, and 6</h2>
<p><embed src="static/figure/cfRNA.SI.Fig4.5.6.pdf#zoom=50&amp;&amp;navpanes=0&amp;statusbar=0&amp;messages=0&amp;pagemode=none" style="width:100.0%" height="475"> <a href="static/figure/cfRNA.SI.Fig4.5.6.pdf">Full Screen</a></p>
<p>You may need to refer to <a href="change_by_gestational_age.html#sec-ga-long" class="quarto-xref"><span>Section 4.2</span></a> and <a href="main_figure.html#sec-fig5b" class="quarto-xref"><span>Section 8.5.2</span></a></p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>targets<span class="ot">&lt;-</span><span class="fu">c</span>(</span>
<span id="cb7-2"><a href="#cb7-2"></a>  dt.R2.GA2[<span class="fu">order</span>(<span class="sc">-</span>r2m)][[<span class="st">"gene"</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>], <span class="co"># top 100 by r2</span></span>
<span id="cb7-3"><a href="#cb7-3"></a>  dt.R2.GA2[<span class="fu">order</span>(p.combined)][[<span class="st">"gene"</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>], <span class="co"># top 100 by the combined p-value</span></span>
<span id="cb7-4"><a href="#cb7-4"></a>  dt.R2.GA[BH.GA<span class="sc">&lt;</span><span class="fl">0.05</span>][<span class="fu">order</span>(p.GA)][coef.GA<span class="sc">&lt;</span><span class="dv">0</span>][[<span class="st">"gene"</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>] <span class="co"># top 100 decreasing</span></span>
<span id="cb7-5"><a href="#cb7-5"></a>  ) <span class="sc">%&gt;%</span> unique</span>
<span id="cb7-6"><a href="#cb7-6"></a><span class="co">#</span></span>
<span id="cb7-7"><a href="#cb7-7"></a>li.plots<span class="ot">&lt;-</span><span class="fu">lapply</span>(dl.logCPM[targets], <span class="cf">function</span>(dt.foo){</span>
<span id="cb7-8"><a href="#cb7-8"></a>    my.gene<span class="ot">&lt;-</span><span class="fu">unique</span>(dt.foo[[<span class="st">"geneName"</span>]])</span>
<span id="cb7-9"><a href="#cb7-9"></a>    <span class="fu">message</span>(my.gene)</span>
<span id="cb7-10"><a href="#cb7-10"></a></span>
<span id="cb7-11"><a href="#cb7-11"></a>    r2m<span class="ot">&lt;-</span>li.lmer[[my.gene]][[<span class="st">"stat"</span>]][Model<span class="sc">==</span><span class="st">"GA2"</span>][[<span class="st">"r2m"</span>]] <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="dv">3</span>)</span>
<span id="cb7-12"><a href="#cb7-12"></a>    r2c<span class="ot">&lt;-</span>li.lmer[[my.gene]][[<span class="st">"stat"</span>]][Model<span class="sc">==</span><span class="st">"GA2"</span>][[<span class="st">"r2c"</span>]] <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="dv">3</span>)</span>
<span id="cb7-13"><a href="#cb7-13"></a>    <span class="co">#my.title=paste0(my.gene," (r2m:",r2m,", r2c:",r2c,")")</span></span>
<span id="cb7-14"><a href="#cb7-14"></a>    my.title<span class="ot">=</span><span class="fu">paste0</span>(my.gene,<span class="st">" (r2m:"</span>,r2m,<span class="st">")"</span>)</span>
<span id="cb7-15"><a href="#cb7-15"></a></span>
<span id="cb7-16"><a href="#cb7-16"></a>    my.title.y<span class="ot">&lt;-</span>latex2exp<span class="sc">::</span><span class="fu">TeX</span>(r<span class="st">"($</span><span class="sc">\t</span><span class="st">extbf{log_2 CPM}$)"</span>)</span>
<span id="cb7-17"><a href="#cb7-17"></a></span>
<span id="cb7-18"><a href="#cb7-18"></a>    <span class="fu">ggplot</span>(dt.foo, <span class="fu">aes</span>(<span class="at">x =</span> GAwk, <span class="at">y =</span> logCPM)) <span class="sc">+</span></span>
<span id="cb7-19"><a href="#cb7-19"></a>    <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">group =</span> POPSID), <span class="at">linewidth =</span> .<span class="dv">5</span>, <span class="at">col=</span><span class="st">"grey80"</span>,<span class="at">alpha=</span>.<span class="dv">7</span>) <span class="sc">+</span></span>
<span id="cb7-20"><a href="#cb7-20"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> <span class="fu">as.factor</span>(GA)), <span class="at">size=</span><span class="dv">2</span>, <span class="at">alpha=</span>.<span class="dv">7</span>) <span class="sc">+</span></span>
<span id="cb7-21"><a href="#cb7-21"></a>    <span class="fu">labs</span>(<span class="at">subtitle=</span>my.title, <span class="at">x =</span> <span class="st">"Gestational Age (weeks)"</span>, <span class="at">y=</span>my.title.y, <span class="at">color =</span> <span class="st">"GA (week)"</span>) <span class="sc">+</span></span>
<span id="cb7-22"><a href="#cb7-22"></a>    <span class="fu">geom_ribbon</span>(<span class="at">data=</span>li.lmer[[my.gene]][[<span class="st">"predict"</span>]], <span class="fu">aes</span>(<span class="at">ymin=</span>conf.low, <span class="at">ymax=</span>conf.high), <span class="at">alpha=</span><span class="fl">0.6</span>, <span class="at">fill =</span> <span class="st">"grey40"</span>) <span class="sc">+</span></span>
<span id="cb7-23"><a href="#cb7-23"></a>    <span class="fu">geom_line</span>(<span class="at">data=</span>li.lmer[[my.gene]][[<span class="st">"predict"</span>]], <span class="at">linewidth=</span><span class="fl">1.5</span>, <span class="at">col=</span><span class="st">"blue"</span>,<span class="at">alpha=</span><span class="fl">0.8</span>) <span class="sc">+</span> </span>
<span id="cb7-24"><a href="#cb7-24"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">breaks=</span><span class="fu">c</span>(<span class="dv">12</span>,<span class="dv">20</span>,<span class="dv">28</span>,<span class="dv">36</span>)) <span class="sc">+</span></span>
<span id="cb7-25"><a href="#cb7-25"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values=</span>cbPalette2) <span class="sc">+</span></span>
<span id="cb7-26"><a href="#cb7-26"></a>    <span class="fu">theme_Publication</span>() <span class="sc">+</span></span>
<span id="cb7-27"><a href="#cb7-27"></a>    <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>)</span>
<span id="cb7-28"><a href="#cb7-28"></a>})</span>
<span id="cb7-29"><a href="#cb7-29"></a></span>
<span id="cb7-30"><a href="#cb7-30"></a><span class="co"># Top 100 by R2M</span></span>
<span id="cb7-31"><a href="#cb7-31"></a>file.name1<span class="ot">&lt;-</span><span class="fu">file.path</span>(<span class="st">'Figures/Suppl'</span>,<span class="fu">paste0</span>(<span class="st">'cfRNA.Suppl.Data.Fig.logCPM_by_GA_lmer.pdf'</span>))</span>
<span id="cb7-32"><a href="#cb7-32"></a><span class="fu">cairo_pdf</span>(<span class="at">filename=</span>file.name1, <span class="at">width=</span><span class="dv">14</span>, <span class="at">height=</span><span class="dv">9</span>, <span class="at">onefile=</span>T)</span>
<span id="cb7-33"><a href="#cb7-33"></a><span class="fu">grid.text</span>(<span class="at">label=</span><span class="fu">paste0</span>(<span class="st">"Supplementary Figure 4.</span><span class="sc">\n</span><span class="st">Top 100 cfRNAs</span><span class="sc">\n</span><span class="st">by the coefficient of determination"</span>),<span class="at">gp=</span><span class="fu">gpar</span>(<span class="at">fontsize=</span><span class="dv">50</span>,<span class="at">fontface=</span><span class="st">"bold"</span>),<span class="at">x=</span>.<span class="dv">5</span>,<span class="at">y=</span>.<span class="dv">6</span>,<span class="at">just=</span><span class="st">"center"</span>)</span>
<span id="cb7-34"><a href="#cb7-34"></a>gridExtra<span class="sc">::</span><span class="fu">marrangeGrob</span>(li.plots[dt.R2.GA2[<span class="fu">order</span>(<span class="sc">-</span>r2m)][[<span class="st">"gene"</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>]],<span class="at">nrow=</span><span class="dv">2</span>,<span class="at">ncol=</span><span class="dv">4</span>,<span class="at">top=</span><span class="cn">NULL</span>)</span>
<span id="cb7-35"><a href="#cb7-35"></a></span>
<span id="cb7-36"><a href="#cb7-36"></a><span class="co"># Top 100 by p.combined</span></span>
<span id="cb7-37"><a href="#cb7-37"></a><span class="fu">grid.newpage</span>()</span>
<span id="cb7-38"><a href="#cb7-38"></a><span class="fu">grid.text</span>(<span class="at">label=</span><span class="fu">paste0</span>(<span class="st">"Supplementary Figure 5.</span><span class="sc">\n</span><span class="st">Top 100 cfRNAs</span><span class="sc">\n</span><span class="st">by the combined p-values"</span>),<span class="at">gp=</span><span class="fu">gpar</span>(<span class="at">fontsize=</span><span class="dv">50</span>,<span class="at">fontface=</span><span class="st">"bold"</span>),<span class="at">x=</span>.<span class="dv">5</span>,<span class="at">y=</span>.<span class="dv">6</span>,<span class="at">just=</span><span class="st">"center"</span>)</span>
<span id="cb7-39"><a href="#cb7-39"></a>gridExtra<span class="sc">::</span><span class="fu">marrangeGrob</span>(li.plots[dt.R2.GA2[<span class="fu">order</span>(p.combined)][[<span class="st">"gene"</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>]],<span class="at">nrow=</span><span class="dv">2</span>,<span class="at">ncol=</span><span class="dv">4</span>,<span class="at">top=</span><span class="cn">NULL</span>)</span>
<span id="cb7-40"><a href="#cb7-40"></a></span>
<span id="cb7-41"><a href="#cb7-41"></a><span class="co"># Top 100 by negative coefficient in linear mixed effect model (GA only)</span></span>
<span id="cb7-42"><a href="#cb7-42"></a><span class="fu">grid.newpage</span>()</span>
<span id="cb7-43"><a href="#cb7-43"></a><span class="fu">grid.text</span>(<span class="at">label=</span><span class="fu">paste0</span>(<span class="st">"Supplementary Figure 6.</span><span class="sc">\n</span><span class="st">Top 100 cfRNAs</span><span class="sc">\n</span><span class="st">significantly decreasing</span><span class="sc">\n</span><span class="st">(negative coefficient)</span><span class="sc">\n</span><span class="st">by gestational age"</span>),<span class="at">gp=</span><span class="fu">gpar</span>(<span class="at">fontsize=</span><span class="dv">50</span>,<span class="at">fontface=</span><span class="st">"bold"</span>),<span class="at">x=</span>.<span class="dv">5</span>,<span class="at">y=</span>.<span class="dv">6</span>,<span class="at">just=</span><span class="st">"center"</span>)</span>
<span id="cb7-44"><a href="#cb7-44"></a>gridExtra<span class="sc">::</span><span class="fu">marrangeGrob</span>(li.plots[dt.R2.GA[BH.GA<span class="sc">&lt;</span><span class="fl">0.05</span>][<span class="fu">order</span>(p.GA)][coef.GA<span class="sc">&lt;</span><span class="dv">0</span>][[<span class="st">"gene"</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>]],<span class="at">nrow=</span><span class="dv">2</span>,<span class="at">ncol=</span><span class="dv">4</span>,<span class="at">top=</span><span class="cn">NULL</span>)</span>
<span id="cb7-45"><a href="#cb7-45"></a><span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>


<!-- -->

</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    // Ensure there is a toggle, if there isn't float one in the top right
    if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
      const a = window.document.createElement('a');
      a.classList.add('top-right');
      a.classList.add('quarto-color-scheme-toggle');
      a.href = "";
      a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
      const i = window.document.createElement("i");
      i.classList.add('bi');
      a.appendChild(i);
      window.document.body.appendChild(a);
    }
    setColorSchemeToggle(hasAlternateSentinel())
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/ObsGynaeCam\.github\.io\/cell-free-rna-2024");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./main_figure.html" class="pagination-link" aria-label="Main Figures">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Main Figures</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb8" data-shortcodes="false"><pre class="sourceCode numberSource markdown number-lines code-with-copy"><code class="sourceCode markdown"><span id="cb8-1"><a href="#cb8-1"></a><span class="fu"># Supplementary Text Figures {#sec-suppl-figure}</span></span>
<span id="cb8-2"><a href="#cb8-2"></a></span>
<span id="cb8-3"><a href="#cb8-3"></a>This chapter demonstrates the <span class="in">`R`</span> codes that make supplementary text figures shown in the paper. </span>
<span id="cb8-4"><a href="#cb8-4"></a>Making the main figures are shown in @sec-main-figure.</span>
<span id="cb8-5"><a href="#cb8-5"></a></span>
<span id="cb8-6"><a href="#cb8-6"></a><span class="fu">## Supplementary Text Figure 1 {#sec-si-figure1}</span></span>
<span id="cb8-7"><a href="#cb8-7"></a></span>
<span id="cb8-8"><a href="#cb8-8"></a><span class="fu">### Supplementary Text Figure 1A {#sec-si-figure1-a}</span></span>
<span id="cb8-9"><a href="#cb8-9"></a><span class="al">![](static/figure/cfRNA.SI.FigA.png)</span>{#fig-si-fig1a}</span>
<span id="cb8-10"><a href="#cb8-10"></a></span>
<span id="cb8-11"><a href="#cb8-11"></a><span class="in">```{r }</span></span>
<span id="cb8-12"><a href="#cb8-12"></a><span class="co">#| label: si-fig-1a</span></span>
<span id="cb8-13"><a href="#cb8-13"></a><span class="co">#| eval: false</span></span>
<span id="cb8-14"><a href="#cb8-14"></a>  <span class="fu">lazyLoad</span>(<span class="st">"downsample.chrY.TPM_cache/beamer/load_ds_salmon_4b8dd727729fd0ae5deb423d69e7c16d"</span>) <span class="co"># dl.chrY.TPM and dl.stat</span></span>
<span id="cb8-15"><a href="#cb8-15"></a></span>
<span id="cb8-16"><a href="#cb8-16"></a>  dt.ds.chrY <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(dl.stat), <span class="cf">function</span>(DS) dl.stat[[DS]][,<span class="at">DS:=</span>DS]) <span class="sc">%&gt;%</span> rbindlist</span>
<span id="cb8-17"><a href="#cb8-17"></a>  dt.ds.chrY<span class="sc">$</span>DS<span class="ot">&lt;-</span><span class="fu">factor</span>(dt.ds.chrY<span class="sc">$</span>DS, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"1M"</span>,<span class="st">"5M"</span>,<span class="st">"10M"</span>))</span>
<span id="cb8-18"><a href="#cb8-18"></a>  dt.ds.chrY[,GA<span class="sc">:</span><span class="er">=</span><span class="fu">gsub</span>(<span class="st">" weeks"</span>,<span class="st">"wk"</span>,GA)]</span>
<span id="cb8-19"><a href="#cb8-19"></a>  <span class="fu">setnames</span>(dt.ds.chrY, <span class="st">"Type"</span>, <span class="st">"Method"</span>)</span>
<span id="cb8-20"><a href="#cb8-20"></a></span>
<span id="cb8-21"><a href="#cb8-21"></a>  p1.chrY<span class="ot">&lt;-</span><span class="fu">ggplot</span>(dt.ds.chrY, <span class="fu">aes</span>(Method,SumCount)) <span class="sc">+</span></span>
<span id="cb8-22"><a href="#cb8-22"></a>  <span class="fu">geom_rect</span>(<span class="at">data =</span> dt.ds.chrY[DS<span class="sc">==</span><span class="st">"5M"</span>],<span class="at">fill=</span><span class="st">"grey60"</span>,<span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">ymax =</span> <span class="cn">Inf</span>,<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb8-23"><a href="#cb8-23"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">fill=</span>Method), <span class="at">stat=</span><span class="st">"identity"</span>,<span class="at">width=</span>.<span class="dv">5</span>, <span class="at">alpha=</span>.<span class="dv">7</span>) <span class="sc">+</span></span>
<span id="cb8-24"><a href="#cb8-24"></a>  ggsci<span class="sc">::</span><span class="fu">scale_fill_jama</span>() <span class="sc">+</span></span>
<span id="cb8-25"><a href="#cb8-25"></a>  <span class="fu">facet_grid</span>(Sex<span class="sc">~</span>DS<span class="sc">+</span>GA,<span class="at">scales=</span><span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb8-26"><a href="#cb8-26"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">"Sum of read count on chrY genes"</span>) <span class="sc">+</span></span>
<span id="cb8-27"><a href="#cb8-27"></a>  <span class="fu">theme_Publication</span>() <span class="sc">+</span></span>
<span id="cb8-28"><a href="#cb8-28"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"top"</span>,</span>
<span id="cb8-29"><a href="#cb8-29"></a>        <span class="at">panel.border=</span><span class="fu">element_rect</span>(<span class="at">colour=</span><span class="st">"black"</span>),</span>
<span id="cb8-30"><a href="#cb8-30"></a>        <span class="at">axis.title.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb8-31"><a href="#cb8-31"></a>        <span class="at">axis.text.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb8-32"><a href="#cb8-32"></a>        <span class="at">axis.ticks.x=</span><span class="fu">element_blank</span>())</span>
<span id="cb8-33"><a href="#cb8-33"></a></span>
<span id="cb8-34"><a href="#cb8-34"></a>  <span class="co">#</span></span>
<span id="cb8-35"><a href="#cb8-35"></a>  <span class="fu">pdf</span>(<span class="at">file=</span><span class="st">"Figures/Suppl/cfRNA.Suppl.Text.Fig1A.pdf"</span>, <span class="at">width=</span><span class="dv">11</span>, <span class="at">height=</span><span class="dv">8</span>, <span class="at">title=</span><span class="st">"Abundance of chrY"</span>)</span>
<span id="cb8-36"><a href="#cb8-36"></a>  p1.chrY</span>
<span id="cb8-37"><a href="#cb8-37"></a>  <span class="fu">dev.off</span>()</span>
<span id="cb8-38"><a href="#cb8-38"></a><span class="in">```</span></span>
<span id="cb8-39"><a href="#cb8-39"></a></span>
<span id="cb8-40"><a href="#cb8-40"></a><span class="fu">### Supplementary Text Figure 1B {#sec-si-figure-1b}</span></span>
<span id="cb8-41"><a href="#cb8-41"></a><span class="al">![](static/figure/cfRNA.SI.FigB.png)</span>{#fig-si-fig1b}</span>
<span id="cb8-42"><a href="#cb8-42"></a></span>
<span id="cb8-43"><a href="#cb8-43"></a><span class="in">```{r }</span></span>
<span id="cb8-44"><a href="#cb8-44"></a><span class="co">#| label: si-fig-1b</span></span>
<span id="cb8-45"><a href="#cb8-45"></a><span class="co">#| eval: false</span></span>
<span id="cb8-46"><a href="#cb8-46"></a>  <span class="fu">lazyLoad</span>(<span class="st">"downsample.chrY.TPM_cache/beamer/load_ds_salmon_4b8dd727729fd0ae5deb423d69e7c16d"</span>) <span class="co"># dl.chrY.TPM and dl.stat</span></span>
<span id="cb8-47"><a href="#cb8-47"></a></span>
<span id="cb8-48"><a href="#cb8-48"></a>  <span class="fu">lapply</span>(dl.chrY.TPM, <span class="cf">function</span>(i) i[,.(<span class="fu">sum</span>(Counts),<span class="fu">sum</span>(TPM))])</span>
<span id="cb8-49"><a href="#cb8-49"></a></span>
<span id="cb8-50"><a href="#cb8-50"></a>  dl.chrY.TPM<span class="ot">&lt;-</span><span class="fu">lapply</span>(dl.chrY.TPM, <span class="cf">function</span>(DT) {</span>
<span id="cb8-51"><a href="#cb8-51"></a>                  <span class="co">#DT[Sex=="Male",Class:=ifelse(TPM&gt;0,"TP","FN")]</span></span>
<span id="cb8-52"><a href="#cb8-52"></a>                  <span class="co">#DT[Sex=="Female",Class:=ifelse(TPM==0,"TN","FP")] </span></span>
<span id="cb8-53"><a href="#cb8-53"></a>                  DT[,<span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span><span class="sc">:</span><span class="er">=</span><span class="fu">ifelse</span>(TPM<span class="sc">&gt;</span><span class="dv">0</span>, <span class="st">"Male"</span>,<span class="st">"Female"</span>),Type]</span>
<span id="cb8-54"><a href="#cb8-54"></a>                  DT<span class="sc">$</span>Sex<span class="ot">&lt;-</span><span class="fu">factor</span>(DT<span class="sc">$</span>Sex, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"Male"</span>,<span class="st">"Female"</span>))</span>
<span id="cb8-55"><a href="#cb8-55"></a>                  DT<span class="sc">$</span><span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span><span class="ot">&lt;-</span><span class="fu">factor</span>(DT<span class="sc">$</span><span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span>, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"Male"</span>,<span class="st">"Female"</span>))</span>
<span id="cb8-56"><a href="#cb8-56"></a>                  DT </span>
<span id="cb8-57"><a href="#cb8-57"></a>  }) </span>
<span id="cb8-58"><a href="#cb8-58"></a></span>
<span id="cb8-59"><a href="#cb8-59"></a>  <span class="do">#####################################</span></span>
<span id="cb8-60"><a href="#cb8-60"></a>  <span class="co"># Predictive performance by DS &amp; GA #</span></span>
<span id="cb8-61"><a href="#cb8-61"></a>  <span class="do">#####################################</span></span>
<span id="cb8-62"><a href="#cb8-62"></a>  dt.chrY.con.ga<span class="ot">&lt;-</span><span class="fu">lapply</span>(<span class="fu">names</span>(dl.chrY.TPM), <span class="cf">function</span>(DS){</span>
<span id="cb8-63"><a href="#cb8-63"></a>    <span class="fu">lapply</span>(<span class="fu">split</span>(dl.chrY.TPM[[DS]], dl.chrY.TPM[[DS]]<span class="sc">$</span>GA), <span class="cf">function</span>(DT){</span>
<span id="cb8-64"><a href="#cb8-64"></a>    my.GA<span class="ot">&lt;-</span>DT[,.N,GA]<span class="sc">$</span>GA</span>
<span id="cb8-65"><a href="#cb8-65"></a></span>
<span id="cb8-66"><a href="#cb8-66"></a>    cm1<span class="ot">&lt;-</span>caret<span class="sc">::</span><span class="fu">confusionMatrix</span>(DT[Type<span class="sc">==</span><span class="st">"HiSat2+Salmon"</span>][[<span class="st">"Predicted Sex"</span>]], <span class="at">reference=</span>DT[Type<span class="sc">==</span><span class="st">"HiSat2+Salmon"</span>][[<span class="st">"Sex"</span>]])</span>
<span id="cb8-67"><a href="#cb8-67"></a>    cm2<span class="ot">&lt;-</span>caret<span class="sc">::</span><span class="fu">confusionMatrix</span>(DT[Type<span class="sc">==</span><span class="st">"Salmon (SA mode)"</span>][[<span class="st">"Predicted Sex"</span>]], <span class="at">reference=</span>DT[Type<span class="sc">==</span><span class="st">"Salmon (SA mode)"</span>][[<span class="st">"Sex"</span>]])</span>
<span id="cb8-68"><a href="#cb8-68"></a></span>
<span id="cb8-69"><a href="#cb8-69"></a>    <span class="fu">rbind</span>(</span>
<span id="cb8-70"><a href="#cb8-70"></a>    <span class="fu">data.table</span>(<span class="at">DS=</span>DS, <span class="at">GA=</span>my.GA,<span class="at">Type=</span><span class="st">"HiSat2+Salmon"</span>, <span class="at">Measure=</span><span class="fu">names</span>(cm1<span class="sc">$</span>byClass), <span class="at">Value=</span>cm1<span class="sc">$</span>byClass),</span>
<span id="cb8-71"><a href="#cb8-71"></a>    <span class="fu">data.table</span>(<span class="at">DS=</span>DS, <span class="at">GA=</span>my.GA,<span class="at">Type=</span><span class="st">"Salmon (SA mode)"</span>, <span class="at">Measure=</span><span class="fu">names</span>(cm2<span class="sc">$</span>byClass), <span class="at">Value=</span>cm2<span class="sc">$</span>byClass)</span>
<span id="cb8-72"><a href="#cb8-72"></a>    )[Measure <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"F1"</span>,<span class="st">"Precision"</span>, <span class="st">"Recall"</span>)]</span>
<span id="cb8-73"><a href="#cb8-73"></a>    }) <span class="sc">%&gt;%</span> rbindlist </span>
<span id="cb8-74"><a href="#cb8-74"></a>  }) <span class="sc">%&gt;%</span> rbindlist </span>
<span id="cb8-75"><a href="#cb8-75"></a></span>
<span id="cb8-76"><a href="#cb8-76"></a>  dt.chrY.con.ga<span class="sc">$</span>DS<span class="ot">&lt;-</span><span class="fu">factor</span>(dt.chrY.con.ga<span class="sc">$</span>DS, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"1M"</span>,<span class="st">"5M"</span>,<span class="st">"10M"</span>))</span>
<span id="cb8-77"><a href="#cb8-77"></a>  dt.chrY.con.ga[,GA<span class="sc">:</span><span class="er">=</span><span class="fu">gsub</span>(<span class="st">" weeks"</span>,<span class="st">"wk"</span>,GA)]</span>
<span id="cb8-78"><a href="#cb8-78"></a>  <span class="fu">setnames</span>(dt.chrY.con.ga, <span class="st">"Type"</span>, <span class="st">"Method"</span>)</span>
<span id="cb8-79"><a href="#cb8-79"></a></span>
<span id="cb8-80"><a href="#cb8-80"></a>  p1.chrY.con<span class="ot">&lt;-</span><span class="fu">ggplot</span>(dt.chrY.con.ga, <span class="fu">aes</span>(Method,Value)) <span class="sc">+</span></span>
<span id="cb8-81"><a href="#cb8-81"></a>  <span class="fu">geom_rect</span>(<span class="at">data =</span> dt.chrY.con.ga[DS<span class="sc">==</span><span class="st">"5M"</span>],<span class="at">fill=</span><span class="st">"grey60"</span>,<span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">ymax =</span> <span class="cn">Inf</span>,<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb8-82"><a href="#cb8-82"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">fill=</span>Method), <span class="at">stat=</span><span class="st">"identity"</span>,<span class="at">width=</span>.<span class="dv">5</span>, <span class="at">alpha=</span>.<span class="dv">7</span>) <span class="sc">+</span></span>
<span id="cb8-83"><a href="#cb8-83"></a>  ggsci<span class="sc">::</span><span class="fu">scale_fill_jama</span>() <span class="sc">+</span></span>
<span id="cb8-84"><a href="#cb8-84"></a>  <span class="fu">facet_grid</span>(Measure<span class="sc">~</span>DS<span class="sc">+</span>GA,<span class="at">scales=</span><span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb8-85"><a href="#cb8-85"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">"Score"</span>) <span class="sc">+</span></span>
<span id="cb8-86"><a href="#cb8-86"></a>  <span class="fu">theme_Publication</span>() <span class="sc">+</span></span>
<span id="cb8-87"><a href="#cb8-87"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>,</span>
<span id="cb8-88"><a href="#cb8-88"></a>        <span class="at">panel.border=</span><span class="fu">element_rect</span>(<span class="at">colour=</span><span class="st">"black"</span>),</span>
<span id="cb8-89"><a href="#cb8-89"></a>        <span class="at">axis.title.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb8-90"><a href="#cb8-90"></a>        <span class="at">axis.text.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb8-91"><a href="#cb8-91"></a>        <span class="at">axis.ticks.x=</span><span class="fu">element_blank</span>())</span>
<span id="cb8-92"><a href="#cb8-92"></a></span>
<span id="cb8-93"><a href="#cb8-93"></a></span>
<span id="cb8-94"><a href="#cb8-94"></a>  <span class="co"># merge SI.Fig1A + SI.Fig1B</span></span>
<span id="cb8-95"><a href="#cb8-95"></a>  <span class="fu">pdf</span>(<span class="at">file=</span><span class="st">"Figures/Suppl/cfRNA.Suppl.Text.Fig1AB.pdf"</span>, <span class="at">width=</span><span class="dv">11</span>, <span class="at">height=</span><span class="dv">14</span>, <span class="at">title=</span><span class="st">"Abundance of chrY"</span>)</span>
<span id="cb8-96"><a href="#cb8-96"></a>  cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(p1.chrY, p1.chrY.con, <span class="at">labels=</span><span class="st">"AUTO"</span>, <span class="at">ncol=</span><span class="dv">1</span>, <span class="at">label_size=</span><span class="dv">27</span>, <span class="at">align=</span><span class="st">"v"</span>, <span class="at">axis=</span><span class="st">"l"</span>)</span>
<span id="cb8-97"><a href="#cb8-97"></a>  <span class="fu">dev.off</span>()</span>
<span id="cb8-98"><a href="#cb8-98"></a><span class="in">```</span></span>
<span id="cb8-99"><a href="#cb8-99"></a></span>
<span id="cb8-100"><a href="#cb8-100"></a><span class="fu">### Supplementary Text Figure 1C {#sec-si-figure-1c}</span></span>
<span id="cb8-101"><a href="#cb8-101"></a><span class="al">![](static/figure/cfRNA.SI.FigC.png)</span>{#fig-si-fig1c}</span>
<span id="cb8-102"><a href="#cb8-102"></a></span>
<span id="cb8-103"><a href="#cb8-103"></a><span class="in">```{r }</span></span>
<span id="cb8-104"><a href="#cb8-104"></a><span class="co">#| label: si-fig-1c</span></span>
<span id="cb8-105"><a href="#cb8-105"></a><span class="co">#| eval: false</span></span>
<span id="cb8-106"><a href="#cb8-106"></a>  <span class="co"># load the transcript definition used in Salmon (SA)</span></span>
<span id="cb8-107"><a href="#cb8-107"></a>  <span class="fu">load</span>(<span class="st">"RData/gr.ensg.Homo_sapiens.GRCh38.88.cdna.all.ncrna.fa.gz.RData"</span>) <span class="sc">%&gt;%</span> system.time <span class="co"># isa GRanges</span></span>
<span id="cb8-108"><a href="#cb8-108"></a>  gr.ensg[<span class="fu">seqnames</span>(gr.ensg)<span class="sc">==</span><span class="st">"Y"</span> <span class="sc">&amp;</span> gr.ensg<span class="sc">$</span>gene_biotype<span class="sc">==</span><span class="st">"protein_coding"</span>,] <span class="co"># chrY protein-coding; n=42 </span></span>
<span id="cb8-109"><a href="#cb8-109"></a>  this.gene<span class="ot">&lt;-</span>(gr.ensg[<span class="fu">seqnames</span>(gr.ensg)<span class="sc">==</span><span class="st">"Y"</span> <span class="sc">&amp;</span> gr.ensg<span class="sc">$</span>gene_biotype<span class="sc">==</span><span class="st">"protein_coding"</span>,] <span class="sc">%&gt;%</span> names) <span class="co"># chrY protein-coding; n=42 (without the training version e.g. ENSG00000176679)</span></span>
<span id="cb8-110"><a href="#cb8-110"></a></span>
<span id="cb8-111"><a href="#cb8-111"></a>  <span class="co"># load the TPM based om 10M-ds dataset</span></span>
<span id="cb8-112"><a href="#cb8-112"></a>  <span class="fu">load</span>(<span class="st">"RData/li.TPM.10M.RData"</span>)</span>
<span id="cb8-113"><a href="#cb8-113"></a>  li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"TPM"</span>]] <span class="sc">%&gt;%</span> dim <span class="co"># 62803 genes x 100 samples</span></span>
<span id="cb8-114"><a href="#cb8-114"></a></span>
<span id="cb8-115"><a href="#cb8-115"></a>  <span class="do">## Heatamp of TPM using `HiSat2+Salmon` (based on 10M)</span></span>
<span id="cb8-116"><a href="#cb8-116"></a>  li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> rowSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="co"># count of 42 chrY genes</span></span>
<span id="cb8-117"><a href="#cb8-117"></a>  filter.g2<span class="ot">&lt;-</span>li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> rowSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="sc">&gt;</span><span class="dv">0</span> <span class="co"># filter for 42 chrY genes with TPM &gt;0</span></span>
<span id="cb8-118"><a href="#cb8-118"></a>  this.gene2<span class="ot">&lt;-</span>(li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> rowSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev)[filter.g2] <span class="sc">%&gt;%</span> names <span class="co"># n=24 genes</span></span>
<span id="cb8-119"><a href="#cb8-119"></a>  this.sample2<span class="ot">&lt;-</span>li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="sc">%&gt;%</span> names <span class="co"># n=100 samples sorted </span></span>
<span id="cb8-120"><a href="#cb8-120"></a>  filter.s2<span class="ot">&lt;-</span>li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene2,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="sc">&gt;</span><span class="dv">0</span> <span class="co"># filter for samples with TPM &gt;0 </span></span>
<span id="cb8-121"><a href="#cb8-121"></a>  predicted.male2<span class="ot">&lt;-</span>(li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev)[filter.s2] <span class="sc">%&gt;%</span> names <span class="co"># n=60 samples </span></span>
<span id="cb8-122"><a href="#cb8-122"></a></span>
<span id="cb8-123"><a href="#cb8-123"></a>  my.mat2<span class="ot">&lt;-</span><span class="fu">log2</span>(li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene2,this.sample2]<span class="sc">+</span><span class="dv">1</span>)</span>
<span id="cb8-124"><a href="#cb8-124"></a>  <span class="fu">rownames</span>(my.mat2)<span class="ot">&lt;-</span>gr.ensg[this.gene2]<span class="sc">$</span>gene_name</span>
<span id="cb8-125"><a href="#cb8-125"></a></span>
<span id="cb8-126"><a href="#cb8-126"></a>  col.anno<span class="ot">&lt;-</span>mat.samples[this.sample2,<span class="fu">c</span>(<span class="st">"GA"</span>,<span class="st">"Sex"</span>),drop<span class="ot">=</span>F] <span class="co"># isa `data.frame`</span></span>
<span id="cb8-127"><a href="#cb8-127"></a>  col.anno<span class="sc">$</span><span class="st">"Predicted Sex"</span><span class="ot">&lt;-</span><span class="fu">ifelse</span>(this.sample2 <span class="sc">%in%</span> predicted.male2,<span class="st">"Male"</span>,<span class="st">"Female"</span>)</span>
<span id="cb8-128"><a href="#cb8-128"></a>  col.anno<span class="ot">&lt;-</span>col.anno[,<span class="fu">c</span>(<span class="st">"Predicted Sex"</span>,<span class="st">"Sex"</span>,<span class="st">"GA"</span>)]</span>
<span id="cb8-129"><a href="#cb8-129"></a></span>
<span id="cb8-130"><a href="#cb8-130"></a>  GA.color<span class="ot">&lt;-</span>ggsci<span class="sc">::</span><span class="fu">pal_jama</span>(<span class="st">"default"</span>,<span class="at">alpha=</span>.<span class="dv">9</span>)(<span class="dv">4</span>)</span>
<span id="cb8-131"><a href="#cb8-131"></a>  <span class="fu">names</span>(GA.color)<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"12wk"</span>,<span class="st">"20wk"</span>,<span class="st">"28wk"</span>,<span class="st">"36wk"</span>)</span>
<span id="cb8-132"><a href="#cb8-132"></a>  my.color<span class="ot">=</span><span class="fu">list</span>(</span>
<span id="cb8-133"><a href="#cb8-133"></a>      <span class="st">`</span><span class="at">Sex</span><span class="st">`</span><span class="ot">=</span><span class="fu">c</span>(<span class="at">Female=</span><span class="st">"hotpink"</span>,<span class="at">Male=</span><span class="st">'skyblue'</span>),</span>
<span id="cb8-134"><a href="#cb8-134"></a>      <span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span><span class="ot">=</span><span class="fu">c</span>(<span class="at">Female=</span><span class="st">"hotpink3"</span>,<span class="at">Male=</span><span class="st">'skyblue3'</span>),</span>
<span id="cb8-135"><a href="#cb8-135"></a>      <span class="co">#`GA`=c("12wk"=cbPalette2[1],"20wk"=cbPalette2[2],"28wk"=cbPalette2[3],"36wk"=cbPalette2[4])</span></span>
<span id="cb8-136"><a href="#cb8-136"></a>      <span class="st">`</span><span class="at">GA</span><span class="st">`</span><span class="ot">=</span>GA.color</span>
<span id="cb8-137"><a href="#cb8-137"></a>  )</span>
<span id="cb8-138"><a href="#cb8-138"></a></span>
<span id="cb8-139"><a href="#cb8-139"></a>  cph1<span class="ot">&lt;-</span>ComplexHeatmap<span class="sc">::</span><span class="fu">pheatmap</span>(my.mat2,</span>
<span id="cb8-140"><a href="#cb8-140"></a>                                 <span class="at">annotation_col=</span>col.anno, </span>
<span id="cb8-141"><a href="#cb8-141"></a>                                 <span class="at">annotation_colors=</span>my.color, </span>
<span id="cb8-142"><a href="#cb8-142"></a>                                 <span class="at">cluster_cols=</span>F,</span>
<span id="cb8-143"><a href="#cb8-143"></a>                                 <span class="at">cluster_rows=</span>F,</span>
<span id="cb8-144"><a href="#cb8-144"></a>                                 <span class="at">show_colnames=</span>F,</span>
<span id="cb8-145"><a href="#cb8-145"></a>                                 <span class="at">border_color=</span><span class="cn">NA</span>,</span>
<span id="cb8-146"><a href="#cb8-146"></a>                                 <span class="at">fontsize=</span><span class="dv">12</span>,</span>
<span id="cb8-147"><a href="#cb8-147"></a>                                 <span class="at">name=</span><span class="st">"log2(count)"</span>,</span>
<span id="cb8-148"><a href="#cb8-148"></a>                                 <span class="at">row_title=</span><span class="st">"</span><span class="sc">\n</span><span class="st">HiSat2+Salmon"</span>,</span>
<span id="cb8-149"><a href="#cb8-149"></a>                                 <span class="at">row_title_gp =</span> <span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">20</span>),</span>
<span id="cb8-150"><a href="#cb8-150"></a>  )</span>
<span id="cb8-151"><a href="#cb8-151"></a></span>
<span id="cb8-152"><a href="#cb8-152"></a>  filter.g3<span class="ot">&lt;-</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> rowSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="sc">&gt;</span><span class="dv">0</span> <span class="co"># filter for 42 chrY genes with count &gt;0</span></span>
<span id="cb8-153"><a href="#cb8-153"></a>  this.gene3<span class="ot">&lt;-</span>(li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> rowSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev)[filter.g3] <span class="sc">%&gt;%</span> names <span class="co"># n=16 genes (out of 42) &gt;0 count </span></span>
<span id="cb8-154"><a href="#cb8-154"></a>  this.sample3<span class="ot">&lt;-</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="sc">%&gt;%</span> names <span class="co"># 100 sample names</span></span>
<span id="cb8-155"><a href="#cb8-155"></a>  filter.s3<span class="ot">&lt;-</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene3,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="sc">&gt;</span><span class="dv">0</span> <span class="co"># samples with count&gt;0 </span></span>
<span id="cb8-156"><a href="#cb8-156"></a>  <span class="co">#(li.TPM[["Salmon"]][["counts"]][this.gene,] %&gt;% colSums %&gt;% sort %&gt;% rev)[filter.s3] # n=38 samples with count&gt;0 (predicted as Male)</span></span>
<span id="cb8-157"><a href="#cb8-157"></a>  predicted.male3<span class="ot">&lt;-</span>(li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev)[filter.s3] <span class="sc">%&gt;%</span> names <span class="co"># n=38 samples with count&gt;0 (predicted as Male)</span></span>
<span id="cb8-158"><a href="#cb8-158"></a></span>
<span id="cb8-159"><a href="#cb8-159"></a>  my.mat3<span class="ot">&lt;-</span><span class="fu">log2</span>(li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene3,this.sample3]<span class="sc">+</span><span class="dv">1</span>)</span>
<span id="cb8-160"><a href="#cb8-160"></a>  <span class="fu">rownames</span>(my.mat3)<span class="ot">&lt;-</span>gr.ensg[this.gene3]<span class="sc">$</span>gene_name</span>
<span id="cb8-161"><a href="#cb8-161"></a></span>
<span id="cb8-162"><a href="#cb8-162"></a>  col.anno3<span class="ot">&lt;-</span>mat.samples[this.sample3,<span class="fu">c</span>(<span class="st">"GA"</span>,<span class="st">"Sex"</span>),drop<span class="ot">=</span>F]</span>
<span id="cb8-163"><a href="#cb8-163"></a>  col.anno3<span class="sc">$</span><span class="st">"Predicted Sex"</span><span class="ot">&lt;-</span><span class="fu">ifelse</span>(this.sample3 <span class="sc">%in%</span> predicted.male3,<span class="st">"Male"</span>,<span class="st">"Female"</span>)</span>
<span id="cb8-164"><a href="#cb8-164"></a>  col.anno3<span class="ot">&lt;-</span>col.anno3[,<span class="fu">c</span>(<span class="st">"Predicted Sex"</span>,<span class="st">"Sex"</span>,<span class="st">"GA"</span>)]</span>
<span id="cb8-165"><a href="#cb8-165"></a></span>
<span id="cb8-166"><a href="#cb8-166"></a>  <span class="co"># via ComplexHeatmap</span></span>
<span id="cb8-167"><a href="#cb8-167"></a>  cph3<span class="ot">&lt;-</span>ComplexHeatmap<span class="sc">::</span><span class="fu">pheatmap</span>(my.mat3,</span>
<span id="cb8-168"><a href="#cb8-168"></a>                                 <span class="at">annotation_col=</span>col.anno3, </span>
<span id="cb8-169"><a href="#cb8-169"></a>                                 <span class="at">annotation_colors=</span>my.color, </span>
<span id="cb8-170"><a href="#cb8-170"></a>                                 <span class="at">cluster_cols=</span>F,</span>
<span id="cb8-171"><a href="#cb8-171"></a>                                 <span class="at">cluster_rows=</span>F,</span>
<span id="cb8-172"><a href="#cb8-172"></a>                                 <span class="at">show_colnames=</span>F,</span>
<span id="cb8-173"><a href="#cb8-173"></a>                                 <span class="at">border_color=</span><span class="cn">NA</span>,</span>
<span id="cb8-174"><a href="#cb8-174"></a>                                 <span class="at">fontsize=</span><span class="dv">12</span>,</span>
<span id="cb8-175"><a href="#cb8-175"></a>                                 <span class="at">name=</span><span class="st">"log2(count)"</span>,</span>
<span id="cb8-176"><a href="#cb8-176"></a>                                 <span class="at">row_title=</span><span class="st">"</span><span class="sc">\n</span><span class="st">Salmon (SA mode)"</span>,</span>
<span id="cb8-177"><a href="#cb8-177"></a>                                 <span class="at">row_title_gp =</span> <span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">20</span>),</span>
<span id="cb8-178"><a href="#cb8-178"></a>                                 <span class="at">heatmap_legend_param =</span> <span class="fu">list</span>(</span>
<span id="cb8-179"><a href="#cb8-179"></a>                                                            <span class="at">title_gp=</span><span class="fu">gpar</span>(<span class="at">fontsize=</span><span class="dv">13</span>,<span class="at">fontface=</span><span class="st">"bold"</span>),</span>
<span id="cb8-180"><a href="#cb8-180"></a>                                                            <span class="at">labels_gp=</span><span class="fu">gpar</span>(<span class="at">fontsize=</span><span class="dv">11</span>),</span>
<span id="cb8-181"><a href="#cb8-181"></a>                            <span class="at">annotation_legend_param =</span> <span class="fu">list</span>(<span class="at">title_gp=</span><span class="fu">gpar</span>(<span class="at">fontsize=</span><span class="dv">13</span>,<span class="at">fontface=</span><span class="st">"bold"</span>),<span class="at">labels_gp =</span> <span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">11</span>))</span>
<span id="cb8-182"><a href="#cb8-182"></a>                                )</span>
<span id="cb8-183"><a href="#cb8-183"></a>  )</span>
<span id="cb8-184"><a href="#cb8-184"></a></span>
<span id="cb8-185"><a href="#cb8-185"></a></span>
<span id="cb8-186"><a href="#cb8-186"></a>  <span class="fu">pdf</span>(<span class="at">file=</span><span class="st">"Figures/Suppl/cfRNA.Suppl.Text.Fig1C.pdf"</span>, <span class="at">width=</span><span class="dv">11</span>, <span class="at">height=</span><span class="dv">14</span>, <span class="at">title=</span><span class="st">"Abundance of chrY"</span>)</span>
<span id="cb8-187"><a href="#cb8-187"></a>  <span class="fu">draw</span>(cph1 <span class="sc">%v%</span> cph3, <span class="at">main_heatmap=</span><span class="st">"log2(count)"</span>, <span class="at">ht_gap=</span><span class="fu">unit</span>(.<span class="dv">7</span>,<span class="st">"cm"</span>), <span class="at">auto_adjust=</span>F)</span>
<span id="cb8-188"><a href="#cb8-188"></a>  <span class="fu">grid.text</span>(<span class="at">label=</span><span class="st">"C"</span>,<span class="at">gp=</span><span class="fu">gpar</span>(<span class="at">fontsize=</span><span class="dv">27</span>,<span class="at">fontface=</span><span class="st">"bold"</span>),<span class="at">x=</span><span class="fl">0.02</span>,<span class="at">y=</span>.<span class="dv">99</span>,<span class="at">just=</span><span class="st">"top"</span>)</span>
<span id="cb8-189"><a href="#cb8-189"></a>  <span class="fu">dev.off</span>()</span>
<span id="cb8-190"><a href="#cb8-190"></a><span class="in">```</span></span>
<span id="cb8-191"><a href="#cb8-191"></a></span>
<span id="cb8-192"><a href="#cb8-192"></a><span class="fu">### Supplementary Text Figure 1D {#sec-si-figure-1d}</span></span>
<span id="cb8-193"><a href="#cb8-193"></a><span class="al">![](static/figure/cfRNA.SI.FigD.png)</span>{#fig-si-fig1d}</span>
<span id="cb8-194"><a href="#cb8-194"></a></span>
<span id="cb8-195"><a href="#cb8-195"></a><span class="in">```{r }</span></span>
<span id="cb8-196"><a href="#cb8-196"></a><span class="co">#| label: si-fig-1d</span></span>
<span id="cb8-197"><a href="#cb8-197"></a><span class="co">#| eval: false</span></span>
<span id="cb8-198"><a href="#cb8-198"></a>  dl.chrY.TPM <span class="sc">%&gt;%</span> names</span>
<span id="cb8-199"><a href="#cb8-199"></a></span>
<span id="cb8-200"><a href="#cb8-200"></a>  <span class="co"># load the TPM based om 10M-ds dataset</span></span>
<span id="cb8-201"><a href="#cb8-201"></a>  <span class="fu">load</span>(<span class="st">"RData/li.TPM.10M.trim.RData"</span>)</span>
<span id="cb8-202"><a href="#cb8-202"></a>  li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"TPM"</span>]] <span class="sc">%&gt;%</span> dim <span class="co"># 59354 genes x 100 samples</span></span>
<span id="cb8-203"><a href="#cb8-203"></a></span>
<span id="cb8-204"><a href="#cb8-204"></a>  dt.trim<span class="fl">.10</span>M<span class="ot">&lt;-</span><span class="fu">data.table</span>(</span>
<span id="cb8-205"><a href="#cb8-205"></a>      <span class="st">`</span><span class="at">Type</span><span class="st">`</span><span class="ot">=</span><span class="st">"Salmon (SA mode)"</span>,</span>
<span id="cb8-206"><a href="#cb8-206"></a>      <span class="st">`</span><span class="at">SampleID</span><span class="st">`</span><span class="ot">=</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> names,</span>
<span id="cb8-207"><a href="#cb8-207"></a>      <span class="st">`</span><span class="at">Counts</span><span class="st">`</span><span class="ot">=</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums, </span>
<span id="cb8-208"><a href="#cb8-208"></a>      <span class="st">`</span><span class="at">TPM</span><span class="st">`</span><span class="ot">=</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"TPM"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums </span>
<span id="cb8-209"><a href="#cb8-209"></a>      )</span>
<span id="cb8-210"><a href="#cb8-210"></a>  <span class="co"># add Sex and GA</span></span>
<span id="cb8-211"><a href="#cb8-211"></a>  dt.trim<span class="fl">.10</span>M<span class="ot">&lt;-</span><span class="fu">merge</span>(dt.trim<span class="fl">.10</span>M, dt.samples[,.(SampleID,GA,Sex,pn_female)])</span>
<span id="cb8-212"><a href="#cb8-212"></a></span>
<span id="cb8-213"><a href="#cb8-213"></a>  dt.trim<span class="fl">.10</span>M[,<span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span><span class="sc">:</span><span class="er">=</span><span class="fu">ifelse</span>(TPM<span class="sc">&gt;</span><span class="dv">0</span>, <span class="st">"Male"</span>,<span class="st">"Female"</span>),Type]</span>
<span id="cb8-214"><a href="#cb8-214"></a>  dt.trim<span class="fl">.10</span>M<span class="sc">$</span>Sex<span class="ot">&lt;-</span><span class="fu">factor</span>(dt.trim<span class="fl">.10</span>M<span class="sc">$</span>Sex, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"Male"</span>,<span class="st">"Female"</span>))</span>
<span id="cb8-215"><a href="#cb8-215"></a>  dt.trim<span class="fl">.10</span>M<span class="sc">$</span><span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span><span class="ot">&lt;-</span><span class="fu">factor</span>(dt.trim<span class="fl">.10</span>M<span class="sc">$</span><span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span>, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"Male"</span>,<span class="st">"Female"</span>))</span>
<span id="cb8-216"><a href="#cb8-216"></a></span>
<span id="cb8-217"><a href="#cb8-217"></a>  dt.chrY.trim<span class="ot">&lt;-</span><span class="fu">rbind</span>(</span>
<span id="cb8-218"><a href="#cb8-218"></a>        dl.chrY.TPM[[<span class="st">"10M"</span>]][Type<span class="sc">==</span><span class="st">"Salmon (SA mode)"</span>][,<span class="at">Trim:=</span><span class="st">"No"</span>],</span>
<span id="cb8-219"><a href="#cb8-219"></a>        dt.trim<span class="fl">.10</span>M[,<span class="at">Trim:=</span><span class="st">"Yes"</span>]</span>
<span id="cb8-220"><a href="#cb8-220"></a>        )</span>
<span id="cb8-221"><a href="#cb8-221"></a>  <span class="do">## Number of false positives: chrY signal detected from females </span></span>
<span id="cb8-222"><a href="#cb8-222"></a>  dt.chrY.trim[,.(Trim,Sex,<span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span>)] <span class="sc">%&gt;%</span> ftable</span>
<span id="cb8-223"><a href="#cb8-223"></a></span>
<span id="cb8-224"><a href="#cb8-224"></a>  <span class="do">##</span></span>
<span id="cb8-225"><a href="#cb8-225"></a>  <span class="do">## Read counts by trim or not</span></span>
<span id="cb8-226"><a href="#cb8-226"></a>  dt.bar<span class="ot">&lt;-</span><span class="fu">merge</span>(</span>
<span id="cb8-227"><a href="#cb8-227"></a>    dl.chrY.TPM[[<span class="st">"10M"</span>]][Type<span class="sc">==</span><span class="st">"Salmon (SA mode)"</span>,.(SampleID,<span class="at">GA=</span><span class="fu">gsub</span>(<span class="st">" weeks"</span>,<span class="st">"wk"</span>,GA),Sex,Counts,TPM)], <span class="co"># no-trim</span></span>
<span id="cb8-228"><a href="#cb8-228"></a>    dt.trim<span class="fl">.10</span>M[Type<span class="sc">==</span><span class="st">"Salmon (SA mode)"</span>,.(SampleID,GA,Sex,Counts,TPM)], <span class="co"># trim</span></span>
<span id="cb8-229"><a href="#cb8-229"></a>    <span class="at">by=</span><span class="fu">c</span>(<span class="st">"SampleID"</span>,<span class="st">"Sex"</span>,<span class="st">"GA"</span>)</span>
<span id="cb8-230"><a href="#cb8-230"></a>  )</span>
<span id="cb8-231"><a href="#cb8-231"></a>  <span class="fu">cor.test</span>(<span class="sc">~</span>Counts.x <span class="sc">+</span> Counts.y, dt.bar)</span>
<span id="cb8-232"><a href="#cb8-232"></a>  dt.bar[Counts.x<span class="sc">&gt;</span>Counts.y] <span class="co"># no-trim &gt; trim</span></span>
<span id="cb8-233"><a href="#cb8-233"></a>  dt.bar[Counts.x<span class="sc">&lt;</span>Counts.y] <span class="co"># no-trim &lt; trim</span></span>
<span id="cb8-234"><a href="#cb8-234"></a>  dt.bar[Sex<span class="sc">==</span><span class="st">"Female"</span> <span class="sc">&amp;</span> Counts.x<span class="sc">&lt;</span>Counts.y] <span class="co"># more counts for Trimmed Salmon</span></span>
<span id="cb8-235"><a href="#cb8-235"></a></span>
<span id="cb8-236"><a href="#cb8-236"></a>  my.limit<span class="ot">&lt;-</span>dt.bar[,<span class="fu">max</span>(Counts.x,Counts.y)]  <span class="sc">%&gt;%</span> ceiling</span>
<span id="cb8-237"><a href="#cb8-237"></a>  p1.chrY.trim<span class="ot">&lt;-</span><span class="fu">ggplot</span>(dt.bar, <span class="fu">aes</span>(Counts.x<span class="sc">+</span><span class="dv">1</span>,Counts.y<span class="sc">+</span><span class="dv">1</span>)) <span class="sc">+</span> </span>
<span id="cb8-238"><a href="#cb8-238"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>dt.bar, <span class="at">size=</span><span class="dv">5</span>, <span class="at">position=</span><span class="fu">position_jitter</span>(<span class="at">width=</span><span class="fl">0.2</span>,<span class="at">height=</span><span class="fl">0.2</span>),<span class="at">shape=</span><span class="dv">21</span>) <span class="sc">+</span></span>
<span id="cb8-239"><a href="#cb8-239"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">trans =</span> scales<span class="sc">::</span><span class="fu">log2_trans</span>(),</span>
<span id="cb8-240"><a href="#cb8-240"></a>                     <span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">trans_breaks</span>(<span class="st">"log2"</span>,<span class="cf">function</span>(x) <span class="dv">2</span><span class="sc">^</span>x),</span>
<span id="cb8-241"><a href="#cb8-241"></a>                     <span class="at">labels =</span> scales<span class="sc">::</span><span class="fu">trans_format</span>(<span class="st">"log2"</span>, <span class="fu">math_format</span>(<span class="dv">2</span><span class="sc">^</span>.x)),</span>
<span id="cb8-242"><a href="#cb8-242"></a>                     <span class="at">limits=</span><span class="fu">c</span>(.<span class="dv">5</span>,my.limit<span class="sc">*</span><span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb8-243"><a href="#cb8-243"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">trans =</span> scales<span class="sc">::</span><span class="fu">log2_trans</span>(),</span>
<span id="cb8-244"><a href="#cb8-244"></a>                     <span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">trans_breaks</span>(<span class="st">"log2"</span>,<span class="cf">function</span>(x) <span class="dv">2</span><span class="sc">^</span>x),</span>
<span id="cb8-245"><a href="#cb8-245"></a>                     <span class="at">labels =</span> scales<span class="sc">::</span><span class="fu">trans_format</span>(<span class="st">"log2"</span>, <span class="fu">math_format</span>(<span class="dv">2</span><span class="sc">^</span>.x)),</span>
<span id="cb8-246"><a href="#cb8-246"></a>                     <span class="at">limits=</span><span class="fu">c</span>(.<span class="dv">5</span>,my.limit<span class="sc">*</span><span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb8-247"><a href="#cb8-247"></a>  <span class="co">#coord_fixed() + </span></span>
<span id="cb8-248"><a href="#cb8-248"></a>  <span class="fu">facet_grid</span>(Sex<span class="sc">~</span>GA,<span class="at">scales=</span><span class="st">"free_x"</span>) <span class="sc">+</span></span>
<span id="cb8-249"><a href="#cb8-249"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Sum of read count (No trimmed reads)"</span>, <span class="at">y=</span><span class="st">"Sum of read count (trimmed reads)"</span>) <span class="sc">+</span></span>
<span id="cb8-250"><a href="#cb8-250"></a>  <span class="fu">theme_Publication</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"black"</span>))</span>
<span id="cb8-251"><a href="#cb8-251"></a></span>
<span id="cb8-252"><a href="#cb8-252"></a>  <span class="fu">pdf</span>(<span class="at">file=</span><span class="st">"Figures/Suppl/cfRNA.Suppl.Text.Fig1D.v2.pdf"</span>, <span class="at">width=</span><span class="dv">11</span>, <span class="at">height=</span><span class="dv">7</span>, <span class="at">title=</span><span class="st">"Trim vs No trim"</span>)</span>
<span id="cb8-253"><a href="#cb8-253"></a>  cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(p1.chrY.trim, <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"D"</span>),<span class="at">label_size=</span><span class="dv">27</span>)</span>
<span id="cb8-254"><a href="#cb8-254"></a>  <span class="fu">dev.off</span>()</span>
<span id="cb8-255"><a href="#cb8-255"></a><span class="in">```</span></span>
<span id="cb8-256"><a href="#cb8-256"></a></span>
<span id="cb8-257"><a href="#cb8-257"></a><span class="fu">## Supplementary Text Figure 2 {#sec-si-figure2}</span></span>
<span id="cb8-258"><a href="#cb8-258"></a><span class="al">![](static/figure/cfRNA.SI.Fig2.png)</span>{#fig-si-fig2}</span>
<span id="cb8-259"><a href="#cb8-259"></a></span>
<span id="cb8-262"><a href="#cb8-262"></a><span class="in">```{r}</span></span>
<span id="cb8-263"><a href="#cb8-263"></a><span class="co">#| label: si-fig2</span></span>
<span id="cb8-264"><a href="#cb8-264"></a><span class="co">#| eval: false</span></span>
<span id="cb8-265"><a href="#cb8-265"></a><span class="fu">load</span>(<span class="st">"RData/dl.enet.result.core17.RData"</span>) <span class="co"># Validation of ENet models from 28wk, discovery dataset (preterm)</span></span>
<span id="cb8-266"><a href="#cb8-266"></a><span class="fu">load</span>(<span class="st">"RData/dl.combined.result.core17.RData"</span>) <span class="co"># Validation of ENet models from the combined model (preterm+term)</span></span>
<span id="cb8-267"><a href="#cb8-267"></a></span>
<span id="cb8-268"><a href="#cb8-268"></a><span class="do">##</span></span>
<span id="cb8-269"><a href="#cb8-269"></a><span class="do">## AUC in All </span></span>
<span id="cb8-270"><a href="#cb8-270"></a><span class="do">##</span></span>
<span id="cb8-271"><a href="#cb8-271"></a>ga<span class="ot">&lt;-</span><span class="fu">paste0</span>(<span class="fu">c</span>(<span class="dv">12</span>,<span class="dv">20</span>,<span class="dv">28</span>,<span class="dv">36</span>), <span class="st">"wk"</span>)</span>
<span id="cb8-272"><a href="#cb8-272"></a>term<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"(preterm)"</span>,<span class="st">"(term)"</span>)</span>
<span id="cb8-273"><a href="#cb8-273"></a>my.dataset<span class="ot">&lt;-</span><span class="fu">data.table</span>(<span class="fu">expand.grid</span>(ga,term))[,Var3<span class="sc">:</span><span class="er">=</span><span class="fu">paste0</span>(Var1,Var2)][Var3<span class="sc">!=</span><span class="st">"36wk(preterm)"</span>]<span class="sc">$</span>Var3</span>
<span id="cb8-274"><a href="#cb8-274"></a>my.dataset<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"Munchel"</span>,my.dataset)</span>
<span id="cb8-275"><a href="#cb8-275"></a></span>
<span id="cb8-276"><a href="#cb8-276"></a>dt.auc.long<span class="ot">&lt;-</span></span>
<span id="cb8-277"><a href="#cb8-277"></a><span class="fu">lapply</span>(my.dataset, <span class="cf">function</span>(this.dataset){</span>
<span id="cb8-278"><a href="#cb8-278"></a>  <span class="fu">rbind</span>(</span>
<span id="cb8-279"><a href="#cb8-279"></a>        (<span class="fu">lapply</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>, <span class="cf">function</span>(my.num) dl.combined.result[[my.num<span class="dv">-1</span>]][fold<span class="sc">==</span>this.dataset,.(<span class="at">Num=</span>my.num,<span class="at">Predictor=</span>predictor,<span class="at">AUC_test=</span>AUC_test<span class="sc">/</span><span class="dv">100</span>,<span class="at">AUC_test_lo=</span>AUC_test_lo<span class="sc">/</span><span class="dv">100</span>,<span class="at">AUC_test_hi=</span>AUC_test_hi<span class="sc">/</span><span class="dv">100</span>)]) <span class="sc">%&gt;%</span> rbindlist)[,<span class="at">Model:=</span><span class="st">"preterm+term (28wk)"</span>]</span>
<span id="cb8-280"><a href="#cb8-280"></a>        ,</span>
<span id="cb8-281"><a href="#cb8-281"></a>        (<span class="fu">lapply</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>, <span class="cf">function</span>(my.num) dl.enet.result[[my.num<span class="dv">-1</span>]][fold<span class="sc">==</span>this.dataset <span class="sc">&amp;</span> <span class="fu">grepl</span>(<span class="st">"ENet"</span>,methods),.(<span class="at">Num=</span>my.num,<span class="at">Predictor=</span>predictor,<span class="at">AUC_test=</span>AUC_test<span class="sc">/</span><span class="dv">100</span>,<span class="at">AUC_test_lo=</span>AUC_test_lo<span class="sc">/</span><span class="dv">100</span>,<span class="at">AUC_test_hi=</span>AUC_test_hi<span class="sc">/</span><span class="dv">100</span>)]) <span class="sc">%&gt;%</span> rbindlist)[,<span class="at">Model:=</span><span class="st">"preterm (28wk)"</span>]</span>
<span id="cb8-282"><a href="#cb8-282"></a>        </span>
<span id="cb8-283"><a href="#cb8-283"></a>  )[,Dataset<span class="sc">:</span><span class="er">=</span>this.dataset] </span>
<span id="cb8-284"><a href="#cb8-284"></a>}) <span class="sc">%&gt;%</span> rbindlist <span class="sc">%&gt;%</span> <span class="fu">setcolorder</span>(<span class="fu">c</span>(<span class="st">"Dataset"</span>,<span class="st">"Model"</span>))</span>
<span id="cb8-285"><a href="#cb8-285"></a></span>
<span id="cb8-286"><a href="#cb8-286"></a>dt.auc.long[Dataset<span class="sc">!=</span><span class="st">"Munchel"</span>,GA<span class="sc">:</span><span class="er">=</span><span class="fu">substr</span>(Dataset,<span class="dv">1</span>,<span class="dv">4</span>)]</span>
<span id="cb8-287"><a href="#cb8-287"></a>dt.auc.long[Dataset<span class="sc">!=</span><span class="st">"Munchel"</span>,Source<span class="sc">:</span><span class="er">=</span><span class="fu">substr</span>(Dataset,<span class="dv">6</span>,<span class="fu">nchar</span>(Dataset)<span class="sc">-</span><span class="dv">1</span>)][,Source<span class="sc">:</span><span class="er">=</span><span class="fu">ifelse</span>(Source<span class="sc">==</span><span class="st">"preterm"</span>,<span class="st">"Discovery (preterm)"</span>,<span class="st">"Validation (term)"</span>)]</span>
<span id="cb8-288"><a href="#cb8-288"></a>dt.auc.long[Dataset<span class="sc">==</span><span class="st">"Munchel"</span>,<span class="st">`</span><span class="at">:=</span><span class="st">`</span>(<span class="at">Source=</span><span class="st">"Munchel"</span>,<span class="at">GA=</span><span class="st">"36wk"</span>)]</span>
<span id="cb8-289"><a href="#cb8-289"></a>dt.auc.long<span class="sc">$</span>Model<span class="ot">&lt;-</span><span class="fu">factor</span>(dt.auc.long<span class="sc">$</span>Model, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"preterm (28wk)"</span>,<span class="st">"preterm+term (28wk)"</span>))</span>
<span id="cb8-290"><a href="#cb8-290"></a>dt.auc.long<span class="sc">$</span>Source<span class="ot">&lt;-</span><span class="fu">factor</span>(dt.auc.long<span class="sc">$</span>Source, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"Discovery (preterm)"</span>,<span class="st">"Validation (term)"</span>,<span class="st">"Munchel"</span>))</span>
<span id="cb8-291"><a href="#cb8-291"></a></span>
<span id="cb8-292"><a href="#cb8-292"></a>p.auc.combined.A<span class="ot">&lt;-</span><span class="fu">ggplot</span>(dt.auc.long[Source<span class="sc">!=</span><span class="st">"Munchel"</span>], <span class="fu">aes</span>(Num, AUC_test,<span class="at">group=</span>Model)) <span class="sc">+</span> </span>
<span id="cb8-293"><a href="#cb8-293"></a>  <span class="fu">geom_rect</span>(<span class="at">data =</span> dt.auc.long[Source<span class="sc">!=</span><span class="st">"Munchel"</span> <span class="sc">&amp;</span> GA<span class="sc">==</span><span class="st">"28wk"</span>],<span class="fu">aes</span>(<span class="at">fill =</span> GA),<span class="at">fill=</span><span class="st">"grey90"</span>,<span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">ymax =</span> <span class="cn">Inf</span>,<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb8-294"><a href="#cb8-294"></a>  <span class="fu">geom_pointrange</span>(<span class="fu">aes</span>(<span class="at">col=</span>Model,<span class="at">ymin=</span>AUC_test_lo, <span class="at">ymax=</span>AUC_test_hi),</span>
<span id="cb8-295"><a href="#cb8-295"></a>                  <span class="at">position=</span><span class="fu">position_dodge</span>(<span class="at">width=</span><span class="fl">0.4</span>),</span>
<span id="cb8-296"><a href="#cb8-296"></a>                  <span class="at">size=</span>.<span class="dv">8</span>,<span class="at">alpha=</span>.<span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb8-297"><a href="#cb8-297"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>),<span class="at">breaks=</span><span class="fu">c</span>(<span class="dv">0</span>,.<span class="dv">2</span>,.<span class="dv">4</span>,.<span class="dv">6</span>,.<span class="dv">8</span>,<span class="dv">1</span>), <span class="at">limit=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">1.05</span>)) <span class="sc">+</span></span>
<span id="cb8-298"><a href="#cb8-298"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks=</span><span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb8-299"><a href="#cb8-299"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">"AUC"</span>, <span class="at">x=</span><span class="st">"Number of predictor mRNA in the training model"</span>) <span class="sc">+</span></span>
<span id="cb8-300"><a href="#cb8-300"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"grey30"</span>,cbPalette2[<span class="dv">2</span>])) <span class="sc">+</span></span>
<span id="cb8-301"><a href="#cb8-301"></a>  <span class="fu">facet_grid</span>(Source<span class="sc">~</span>GA) <span class="sc">+</span> </span>
<span id="cb8-302"><a href="#cb8-302"></a>  <span class="fu">theme_Publication</span>() <span class="sc">+</span></span>
<span id="cb8-303"><a href="#cb8-303"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">""</span>, <span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"black"</span>))</span>
<span id="cb8-304"><a href="#cb8-304"></a></span>
<span id="cb8-305"><a href="#cb8-305"></a>p.auc.combined.B<span class="ot">&lt;-</span><span class="fu">ggplot</span>(dt.auc.long[Source<span class="sc">==</span><span class="st">"Munchel"</span>], <span class="fu">aes</span>(Num, AUC_test,<span class="at">group=</span>Model)) <span class="sc">+</span> </span>
<span id="cb8-306"><a href="#cb8-306"></a>  <span class="fu">geom_pointrange</span>(<span class="fu">aes</span>(<span class="at">col=</span>Model,<span class="at">ymin=</span>AUC_test_lo, <span class="at">ymax=</span>AUC_test_hi),</span>
<span id="cb8-307"><a href="#cb8-307"></a>                  <span class="at">position=</span><span class="fu">position_dodge</span>(<span class="at">width=</span><span class="fl">0.4</span>),</span>
<span id="cb8-308"><a href="#cb8-308"></a>                  <span class="at">size=</span>.<span class="dv">8</span>,<span class="at">alpha=</span>.<span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb8-309"><a href="#cb8-309"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>),<span class="at">breaks=</span><span class="fu">c</span>(<span class="dv">0</span>,.<span class="dv">2</span>,.<span class="dv">4</span>,.<span class="dv">6</span>,.<span class="dv">8</span>,<span class="dv">1</span>), <span class="at">limit=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">1.05</span>)) <span class="sc">+</span></span>
<span id="cb8-310"><a href="#cb8-310"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks=</span><span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb8-311"><a href="#cb8-311"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">"AUC"</span>, <span class="at">x=</span><span class="st">"Number of predictor mRNA in the training model                                     "</span>) <span class="sc">+</span></span>
<span id="cb8-312"><a href="#cb8-312"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"grey30"</span>,cbPalette2[<span class="dv">2</span>])) <span class="sc">+</span></span>
<span id="cb8-313"><a href="#cb8-313"></a>  <span class="fu">facet_grid</span>(Source<span class="sc">~</span>.) <span class="sc">+</span> </span>
<span id="cb8-314"><a href="#cb8-314"></a>  <span class="fu">theme_Publication</span>() <span class="sc">+</span></span>
<span id="cb8-315"><a href="#cb8-315"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"left"</span>, </span>
<span id="cb8-316"><a href="#cb8-316"></a>        <span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"black"</span>),</span>
<span id="cb8-317"><a href="#cb8-317"></a>        <span class="at">plot.margin=</span><span class="fu">margin</span>(<span class="dv">1</span>,<span class="sc">-</span><span class="dv">15</span>,<span class="dv">7</span>,<span class="dv">0</span>)</span>
<span id="cb8-318"><a href="#cb8-318"></a>  )</span>
<span id="cb8-319"><a href="#cb8-319"></a></span>
<span id="cb8-320"><a href="#cb8-320"></a>p.auc.bottom<span class="ot">&lt;-</span>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(<span class="cn">NULL</span>,p.auc.combined.B, <span class="at">nrow=</span><span class="dv">1</span>,<span class="at">labels=</span><span class="fu">c</span>(<span class="st">""</span>,<span class="st">"B"</span>),<span class="at">label_size=</span><span class="dv">27</span>,<span class="at">rel_widths=</span><span class="fu">c</span>(.<span class="dv">87</span>,<span class="dv">1</span>))</span>
<span id="cb8-321"><a href="#cb8-321"></a></span>
<span id="cb8-322"><a href="#cb8-322"></a><span class="fu">pdf</span>(<span class="at">file=</span><span class="st">"Figures/Suppl/cfRNA.Suppl.Data.Fig.combined.AUC.AB.pdf"</span>, <span class="at">width=</span><span class="dv">12</span>, <span class="at">height=</span><span class="dv">10</span>, <span class="at">title=</span><span class="st">"Suppl Fig: AUC from the combined model vs preterm model"</span>)</span>
<span id="cb8-323"><a href="#cb8-323"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(p.auc.combined.A, p.auc.bottom,<span class="at">ncol=</span><span class="dv">1</span>, <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"A"</span>,<span class="st">""</span>),<span class="at">label_size=</span><span class="dv">27</span>,<span class="at">rel_heights=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">1</span>), <span class="at">align=</span><span class="st">"v"</span>, <span class="at">axis=</span><span class="st">"lr"</span>)</span>
<span id="cb8-324"><a href="#cb8-324"></a><span class="fu">dev.off</span>()</span>
<span id="cb8-325"><a href="#cb8-325"></a><span class="in">```</span></span>
<span id="cb8-326"><a href="#cb8-326"></a></span>
<span id="cb8-327"><a href="#cb8-327"></a><span class="fu">## Supplementary Text Figure 3 {#sec-si-figure3}</span></span>
<span id="cb8-328"><a href="#cb8-328"></a><span class="al">![](static/figure/cfRNA.SI.Fig3.png)</span>{#fig-si-fig3}</span>
<span id="cb8-329"><a href="#cb8-329"></a></span>
<span id="cb8-332"><a href="#cb8-332"></a><span class="in">```{r}</span></span>
<span id="cb8-333"><a href="#cb8-333"></a><span class="co">#| label: si-fig3</span></span>
<span id="cb8-334"><a href="#cb8-334"></a><span class="co">#| eval: false</span></span>
<span id="cb8-335"><a href="#cb8-335"></a><span class="fu">load</span>(<span class="st">"RData/dt.cpmZ.preterm.POPS-2022.GRCh38.88.RData"</span>) <span class="co"># dt.cpmZ (preterm)</span></span>
<span id="cb8-336"><a href="#cb8-336"></a><span class="fu">load</span>(<span class="st">"RData/dt.cpmZ.term.POPS-2022.GRCh38.88.RData"</span>) <span class="co"># dt.cpmZ.term (term)</span></span>
<span id="cb8-337"><a href="#cb8-337"></a><span class="fu">load</span>(<span class="st">"RData/dt.cpmZ.munchel.RData"</span>) <span class="co"># dt.cpmZ.munchel (Munchel)</span></span>
<span id="cb8-338"><a href="#cb8-338"></a></span>
<span id="cb8-339"><a href="#cb8-339"></a>dt.lep.pappa2<span class="ot">&lt;-</span><span class="fu">rbind</span>(</span>
<span id="cb8-340"><a href="#cb8-340"></a>                      (dt.cpmZ[geneName <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"LEP"</span>,<span class="st">"PAPPA2"</span>)] <span class="sc">%&gt;%</span> <span class="fu">dcast.data.table</span>(SampleID<span class="sc">+</span>GA<span class="sc">+</span>Condition<span class="sc">~</span>geneName, <span class="at">value.var=</span><span class="st">"logCPMZ"</span>))[,<span class="at">Source:=</span><span class="st">"Discovery"</span>],</span>
<span id="cb8-341"><a href="#cb8-341"></a>                      (dt.cpmZ.term[geneName <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"LEP"</span>,<span class="st">"PAPPA2"</span>)] <span class="sc">%&gt;%</span> <span class="fu">dcast.data.table</span>(SampleID<span class="sc">+</span>GA<span class="sc">+</span>Condition<span class="sc">~</span>geneName, <span class="at">value.var=</span><span class="st">"logCPMZ"</span>))[,<span class="at">Source:=</span><span class="st">"Validation"</span>],</span>
<span id="cb8-342"><a href="#cb8-342"></a>                      (dt.cpmZ.munchel[geneName <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"LEP"</span>,<span class="st">"PAPPA2"</span>)] <span class="sc">%&gt;%</span> <span class="fu">dcast.data.table</span>(SampleID<span class="sc">+</span>Condition<span class="sc">~</span>geneName, <span class="at">value.var=</span><span class="st">"logCPMZ"</span>))[,<span class="at">Source:=</span><span class="st">"Munchel"</span>]</span>
<span id="cb8-343"><a href="#cb8-343"></a>    ,<span class="at">fill=</span>T)</span>
<span id="cb8-344"><a href="#cb8-344"></a></span>
<span id="cb8-345"><a href="#cb8-345"></a>dt.lep.pappa2[,Dataset<span class="sc">:</span><span class="er">=</span><span class="fu">ifelse</span>(<span class="fu">is.na</span>(GA),Source,<span class="fu">paste0</span>(Source,<span class="st">"("</span>,GA,<span class="st">")"</span>))]</span>
<span id="cb8-346"><a href="#cb8-346"></a>dt.lep.pappa2[,.(<span class="at">Cor=</span><span class="fu">cor</span>(LEP,PAPPA2),<span class="at">Pval=</span><span class="fu">cor.test</span>(LEP,PAPPA2)<span class="sc">$</span>p.value),.(Source,GA)][<span class="fu">order</span>(Source,GA)] </span>
<span id="cb8-347"><a href="#cb8-347"></a>dt.lep.pappa2[Condition<span class="sc">==</span><span class="st">"Case"</span>,.(<span class="at">Cor=</span><span class="fu">cor</span>(LEP,PAPPA2),<span class="at">Pval=</span><span class="fu">cor.test</span>(LEP,PAPPA2)<span class="sc">$</span>p.value),.(Source,GA)][<span class="fu">order</span>(Source,GA)] </span>
<span id="cb8-348"><a href="#cb8-348"></a></span>
<span id="cb8-349"><a href="#cb8-349"></a>p1<span class="ot">&lt;-</span></span>
<span id="cb8-350"><a href="#cb8-350"></a><span class="fu">ggplot</span>(dt.lep.pappa2[Source<span class="sc">!=</span><span class="st">"Munchel"</span>], <span class="fu">aes</span>(LEP,PAPPA2)) <span class="sc">+</span></span>
<span id="cb8-351"><a href="#cb8-351"></a>  <span class="fu">geom_rect</span>(<span class="at">data =</span> dt.lep.pappa2[Source<span class="sc">==</span><span class="st">"Discovery"</span> <span class="sc">&amp;</span> GA<span class="sc">==</span><span class="st">"28wk"</span>],<span class="fu">aes</span>(<span class="at">fill =</span> GA),<span class="at">fill=</span><span class="st">"grey80"</span>,<span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">ymax =</span> <span class="cn">Inf</span>,<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb8-352"><a href="#cb8-352"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">col=</span>Condition),<span class="at">size=</span><span class="dv">3</span>,<span class="at">shape=</span><span class="dv">21</span>,<span class="at">stroke=</span><span class="fl">1.1</span>,<span class="at">alpha=</span>.<span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb8-353"><a href="#cb8-353"></a>  <span class="fu">facet_grid</span>(Source<span class="sc">~</span>GA) <span class="sc">+</span> </span>
<span id="cb8-354"><a href="#cb8-354"></a>  ggsci<span class="sc">::</span><span class="fu">scale_color_jama</span>() <span class="sc">+</span></span>
<span id="cb8-355"><a href="#cb8-355"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Z-score (LEP)"</span>,<span class="at">y=</span><span class="st">"Z-score (PAPPA2)"</span>) <span class="sc">+</span></span>
<span id="cb8-356"><a href="#cb8-356"></a>  <span class="fu">theme_Publication</span>() <span class="sc">+</span><span class="fu">theme</span> (<span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"black"</span>),<span class="at">legend.position=</span><span class="st">"top"</span>)</span>
<span id="cb8-357"><a href="#cb8-357"></a></span>
<span id="cb8-358"><a href="#cb8-358"></a>p2<span class="ot">&lt;-</span></span>
<span id="cb8-359"><a href="#cb8-359"></a><span class="fu">ggplot</span>(dt.lep.pappa2[Source<span class="sc">==</span><span class="st">"Munchel"</span>], <span class="fu">aes</span>(LEP,PAPPA2)) <span class="sc">+</span></span>
<span id="cb8-360"><a href="#cb8-360"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">col=</span>Condition),<span class="at">size=</span><span class="dv">3</span>,<span class="at">shape=</span><span class="dv">21</span>,<span class="at">stroke=</span><span class="fl">1.1</span>,<span class="at">alpha=</span>.<span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb8-361"><a href="#cb8-361"></a>  <span class="fu">facet_grid</span>(Source<span class="sc">~</span>.) <span class="sc">+</span> </span>
<span id="cb8-362"><a href="#cb8-362"></a>  ggsci<span class="sc">::</span><span class="fu">scale_color_jama</span>() <span class="sc">+</span></span>
<span id="cb8-363"><a href="#cb8-363"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">""</span>,<span class="at">y=</span><span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb8-364"><a href="#cb8-364"></a>  <span class="fu">theme_Publication</span>() <span class="sc">+</span><span class="fu">theme</span> (<span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"black"</span>),<span class="at">legend.position=</span><span class="st">""</span>)</span>
<span id="cb8-365"><a href="#cb8-365"></a></span>
<span id="cb8-366"><a href="#cb8-366"></a>p2.right<span class="ot">&lt;-</span>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(<span class="cn">NULL</span>,p2,<span class="at">labels=</span><span class="fu">c</span>(<span class="st">""</span>,<span class="st">"B"</span>),<span class="at">ncol=</span><span class="dv">1</span>,<span class="at">label_size=</span><span class="dv">27</span>,<span class="at">rel_heights=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb8-367"><a href="#cb8-367"></a>p.corr2<span class="ot">&lt;-</span>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(p1, p2.right, <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"A"</span>,<span class="st">""</span>),<span class="at">label_size=</span><span class="dv">27</span>,<span class="at">rel_widths=</span><span class="fu">c</span>(<span class="fl">2.8</span>,<span class="dv">1</span>))</span>
<span id="cb8-368"><a href="#cb8-368"></a></span>
<span id="cb8-369"><a href="#cb8-369"></a><span class="fu">cairo_pdf</span>(<span class="at">file=</span><span class="st">"Figures/Suppl/cfRNA.Suppl.LEP.PAPPA2.cor2.pdf"</span>, <span class="at">width=</span><span class="dv">12</span>,<span class="at">height=</span><span class="dv">6</span>, <span class="at">onefile=</span>T)</span>
<span id="cb8-370"><a href="#cb8-370"></a>p.corr2</span>
<span id="cb8-371"><a href="#cb8-371"></a><span class="fu">dev.off</span>()</span>
<span id="cb8-372"><a href="#cb8-372"></a><span class="in">```</span></span>
<span id="cb8-373"><a href="#cb8-373"></a></span>
<span id="cb8-374"><a href="#cb8-374"></a><span class="fu">## Supplementary Text Figure 4, 5, and 6 {#sec-si-figure4-6}</span></span>
<span id="cb8-375"><a href="#cb8-375"></a><span class="al">![](static/figure/cfRNA.SI.Fig4.5.6.pdf#zoom=50&amp;&amp;navpanes=0&amp;statusbar=0&amp;messages=0&amp;pagemode=none)</span>{width=100% height=475px}</span>
<span id="cb8-376"><a href="#cb8-376"></a><span class="co">[</span><span class="ot">Full Screen</span><span class="co">](static/figure/cfRNA.SI.Fig4.5.6.pdf)</span></span>
<span id="cb8-377"><a href="#cb8-377"></a></span>
<span id="cb8-378"><a href="#cb8-378"></a>You may need to refer to @sec-ga-long and @sec-fig5b</span>
<span id="cb8-379"><a href="#cb8-379"></a></span>
<span id="cb8-380"><a href="#cb8-380"></a><span class="in">```{r rlmer_csh1_plot}</span></span>
<span id="cb8-381"><a href="#cb8-381"></a><span class="co">#| label: si-fig456</span></span>
<span id="cb8-382"><a href="#cb8-382"></a><span class="co">#| eval: false</span></span>
<span id="cb8-383"><a href="#cb8-383"></a></span>
<span id="cb8-384"><a href="#cb8-384"></a>targets<span class="ot">&lt;-</span><span class="fu">c</span>(</span>
<span id="cb8-385"><a href="#cb8-385"></a>  dt.R2.GA2[<span class="fu">order</span>(<span class="sc">-</span>r2m)][[<span class="st">"gene"</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>], <span class="co"># top 100 by r2</span></span>
<span id="cb8-386"><a href="#cb8-386"></a>  dt.R2.GA2[<span class="fu">order</span>(p.combined)][[<span class="st">"gene"</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>], <span class="co"># top 100 by the combined p-value</span></span>
<span id="cb8-387"><a href="#cb8-387"></a>  dt.R2.GA[BH.GA<span class="sc">&lt;</span><span class="fl">0.05</span>][<span class="fu">order</span>(p.GA)][coef.GA<span class="sc">&lt;</span><span class="dv">0</span>][[<span class="st">"gene"</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>] <span class="co"># top 100 decreasing</span></span>
<span id="cb8-388"><a href="#cb8-388"></a>  ) <span class="sc">%&gt;%</span> unique</span>
<span id="cb8-389"><a href="#cb8-389"></a><span class="co">#</span></span>
<span id="cb8-390"><a href="#cb8-390"></a>li.plots<span class="ot">&lt;-</span><span class="fu">lapply</span>(dl.logCPM[targets], <span class="cf">function</span>(dt.foo){</span>
<span id="cb8-391"><a href="#cb8-391"></a>    my.gene<span class="ot">&lt;-</span><span class="fu">unique</span>(dt.foo[[<span class="st">"geneName"</span>]])</span>
<span id="cb8-392"><a href="#cb8-392"></a>    <span class="fu">message</span>(my.gene)</span>
<span id="cb8-393"><a href="#cb8-393"></a></span>
<span id="cb8-394"><a href="#cb8-394"></a>    r2m<span class="ot">&lt;-</span>li.lmer[[my.gene]][[<span class="st">"stat"</span>]][Model<span class="sc">==</span><span class="st">"GA2"</span>][[<span class="st">"r2m"</span>]] <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="dv">3</span>)</span>
<span id="cb8-395"><a href="#cb8-395"></a>    r2c<span class="ot">&lt;-</span>li.lmer[[my.gene]][[<span class="st">"stat"</span>]][Model<span class="sc">==</span><span class="st">"GA2"</span>][[<span class="st">"r2c"</span>]] <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="dv">3</span>)</span>
<span id="cb8-396"><a href="#cb8-396"></a>    <span class="co">#my.title=paste0(my.gene," (r2m:",r2m,", r2c:",r2c,")")</span></span>
<span id="cb8-397"><a href="#cb8-397"></a>    my.title<span class="ot">=</span><span class="fu">paste0</span>(my.gene,<span class="st">" (r2m:"</span>,r2m,<span class="st">")"</span>)</span>
<span id="cb8-398"><a href="#cb8-398"></a></span>
<span id="cb8-399"><a href="#cb8-399"></a>    my.title.y<span class="ot">&lt;-</span>latex2exp<span class="sc">::</span><span class="fu">TeX</span>(r<span class="st">"($</span><span class="sc">\t</span><span class="st">extbf{log_2 CPM}$)"</span>)</span>
<span id="cb8-400"><a href="#cb8-400"></a></span>
<span id="cb8-401"><a href="#cb8-401"></a>    <span class="fu">ggplot</span>(dt.foo, <span class="fu">aes</span>(<span class="at">x =</span> GAwk, <span class="at">y =</span> logCPM)) <span class="sc">+</span></span>
<span id="cb8-402"><a href="#cb8-402"></a>    <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">group =</span> POPSID), <span class="at">linewidth =</span> .<span class="dv">5</span>, <span class="at">col=</span><span class="st">"grey80"</span>,<span class="at">alpha=</span>.<span class="dv">7</span>) <span class="sc">+</span></span>
<span id="cb8-403"><a href="#cb8-403"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> <span class="fu">as.factor</span>(GA)), <span class="at">size=</span><span class="dv">2</span>, <span class="at">alpha=</span>.<span class="dv">7</span>) <span class="sc">+</span></span>
<span id="cb8-404"><a href="#cb8-404"></a>    <span class="fu">labs</span>(<span class="at">subtitle=</span>my.title, <span class="at">x =</span> <span class="st">"Gestational Age (weeks)"</span>, <span class="at">y=</span>my.title.y, <span class="at">color =</span> <span class="st">"GA (week)"</span>) <span class="sc">+</span></span>
<span id="cb8-405"><a href="#cb8-405"></a>    <span class="fu">geom_ribbon</span>(<span class="at">data=</span>li.lmer[[my.gene]][[<span class="st">"predict"</span>]], <span class="fu">aes</span>(<span class="at">ymin=</span>conf.low, <span class="at">ymax=</span>conf.high), <span class="at">alpha=</span><span class="fl">0.6</span>, <span class="at">fill =</span> <span class="st">"grey40"</span>) <span class="sc">+</span></span>
<span id="cb8-406"><a href="#cb8-406"></a>    <span class="fu">geom_line</span>(<span class="at">data=</span>li.lmer[[my.gene]][[<span class="st">"predict"</span>]], <span class="at">linewidth=</span><span class="fl">1.5</span>, <span class="at">col=</span><span class="st">"blue"</span>,<span class="at">alpha=</span><span class="fl">0.8</span>) <span class="sc">+</span> </span>
<span id="cb8-407"><a href="#cb8-407"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">breaks=</span><span class="fu">c</span>(<span class="dv">12</span>,<span class="dv">20</span>,<span class="dv">28</span>,<span class="dv">36</span>)) <span class="sc">+</span></span>
<span id="cb8-408"><a href="#cb8-408"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values=</span>cbPalette2) <span class="sc">+</span></span>
<span id="cb8-409"><a href="#cb8-409"></a>    <span class="fu">theme_Publication</span>() <span class="sc">+</span></span>
<span id="cb8-410"><a href="#cb8-410"></a>    <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>)</span>
<span id="cb8-411"><a href="#cb8-411"></a>})</span>
<span id="cb8-412"><a href="#cb8-412"></a></span>
<span id="cb8-413"><a href="#cb8-413"></a><span class="co"># Top 100 by R2M</span></span>
<span id="cb8-414"><a href="#cb8-414"></a>file.name1<span class="ot">&lt;-</span><span class="fu">file.path</span>(<span class="st">'Figures/Suppl'</span>,<span class="fu">paste0</span>(<span class="st">'cfRNA.Suppl.Data.Fig.logCPM_by_GA_lmer.pdf'</span>))</span>
<span id="cb8-415"><a href="#cb8-415"></a><span class="fu">cairo_pdf</span>(<span class="at">filename=</span>file.name1, <span class="at">width=</span><span class="dv">14</span>, <span class="at">height=</span><span class="dv">9</span>, <span class="at">onefile=</span>T)</span>
<span id="cb8-416"><a href="#cb8-416"></a><span class="fu">grid.text</span>(<span class="at">label=</span><span class="fu">paste0</span>(<span class="st">"Supplementary Figure 4.</span><span class="sc">\n</span><span class="st">Top 100 cfRNAs</span><span class="sc">\n</span><span class="st">by the coefficient of determination"</span>),<span class="at">gp=</span><span class="fu">gpar</span>(<span class="at">fontsize=</span><span class="dv">50</span>,<span class="at">fontface=</span><span class="st">"bold"</span>),<span class="at">x=</span>.<span class="dv">5</span>,<span class="at">y=</span>.<span class="dv">6</span>,<span class="at">just=</span><span class="st">"center"</span>)</span>
<span id="cb8-417"><a href="#cb8-417"></a>gridExtra<span class="sc">::</span><span class="fu">marrangeGrob</span>(li.plots[dt.R2.GA2[<span class="fu">order</span>(<span class="sc">-</span>r2m)][[<span class="st">"gene"</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>]],<span class="at">nrow=</span><span class="dv">2</span>,<span class="at">ncol=</span><span class="dv">4</span>,<span class="at">top=</span><span class="cn">NULL</span>)</span>
<span id="cb8-418"><a href="#cb8-418"></a></span>
<span id="cb8-419"><a href="#cb8-419"></a><span class="co"># Top 100 by p.combined</span></span>
<span id="cb8-420"><a href="#cb8-420"></a><span class="fu">grid.newpage</span>()</span>
<span id="cb8-421"><a href="#cb8-421"></a><span class="fu">grid.text</span>(<span class="at">label=</span><span class="fu">paste0</span>(<span class="st">"Supplementary Figure 5.</span><span class="sc">\n</span><span class="st">Top 100 cfRNAs</span><span class="sc">\n</span><span class="st">by the combined p-values"</span>),<span class="at">gp=</span><span class="fu">gpar</span>(<span class="at">fontsize=</span><span class="dv">50</span>,<span class="at">fontface=</span><span class="st">"bold"</span>),<span class="at">x=</span>.<span class="dv">5</span>,<span class="at">y=</span>.<span class="dv">6</span>,<span class="at">just=</span><span class="st">"center"</span>)</span>
<span id="cb8-422"><a href="#cb8-422"></a>gridExtra<span class="sc">::</span><span class="fu">marrangeGrob</span>(li.plots[dt.R2.GA2[<span class="fu">order</span>(p.combined)][[<span class="st">"gene"</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>]],<span class="at">nrow=</span><span class="dv">2</span>,<span class="at">ncol=</span><span class="dv">4</span>,<span class="at">top=</span><span class="cn">NULL</span>)</span>
<span id="cb8-423"><a href="#cb8-423"></a></span>
<span id="cb8-424"><a href="#cb8-424"></a><span class="co"># Top 100 by negative coefficient in linear mixed effect model (GA only)</span></span>
<span id="cb8-425"><a href="#cb8-425"></a><span class="fu">grid.newpage</span>()</span>
<span id="cb8-426"><a href="#cb8-426"></a><span class="fu">grid.text</span>(<span class="at">label=</span><span class="fu">paste0</span>(<span class="st">"Supplementary Figure 6.</span><span class="sc">\n</span><span class="st">Top 100 cfRNAs</span><span class="sc">\n</span><span class="st">significantly decreasing</span><span class="sc">\n</span><span class="st">(negative coefficient)</span><span class="sc">\n</span><span class="st">by gestational age"</span>),<span class="at">gp=</span><span class="fu">gpar</span>(<span class="at">fontsize=</span><span class="dv">50</span>,<span class="at">fontface=</span><span class="st">"bold"</span>),<span class="at">x=</span>.<span class="dv">5</span>,<span class="at">y=</span>.<span class="dv">6</span>,<span class="at">just=</span><span class="st">"center"</span>)</span>
<span id="cb8-427"><a href="#cb8-427"></a>gridExtra<span class="sc">::</span><span class="fu">marrangeGrob</span>(li.plots[dt.R2.GA[BH.GA<span class="sc">&lt;</span><span class="fl">0.05</span>][<span class="fu">order</span>(p.GA)][coef.GA<span class="sc">&lt;</span><span class="dv">0</span>][[<span class="st">"gene"</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>]],<span class="at">nrow=</span><span class="dv">2</span>,<span class="at">ncol=</span><span class="dv">4</span>,<span class="at">top=</span><span class="cn">NULL</span>)</span>
<span id="cb8-428"><a href="#cb8-428"></a><span class="fu">dev.off</span>()</span>
<span id="cb8-429"><a href="#cb8-429"></a></span>
<span id="cb8-430"><a href="#cb8-430"></a><span class="in">```</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>Cookbook for the analysis of cell-free RNA, © <a href="https://sung.github.io">Sungsam Gong</a> (2024-2025)</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/obsgynaecam/cell-free-rna-2024/edit/main/suppl_figure.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/obsgynaecam/cell-free-rna-2024/blob/main/suppl_figure.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li><li><a href="https://github.com/obsgynaecam/cell-free-rna-2024/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
<p>This book was built with <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>




</body></html>