<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.551">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Cell-free-rna-2024 - 7&nbsp; Supplementary Text Figures</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./main_figure.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./figure.html">Figures</a></li><li class="breadcrumb-item"><a href="./suppl_figure.html"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Supplementary Text Figures</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Cell-free-rna-2024</a> 
        <div class="sidebar-tools-main">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Methods</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./sample.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Sample Information</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./quant.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">RNA-seq data processing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cross_validation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Cross Validation</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Supplementary Text</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./SI_select_quant_method.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Selection of RNA-seq quantification method</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./SI_trimming.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Assessing the effect of trimming the adaptor sequences</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./figure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figures</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./main_figure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Main Figures</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./suppl_figure.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Supplementary Text Figures</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#sec-si-figure-a" id="toc-sec-si-figure-a" class="nav-link active" data-scroll-target="#sec-si-figure-a"><span class="header-section-number">7.1</span> Supplementary Text Figure A</a></li>
  <li><a href="#sec-si-figure-b" id="toc-sec-si-figure-b" class="nav-link" data-scroll-target="#sec-si-figure-b"><span class="header-section-number">7.2</span> Supplementary Text Figure B</a></li>
  <li><a href="#sec-si-figure-c" id="toc-sec-si-figure-c" class="nav-link" data-scroll-target="#sec-si-figure-c"><span class="header-section-number">7.3</span> Supplementary Text Figure C</a></li>
  <li><a href="#sec-si-figure-d" id="toc-sec-si-figure-d" class="nav-link" data-scroll-target="#sec-si-figure-d"><span class="header-section-number">7.4</span> Supplementary Text Figure D</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./figure.html">Figures</a></li><li class="breadcrumb-item"><a href="./suppl_figure.html"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Supplementary Text Figures</span></a></li></ol></nav>
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-suppl-figure" class="quarto-section-identifier"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Supplementary Text Figures</span></span></h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>This chapter demonstrates the <code>R</code> codes that make supplementary text figures shown in the paper. Making the main figures are shown in <a href="main_figure.html" class="quarto-xref"><span>Chapter 6</span></a>.</p>
<section id="sec-si-figure-a" class="level2" data-number="7.1">
<h2 data-number="7.1" class="anchored" data-anchor-id="sec-si-figure-a"><span class="header-section-number">7.1</span> Supplementary Text Figure A</h2>
<p><img src="static/figure/cfRNA.SI.FigA.png" class="img-fluid"></p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>  <span class="fu">lazyLoad</span>(<span class="st">"downsample.chrY.TPM_cache/beamer/load_ds_salmon_4b8dd727729fd0ae5deb423d69e7c16d"</span>) <span class="co"># dl.chrY.TPM and dl.stat</span></span>
<span id="cb1-2"><a href="#cb1-2"></a></span>
<span id="cb1-3"><a href="#cb1-3"></a>  dt.ds.chrY <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(dl.stat), <span class="cf">function</span>(DS) dl.stat[[DS]][,<span class="at">DS:=</span>DS]) <span class="sc">%&gt;%</span> rbindlist</span>
<span id="cb1-4"><a href="#cb1-4"></a>  dt.ds.chrY<span class="sc">$</span>DS<span class="ot">&lt;-</span><span class="fu">factor</span>(dt.ds.chrY<span class="sc">$</span>DS, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"1M"</span>,<span class="st">"5M"</span>,<span class="st">"10M"</span>))</span>
<span id="cb1-5"><a href="#cb1-5"></a>  dt.ds.chrY[,GA<span class="sc">:</span><span class="er">=</span><span class="fu">gsub</span>(<span class="st">" weeks"</span>,<span class="st">"wk"</span>,GA)]</span>
<span id="cb1-6"><a href="#cb1-6"></a>  <span class="fu">setnames</span>(dt.ds.chrY, <span class="st">"Type"</span>, <span class="st">"Method"</span>)</span>
<span id="cb1-7"><a href="#cb1-7"></a></span>
<span id="cb1-8"><a href="#cb1-8"></a>  p1.chrY<span class="ot">&lt;-</span><span class="fu">ggplot</span>(dt.ds.chrY, <span class="fu">aes</span>(Method,SumCount)) <span class="sc">+</span></span>
<span id="cb1-9"><a href="#cb1-9"></a>  <span class="fu">geom_rect</span>(<span class="at">data =</span> dt.ds.chrY[DS<span class="sc">==</span><span class="st">"5M"</span>],<span class="at">fill=</span><span class="st">"grey60"</span>,<span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">ymax =</span> <span class="cn">Inf</span>,<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb1-10"><a href="#cb1-10"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">fill=</span>Method), <span class="at">stat=</span><span class="st">"identity"</span>,<span class="at">width=</span>.<span class="dv">5</span>, <span class="at">alpha=</span>.<span class="dv">7</span>) <span class="sc">+</span></span>
<span id="cb1-11"><a href="#cb1-11"></a>  ggsci<span class="sc">::</span><span class="fu">scale_fill_jama</span>() <span class="sc">+</span></span>
<span id="cb1-12"><a href="#cb1-12"></a>  <span class="fu">facet_grid</span>(Sex<span class="sc">~</span>DS<span class="sc">+</span>GA,<span class="at">scales=</span><span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb1-13"><a href="#cb1-13"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">"Sum of read count on chrY genes"</span>) <span class="sc">+</span></span>
<span id="cb1-14"><a href="#cb1-14"></a>  <span class="fu">theme_Publication</span>() <span class="sc">+</span></span>
<span id="cb1-15"><a href="#cb1-15"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"top"</span>,</span>
<span id="cb1-16"><a href="#cb1-16"></a>        <span class="at">panel.border=</span><span class="fu">element_rect</span>(<span class="at">colour=</span><span class="st">"black"</span>),</span>
<span id="cb1-17"><a href="#cb1-17"></a>        <span class="at">axis.title.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb1-18"><a href="#cb1-18"></a>        <span class="at">axis.text.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb1-19"><a href="#cb1-19"></a>        <span class="at">axis.ticks.x=</span><span class="fu">element_blank</span>())</span>
<span id="cb1-20"><a href="#cb1-20"></a></span>
<span id="cb1-21"><a href="#cb1-21"></a>  <span class="co">#</span></span>
<span id="cb1-22"><a href="#cb1-22"></a>  <span class="fu">pdf</span>(<span class="at">file=</span><span class="st">"Figures/Suppl/cfRNA.Suppl.Text.Fig1A.pdf"</span>, <span class="at">width=</span><span class="dv">11</span>, <span class="at">height=</span><span class="dv">8</span>, <span class="at">title=</span><span class="st">"Abundance of chrY"</span>)</span>
<span id="cb1-23"><a href="#cb1-23"></a>  p1.chrY</span>
<span id="cb1-24"><a href="#cb1-24"></a>  <span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="sec-si-figure-b" class="level2" data-number="7.2">
<h2 data-number="7.2" class="anchored" data-anchor-id="sec-si-figure-b"><span class="header-section-number">7.2</span> Supplementary Text Figure B</h2>
<p><img src="static/figure/cfRNA.SI.FigB.png" class="img-fluid"></p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>  <span class="fu">lazyLoad</span>(<span class="st">"downsample.chrY.TPM_cache/beamer/load_ds_salmon_4b8dd727729fd0ae5deb423d69e7c16d"</span>) <span class="co"># dl.chrY.TPM and dl.stat</span></span>
<span id="cb2-2"><a href="#cb2-2"></a></span>
<span id="cb2-3"><a href="#cb2-3"></a>  <span class="fu">lapply</span>(dl.chrY.TPM, <span class="cf">function</span>(i) i[,.(<span class="fu">sum</span>(Counts),<span class="fu">sum</span>(TPM))])</span>
<span id="cb2-4"><a href="#cb2-4"></a></span>
<span id="cb2-5"><a href="#cb2-5"></a>  dl.chrY.TPM<span class="ot">&lt;-</span><span class="fu">lapply</span>(dl.chrY.TPM, <span class="cf">function</span>(DT) {</span>
<span id="cb2-6"><a href="#cb2-6"></a>                  <span class="co">#DT[Sex=="Male",Class:=ifelse(TPM&gt;0,"TP","FN")]</span></span>
<span id="cb2-7"><a href="#cb2-7"></a>                  <span class="co">#DT[Sex=="Female",Class:=ifelse(TPM==0,"TN","FP")] </span></span>
<span id="cb2-8"><a href="#cb2-8"></a>                  DT[,<span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span><span class="sc">:</span><span class="er">=</span><span class="fu">ifelse</span>(TPM<span class="sc">&gt;</span><span class="dv">0</span>, <span class="st">"Male"</span>,<span class="st">"Female"</span>),Type]</span>
<span id="cb2-9"><a href="#cb2-9"></a>                  DT<span class="sc">$</span>Sex<span class="ot">&lt;-</span><span class="fu">factor</span>(DT<span class="sc">$</span>Sex, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"Male"</span>,<span class="st">"Female"</span>))</span>
<span id="cb2-10"><a href="#cb2-10"></a>                  DT<span class="sc">$</span><span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span><span class="ot">&lt;-</span><span class="fu">factor</span>(DT<span class="sc">$</span><span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span>, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"Male"</span>,<span class="st">"Female"</span>))</span>
<span id="cb2-11"><a href="#cb2-11"></a>                  DT </span>
<span id="cb2-12"><a href="#cb2-12"></a>  }) </span>
<span id="cb2-13"><a href="#cb2-13"></a></span>
<span id="cb2-14"><a href="#cb2-14"></a>  <span class="do">#####################################</span></span>
<span id="cb2-15"><a href="#cb2-15"></a>  <span class="co"># Predictive performance by DS &amp; GA #</span></span>
<span id="cb2-16"><a href="#cb2-16"></a>  <span class="do">#####################################</span></span>
<span id="cb2-17"><a href="#cb2-17"></a>  dt.chrY.con.ga<span class="ot">&lt;-</span><span class="fu">lapply</span>(<span class="fu">names</span>(dl.chrY.TPM), <span class="cf">function</span>(DS){</span>
<span id="cb2-18"><a href="#cb2-18"></a>    <span class="fu">lapply</span>(<span class="fu">split</span>(dl.chrY.TPM[[DS]], dl.chrY.TPM[[DS]]<span class="sc">$</span>GA), <span class="cf">function</span>(DT){</span>
<span id="cb2-19"><a href="#cb2-19"></a>    my.GA<span class="ot">&lt;-</span>DT[,.N,GA]<span class="sc">$</span>GA</span>
<span id="cb2-20"><a href="#cb2-20"></a></span>
<span id="cb2-21"><a href="#cb2-21"></a>    cm1<span class="ot">&lt;-</span>caret<span class="sc">::</span><span class="fu">confusionMatrix</span>(DT[Type<span class="sc">==</span><span class="st">"HiSat2+Salmon"</span>][[<span class="st">"Predicted Sex"</span>]], <span class="at">reference=</span>DT[Type<span class="sc">==</span><span class="st">"HiSat2+Salmon"</span>][[<span class="st">"Sex"</span>]])</span>
<span id="cb2-22"><a href="#cb2-22"></a>    cm2<span class="ot">&lt;-</span>caret<span class="sc">::</span><span class="fu">confusionMatrix</span>(DT[Type<span class="sc">==</span><span class="st">"Salmon (SA mode)"</span>][[<span class="st">"Predicted Sex"</span>]], <span class="at">reference=</span>DT[Type<span class="sc">==</span><span class="st">"Salmon (SA mode)"</span>][[<span class="st">"Sex"</span>]])</span>
<span id="cb2-23"><a href="#cb2-23"></a></span>
<span id="cb2-24"><a href="#cb2-24"></a>    <span class="fu">rbind</span>(</span>
<span id="cb2-25"><a href="#cb2-25"></a>    <span class="fu">data.table</span>(<span class="at">DS=</span>DS, <span class="at">GA=</span>my.GA,<span class="at">Type=</span><span class="st">"HiSat2+Salmon"</span>, <span class="at">Measure=</span><span class="fu">names</span>(cm1<span class="sc">$</span>byClass), <span class="at">Value=</span>cm1<span class="sc">$</span>byClass),</span>
<span id="cb2-26"><a href="#cb2-26"></a>    <span class="fu">data.table</span>(<span class="at">DS=</span>DS, <span class="at">GA=</span>my.GA,<span class="at">Type=</span><span class="st">"Salmon (SA mode)"</span>, <span class="at">Measure=</span><span class="fu">names</span>(cm2<span class="sc">$</span>byClass), <span class="at">Value=</span>cm2<span class="sc">$</span>byClass)</span>
<span id="cb2-27"><a href="#cb2-27"></a>    )[Measure <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"F1"</span>,<span class="st">"Precision"</span>, <span class="st">"Recall"</span>)]</span>
<span id="cb2-28"><a href="#cb2-28"></a>    }) <span class="sc">%&gt;%</span> rbindlist </span>
<span id="cb2-29"><a href="#cb2-29"></a>  }) <span class="sc">%&gt;%</span> rbindlist </span>
<span id="cb2-30"><a href="#cb2-30"></a></span>
<span id="cb2-31"><a href="#cb2-31"></a>  dt.chrY.con.ga<span class="sc">$</span>DS<span class="ot">&lt;-</span><span class="fu">factor</span>(dt.chrY.con.ga<span class="sc">$</span>DS, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"1M"</span>,<span class="st">"5M"</span>,<span class="st">"10M"</span>))</span>
<span id="cb2-32"><a href="#cb2-32"></a>  dt.chrY.con.ga[,GA<span class="sc">:</span><span class="er">=</span><span class="fu">gsub</span>(<span class="st">" weeks"</span>,<span class="st">"wk"</span>,GA)]</span>
<span id="cb2-33"><a href="#cb2-33"></a>  <span class="fu">setnames</span>(dt.chrY.con.ga, <span class="st">"Type"</span>, <span class="st">"Method"</span>)</span>
<span id="cb2-34"><a href="#cb2-34"></a></span>
<span id="cb2-35"><a href="#cb2-35"></a>  p1.chrY.con<span class="ot">&lt;-</span><span class="fu">ggplot</span>(dt.chrY.con.ga, <span class="fu">aes</span>(Method,Value)) <span class="sc">+</span></span>
<span id="cb2-36"><a href="#cb2-36"></a>  <span class="fu">geom_rect</span>(<span class="at">data =</span> dt.chrY.con.ga[DS<span class="sc">==</span><span class="st">"5M"</span>],<span class="at">fill=</span><span class="st">"grey60"</span>,<span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="sc">-</span><span class="cn">Inf</span>,<span class="at">ymax =</span> <span class="cn">Inf</span>,<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb2-37"><a href="#cb2-37"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">fill=</span>Method), <span class="at">stat=</span><span class="st">"identity"</span>,<span class="at">width=</span>.<span class="dv">5</span>, <span class="at">alpha=</span>.<span class="dv">7</span>) <span class="sc">+</span></span>
<span id="cb2-38"><a href="#cb2-38"></a>  ggsci<span class="sc">::</span><span class="fu">scale_fill_jama</span>() <span class="sc">+</span></span>
<span id="cb2-39"><a href="#cb2-39"></a>  <span class="fu">facet_grid</span>(Measure<span class="sc">~</span>DS<span class="sc">+</span>GA,<span class="at">scales=</span><span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb2-40"><a href="#cb2-40"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">"Score"</span>) <span class="sc">+</span></span>
<span id="cb2-41"><a href="#cb2-41"></a>  <span class="fu">theme_Publication</span>() <span class="sc">+</span></span>
<span id="cb2-42"><a href="#cb2-42"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>,</span>
<span id="cb2-43"><a href="#cb2-43"></a>        <span class="at">panel.border=</span><span class="fu">element_rect</span>(<span class="at">colour=</span><span class="st">"black"</span>),</span>
<span id="cb2-44"><a href="#cb2-44"></a>        <span class="at">axis.title.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb2-45"><a href="#cb2-45"></a>        <span class="at">axis.text.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb2-46"><a href="#cb2-46"></a>        <span class="at">axis.ticks.x=</span><span class="fu">element_blank</span>())</span>
<span id="cb2-47"><a href="#cb2-47"></a></span>
<span id="cb2-48"><a href="#cb2-48"></a></span>
<span id="cb2-49"><a href="#cb2-49"></a>  <span class="co"># merge SI.Fig1A + SI.Fig1B</span></span>
<span id="cb2-50"><a href="#cb2-50"></a>  <span class="fu">pdf</span>(<span class="at">file=</span><span class="st">"Figures/Suppl/cfRNA.Suppl.Text.Fig1AB.pdf"</span>, <span class="at">width=</span><span class="dv">11</span>, <span class="at">height=</span><span class="dv">14</span>, <span class="at">title=</span><span class="st">"Abundance of chrY"</span>)</span>
<span id="cb2-51"><a href="#cb2-51"></a>  cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(p1.chrY, p1.chrY.con, <span class="at">labels=</span><span class="st">"AUTO"</span>, <span class="at">ncol=</span><span class="dv">1</span>, <span class="at">label_size=</span><span class="dv">27</span>, <span class="at">align=</span><span class="st">"v"</span>, <span class="at">axis=</span><span class="st">"l"</span>)</span>
<span id="cb2-52"><a href="#cb2-52"></a>  <span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="sec-si-figure-c" class="level2" data-number="7.3">
<h2 data-number="7.3" class="anchored" data-anchor-id="sec-si-figure-c"><span class="header-section-number">7.3</span> Supplementary Text Figure C</h2>
<p><img src="static/figure/cfRNA.SI.FigC.png" class="img-fluid"></p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>  <span class="co"># load the transcript definition used in Salmon (SA)</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>  <span class="fu">load</span>(<span class="st">"RData/gr.ensg.Homo_sapiens.GRCh38.88.cdna.all.ncrna.fa.gz.RData"</span>) <span class="sc">%&gt;%</span> system.time <span class="co"># isa GRanges</span></span>
<span id="cb3-3"><a href="#cb3-3"></a>  gr.ensg[<span class="fu">seqnames</span>(gr.ensg)<span class="sc">==</span><span class="st">"Y"</span> <span class="sc">&amp;</span> gr.ensg<span class="sc">$</span>gene_biotype<span class="sc">==</span><span class="st">"protein_coding"</span>,] <span class="co"># chrY protein-coding; n=42 </span></span>
<span id="cb3-4"><a href="#cb3-4"></a>  this.gene<span class="ot">&lt;-</span>(gr.ensg[<span class="fu">seqnames</span>(gr.ensg)<span class="sc">==</span><span class="st">"Y"</span> <span class="sc">&amp;</span> gr.ensg<span class="sc">$</span>gene_biotype<span class="sc">==</span><span class="st">"protein_coding"</span>,] <span class="sc">%&gt;%</span> names) <span class="co"># chrY protein-coding; n=42 (without the training version e.g. ENSG00000176679)</span></span>
<span id="cb3-5"><a href="#cb3-5"></a></span>
<span id="cb3-6"><a href="#cb3-6"></a>  <span class="co"># load the TPM based om 10M-ds dataset</span></span>
<span id="cb3-7"><a href="#cb3-7"></a>  <span class="fu">load</span>(<span class="st">"RData/li.TPM.10M.RData"</span>)</span>
<span id="cb3-8"><a href="#cb3-8"></a>  li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"TPM"</span>]] <span class="sc">%&gt;%</span> dim <span class="co"># 62803 genes x 100 samples</span></span>
<span id="cb3-9"><a href="#cb3-9"></a></span>
<span id="cb3-10"><a href="#cb3-10"></a>  <span class="do">## Heatamp of TPM using `HiSat2+Salmon` (based on 10M)</span></span>
<span id="cb3-11"><a href="#cb3-11"></a>  li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> rowSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="co"># count of 42 chrY genes</span></span>
<span id="cb3-12"><a href="#cb3-12"></a>  filter.g2<span class="ot">&lt;-</span>li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> rowSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="sc">&gt;</span><span class="dv">0</span> <span class="co"># filter for 42 chrY genes with TPM &gt;0</span></span>
<span id="cb3-13"><a href="#cb3-13"></a>  this.gene2<span class="ot">&lt;-</span>(li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> rowSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev)[filter.g2] <span class="sc">%&gt;%</span> names <span class="co"># n=24 genes</span></span>
<span id="cb3-14"><a href="#cb3-14"></a>  this.sample2<span class="ot">&lt;-</span>li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="sc">%&gt;%</span> names <span class="co"># n=100 samples sorted </span></span>
<span id="cb3-15"><a href="#cb3-15"></a>  filter.s2<span class="ot">&lt;-</span>li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene2,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="sc">&gt;</span><span class="dv">0</span> <span class="co"># filter for samples with TPM &gt;0 </span></span>
<span id="cb3-16"><a href="#cb3-16"></a>  predicted.male2<span class="ot">&lt;-</span>(li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev)[filter.s2] <span class="sc">%&gt;%</span> names <span class="co"># n=60 samples </span></span>
<span id="cb3-17"><a href="#cb3-17"></a></span>
<span id="cb3-18"><a href="#cb3-18"></a>  my.mat2<span class="ot">&lt;-</span><span class="fu">log2</span>(li.TPM[[<span class="st">"Salmon_aln"</span>]][[<span class="st">"counts"</span>]][this.gene2,this.sample2]<span class="sc">+</span><span class="dv">1</span>)</span>
<span id="cb3-19"><a href="#cb3-19"></a>  <span class="fu">rownames</span>(my.mat2)<span class="ot">&lt;-</span>gr.ensg[this.gene2]<span class="sc">$</span>gene_name</span>
<span id="cb3-20"><a href="#cb3-20"></a></span>
<span id="cb3-21"><a href="#cb3-21"></a>  col.anno<span class="ot">&lt;-</span>mat.samples[this.sample2,<span class="fu">c</span>(<span class="st">"GA"</span>,<span class="st">"Sex"</span>),drop<span class="ot">=</span>F] <span class="co"># isa `data.frame`</span></span>
<span id="cb3-22"><a href="#cb3-22"></a>  col.anno<span class="sc">$</span><span class="st">"Predicted Sex"</span><span class="ot">&lt;-</span><span class="fu">ifelse</span>(this.sample2 <span class="sc">%in%</span> predicted.male2,<span class="st">"Male"</span>,<span class="st">"Female"</span>)</span>
<span id="cb3-23"><a href="#cb3-23"></a>  col.anno<span class="ot">&lt;-</span>col.anno[,<span class="fu">c</span>(<span class="st">"Predicted Sex"</span>,<span class="st">"Sex"</span>,<span class="st">"GA"</span>)]</span>
<span id="cb3-24"><a href="#cb3-24"></a></span>
<span id="cb3-25"><a href="#cb3-25"></a>  GA.color<span class="ot">&lt;-</span>ggsci<span class="sc">::</span><span class="fu">pal_jama</span>(<span class="st">"default"</span>,<span class="at">alpha=</span>.<span class="dv">9</span>)(<span class="dv">4</span>)</span>
<span id="cb3-26"><a href="#cb3-26"></a>  <span class="fu">names</span>(GA.color)<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"12wk"</span>,<span class="st">"20wk"</span>,<span class="st">"28wk"</span>,<span class="st">"36wk"</span>)</span>
<span id="cb3-27"><a href="#cb3-27"></a>  my.color<span class="ot">=</span><span class="fu">list</span>(</span>
<span id="cb3-28"><a href="#cb3-28"></a>      <span class="st">`</span><span class="at">Sex</span><span class="st">`</span><span class="ot">=</span><span class="fu">c</span>(<span class="at">Female=</span><span class="st">"hotpink"</span>,<span class="at">Male=</span><span class="st">'skyblue'</span>),</span>
<span id="cb3-29"><a href="#cb3-29"></a>      <span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span><span class="ot">=</span><span class="fu">c</span>(<span class="at">Female=</span><span class="st">"hotpink3"</span>,<span class="at">Male=</span><span class="st">'skyblue3'</span>),</span>
<span id="cb3-30"><a href="#cb3-30"></a>      <span class="co">#`GA`=c("12wk"=cbPalette2[1],"20wk"=cbPalette2[2],"28wk"=cbPalette2[3],"36wk"=cbPalette2[4])</span></span>
<span id="cb3-31"><a href="#cb3-31"></a>      <span class="st">`</span><span class="at">GA</span><span class="st">`</span><span class="ot">=</span>GA.color</span>
<span id="cb3-32"><a href="#cb3-32"></a>  )</span>
<span id="cb3-33"><a href="#cb3-33"></a></span>
<span id="cb3-34"><a href="#cb3-34"></a>  cph1<span class="ot">&lt;-</span>ComplexHeatmap<span class="sc">::</span><span class="fu">pheatmap</span>(my.mat2,</span>
<span id="cb3-35"><a href="#cb3-35"></a>                                 <span class="at">annotation_col=</span>col.anno, </span>
<span id="cb3-36"><a href="#cb3-36"></a>                                 <span class="at">annotation_colors=</span>my.color, </span>
<span id="cb3-37"><a href="#cb3-37"></a>                                 <span class="at">cluster_cols=</span>F,</span>
<span id="cb3-38"><a href="#cb3-38"></a>                                 <span class="at">cluster_rows=</span>F,</span>
<span id="cb3-39"><a href="#cb3-39"></a>                                 <span class="at">show_colnames=</span>F,</span>
<span id="cb3-40"><a href="#cb3-40"></a>                                 <span class="at">border_color=</span><span class="cn">NA</span>,</span>
<span id="cb3-41"><a href="#cb3-41"></a>                                 <span class="at">fontsize=</span><span class="dv">12</span>,</span>
<span id="cb3-42"><a href="#cb3-42"></a>                                 <span class="at">name=</span><span class="st">"log2(count)"</span>,</span>
<span id="cb3-43"><a href="#cb3-43"></a>                                 <span class="at">row_title=</span><span class="st">"</span><span class="sc">\n</span><span class="st">HiSat2+Salmon"</span>,</span>
<span id="cb3-44"><a href="#cb3-44"></a>                                 <span class="at">row_title_gp =</span> <span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">20</span>),</span>
<span id="cb3-45"><a href="#cb3-45"></a>  )</span>
<span id="cb3-46"><a href="#cb3-46"></a></span>
<span id="cb3-47"><a href="#cb3-47"></a>  filter.g3<span class="ot">&lt;-</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> rowSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="sc">&gt;</span><span class="dv">0</span> <span class="co"># filter for 42 chrY genes with count &gt;0</span></span>
<span id="cb3-48"><a href="#cb3-48"></a>  this.gene3<span class="ot">&lt;-</span>(li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> rowSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev)[filter.g3] <span class="sc">%&gt;%</span> names <span class="co"># n=16 genes (out of 42) &gt;0 count </span></span>
<span id="cb3-49"><a href="#cb3-49"></a>  this.sample3<span class="ot">&lt;-</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="sc">%&gt;%</span> names <span class="co"># 100 sample names</span></span>
<span id="cb3-50"><a href="#cb3-50"></a>  filter.s3<span class="ot">&lt;-</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene3,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev <span class="sc">&gt;</span><span class="dv">0</span> <span class="co"># samples with count&gt;0 </span></span>
<span id="cb3-51"><a href="#cb3-51"></a>  <span class="co">#(li.TPM[["Salmon"]][["counts"]][this.gene,] %&gt;% colSums %&gt;% sort %&gt;% rev)[filter.s3] # n=38 samples with count&gt;0 (predicted as Male)</span></span>
<span id="cb3-52"><a href="#cb3-52"></a>  predicted.male3<span class="ot">&lt;-</span>(li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> sort <span class="sc">%&gt;%</span> rev)[filter.s3] <span class="sc">%&gt;%</span> names <span class="co"># n=38 samples with count&gt;0 (predicted as Male)</span></span>
<span id="cb3-53"><a href="#cb3-53"></a></span>
<span id="cb3-54"><a href="#cb3-54"></a>  my.mat3<span class="ot">&lt;-</span><span class="fu">log2</span>(li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene3,this.sample3]<span class="sc">+</span><span class="dv">1</span>)</span>
<span id="cb3-55"><a href="#cb3-55"></a>  <span class="fu">rownames</span>(my.mat3)<span class="ot">&lt;-</span>gr.ensg[this.gene3]<span class="sc">$</span>gene_name</span>
<span id="cb3-56"><a href="#cb3-56"></a></span>
<span id="cb3-57"><a href="#cb3-57"></a>  col.anno3<span class="ot">&lt;-</span>mat.samples[this.sample3,<span class="fu">c</span>(<span class="st">"GA"</span>,<span class="st">"Sex"</span>),drop<span class="ot">=</span>F]</span>
<span id="cb3-58"><a href="#cb3-58"></a>  col.anno3<span class="sc">$</span><span class="st">"Predicted Sex"</span><span class="ot">&lt;-</span><span class="fu">ifelse</span>(this.sample3 <span class="sc">%in%</span> predicted.male3,<span class="st">"Male"</span>,<span class="st">"Female"</span>)</span>
<span id="cb3-59"><a href="#cb3-59"></a>  col.anno3<span class="ot">&lt;-</span>col.anno3[,<span class="fu">c</span>(<span class="st">"Predicted Sex"</span>,<span class="st">"Sex"</span>,<span class="st">"GA"</span>)]</span>
<span id="cb3-60"><a href="#cb3-60"></a></span>
<span id="cb3-61"><a href="#cb3-61"></a>  <span class="co"># via ComplexHeatmap</span></span>
<span id="cb3-62"><a href="#cb3-62"></a>  cph3<span class="ot">&lt;-</span>ComplexHeatmap<span class="sc">::</span><span class="fu">pheatmap</span>(my.mat3,</span>
<span id="cb3-63"><a href="#cb3-63"></a>                                 <span class="at">annotation_col=</span>col.anno3, </span>
<span id="cb3-64"><a href="#cb3-64"></a>                                 <span class="at">annotation_colors=</span>my.color, </span>
<span id="cb3-65"><a href="#cb3-65"></a>                                 <span class="at">cluster_cols=</span>F,</span>
<span id="cb3-66"><a href="#cb3-66"></a>                                 <span class="at">cluster_rows=</span>F,</span>
<span id="cb3-67"><a href="#cb3-67"></a>                                 <span class="at">show_colnames=</span>F,</span>
<span id="cb3-68"><a href="#cb3-68"></a>                                 <span class="at">border_color=</span><span class="cn">NA</span>,</span>
<span id="cb3-69"><a href="#cb3-69"></a>                                 <span class="at">fontsize=</span><span class="dv">12</span>,</span>
<span id="cb3-70"><a href="#cb3-70"></a>                                 <span class="at">name=</span><span class="st">"log2(count)"</span>,</span>
<span id="cb3-71"><a href="#cb3-71"></a>                                 <span class="at">row_title=</span><span class="st">"</span><span class="sc">\n</span><span class="st">Salmon (SA mode)"</span>,</span>
<span id="cb3-72"><a href="#cb3-72"></a>                                 <span class="at">row_title_gp =</span> <span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">20</span>),</span>
<span id="cb3-73"><a href="#cb3-73"></a>                                 <span class="at">heatmap_legend_param =</span> <span class="fu">list</span>(</span>
<span id="cb3-74"><a href="#cb3-74"></a>                                                            <span class="at">title_gp=</span><span class="fu">gpar</span>(<span class="at">fontsize=</span><span class="dv">13</span>,<span class="at">fontface=</span><span class="st">"bold"</span>),</span>
<span id="cb3-75"><a href="#cb3-75"></a>                                                            <span class="at">labels_gp=</span><span class="fu">gpar</span>(<span class="at">fontsize=</span><span class="dv">11</span>),</span>
<span id="cb3-76"><a href="#cb3-76"></a>                            <span class="at">annotation_legend_param =</span> <span class="fu">list</span>(<span class="at">title_gp=</span><span class="fu">gpar</span>(<span class="at">fontsize=</span><span class="dv">13</span>,<span class="at">fontface=</span><span class="st">"bold"</span>),<span class="at">labels_gp =</span> <span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">11</span>))</span>
<span id="cb3-77"><a href="#cb3-77"></a>                                )</span>
<span id="cb3-78"><a href="#cb3-78"></a>  )</span>
<span id="cb3-79"><a href="#cb3-79"></a></span>
<span id="cb3-80"><a href="#cb3-80"></a></span>
<span id="cb3-81"><a href="#cb3-81"></a>  <span class="fu">pdf</span>(<span class="at">file=</span><span class="st">"Figures/Suppl/cfRNA.Suppl.Text.Fig1C.pdf"</span>, <span class="at">width=</span><span class="dv">11</span>, <span class="at">height=</span><span class="dv">14</span>, <span class="at">title=</span><span class="st">"Abundance of chrY"</span>)</span>
<span id="cb3-82"><a href="#cb3-82"></a>  <span class="fu">draw</span>(cph1 <span class="sc">%v%</span> cph3, <span class="at">main_heatmap=</span><span class="st">"log2(count)"</span>, <span class="at">ht_gap=</span><span class="fu">unit</span>(.<span class="dv">7</span>,<span class="st">"cm"</span>), <span class="at">auto_adjust=</span>F)</span>
<span id="cb3-83"><a href="#cb3-83"></a>  <span class="fu">grid.text</span>(<span class="at">label=</span><span class="st">"C"</span>,<span class="at">gp=</span><span class="fu">gpar</span>(<span class="at">fontsize=</span><span class="dv">27</span>,<span class="at">fontface=</span><span class="st">"bold"</span>),<span class="at">x=</span><span class="fl">0.02</span>,<span class="at">y=</span>.<span class="dv">99</span>,<span class="at">just=</span><span class="st">"top"</span>)</span>
<span id="cb3-84"><a href="#cb3-84"></a>  <span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="sec-si-figure-d" class="level2" data-number="7.4">
<h2 data-number="7.4" class="anchored" data-anchor-id="sec-si-figure-d"><span class="header-section-number">7.4</span> Supplementary Text Figure D</h2>
<p><img src="static/figure/cfRNA.SI.FigD.png" class="img-fluid"></p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>  dl.chrY.TPM <span class="sc">%&gt;%</span> names</span>
<span id="cb4-2"><a href="#cb4-2"></a></span>
<span id="cb4-3"><a href="#cb4-3"></a>  <span class="co"># load the TPM based om 10M-ds dataset</span></span>
<span id="cb4-4"><a href="#cb4-4"></a>  <span class="fu">load</span>(<span class="st">"RData/li.TPM.10M.trim.RData"</span>)</span>
<span id="cb4-5"><a href="#cb4-5"></a>  li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"TPM"</span>]] <span class="sc">%&gt;%</span> dim <span class="co"># 59354 genes x 100 samples</span></span>
<span id="cb4-6"><a href="#cb4-6"></a></span>
<span id="cb4-7"><a href="#cb4-7"></a>  dt.trim<span class="fl">.10</span>M<span class="ot">&lt;-</span><span class="fu">data.table</span>(</span>
<span id="cb4-8"><a href="#cb4-8"></a>      <span class="st">`</span><span class="at">Type</span><span class="st">`</span><span class="ot">=</span><span class="st">"Salmon (SA mode)"</span>,</span>
<span id="cb4-9"><a href="#cb4-9"></a>      <span class="st">`</span><span class="at">SampleID</span><span class="st">`</span><span class="ot">=</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums <span class="sc">%&gt;%</span> names,</span>
<span id="cb4-10"><a href="#cb4-10"></a>      <span class="st">`</span><span class="at">Counts</span><span class="st">`</span><span class="ot">=</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"counts"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums, </span>
<span id="cb4-11"><a href="#cb4-11"></a>      <span class="st">`</span><span class="at">TPM</span><span class="st">`</span><span class="ot">=</span>li.TPM[[<span class="st">"Salmon"</span>]][[<span class="st">"TPM"</span>]][this.gene,] <span class="sc">%&gt;%</span> colSums </span>
<span id="cb4-12"><a href="#cb4-12"></a>      )</span>
<span id="cb4-13"><a href="#cb4-13"></a>  <span class="co"># add Sex and GA</span></span>
<span id="cb4-14"><a href="#cb4-14"></a>  dt.trim<span class="fl">.10</span>M<span class="ot">&lt;-</span><span class="fu">merge</span>(dt.trim<span class="fl">.10</span>M, dt.samples[,.(SampleID,GA,Sex,pn_female)])</span>
<span id="cb4-15"><a href="#cb4-15"></a></span>
<span id="cb4-16"><a href="#cb4-16"></a>  dt.trim<span class="fl">.10</span>M[,<span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span><span class="sc">:</span><span class="er">=</span><span class="fu">ifelse</span>(TPM<span class="sc">&gt;</span><span class="dv">0</span>, <span class="st">"Male"</span>,<span class="st">"Female"</span>),Type]</span>
<span id="cb4-17"><a href="#cb4-17"></a>  dt.trim<span class="fl">.10</span>M<span class="sc">$</span>Sex<span class="ot">&lt;-</span><span class="fu">factor</span>(dt.trim<span class="fl">.10</span>M<span class="sc">$</span>Sex, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"Male"</span>,<span class="st">"Female"</span>))</span>
<span id="cb4-18"><a href="#cb4-18"></a>  dt.trim<span class="fl">.10</span>M<span class="sc">$</span><span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span><span class="ot">&lt;-</span><span class="fu">factor</span>(dt.trim<span class="fl">.10</span>M<span class="sc">$</span><span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span>, <span class="at">level=</span><span class="fu">c</span>(<span class="st">"Male"</span>,<span class="st">"Female"</span>))</span>
<span id="cb4-19"><a href="#cb4-19"></a></span>
<span id="cb4-20"><a href="#cb4-20"></a>  dt.chrY.trim<span class="ot">&lt;-</span><span class="fu">rbind</span>(</span>
<span id="cb4-21"><a href="#cb4-21"></a>        dl.chrY.TPM[[<span class="st">"10M"</span>]][Type<span class="sc">==</span><span class="st">"Salmon (SA mode)"</span>][,<span class="at">Trim:=</span><span class="st">"No"</span>],</span>
<span id="cb4-22"><a href="#cb4-22"></a>        dt.trim<span class="fl">.10</span>M[,<span class="at">Trim:=</span><span class="st">"Yes"</span>]</span>
<span id="cb4-23"><a href="#cb4-23"></a>        )</span>
<span id="cb4-24"><a href="#cb4-24"></a>  <span class="do">## Number of false positives: chrY signal detected from females </span></span>
<span id="cb4-25"><a href="#cb4-25"></a>  dt.chrY.trim[,.(Trim,Sex,<span class="st">`</span><span class="at">Predicted Sex</span><span class="st">`</span>)] <span class="sc">%&gt;%</span> ftable</span>
<span id="cb4-26"><a href="#cb4-26"></a></span>
<span id="cb4-27"><a href="#cb4-27"></a>  <span class="do">##</span></span>
<span id="cb4-28"><a href="#cb4-28"></a>  <span class="do">## Read counts by trim or not</span></span>
<span id="cb4-29"><a href="#cb4-29"></a>  dt.bar<span class="ot">&lt;-</span><span class="fu">merge</span>(</span>
<span id="cb4-30"><a href="#cb4-30"></a>    dl.chrY.TPM[[<span class="st">"10M"</span>]][Type<span class="sc">==</span><span class="st">"Salmon (SA mode)"</span>,.(SampleID,<span class="at">GA=</span><span class="fu">gsub</span>(<span class="st">" weeks"</span>,<span class="st">"wk"</span>,GA),Sex,Counts,TPM)], <span class="co"># no-trim</span></span>
<span id="cb4-31"><a href="#cb4-31"></a>    dt.trim<span class="fl">.10</span>M[Type<span class="sc">==</span><span class="st">"Salmon (SA mode)"</span>,.(SampleID,GA,Sex,Counts,TPM)], <span class="co"># trim</span></span>
<span id="cb4-32"><a href="#cb4-32"></a>    <span class="at">by=</span><span class="fu">c</span>(<span class="st">"SampleID"</span>,<span class="st">"Sex"</span>,<span class="st">"GA"</span>)</span>
<span id="cb4-33"><a href="#cb4-33"></a>  )</span>
<span id="cb4-34"><a href="#cb4-34"></a>  <span class="fu">cor.test</span>(<span class="sc">~</span>Counts.x <span class="sc">+</span> Counts.y, dt.bar)</span>
<span id="cb4-35"><a href="#cb4-35"></a>  dt.bar[Counts.x<span class="sc">&gt;</span>Counts.y] <span class="co"># no-trim &gt; trim</span></span>
<span id="cb4-36"><a href="#cb4-36"></a>  dt.bar[Counts.x<span class="sc">&lt;</span>Counts.y] <span class="co"># no-trim &lt; trim</span></span>
<span id="cb4-37"><a href="#cb4-37"></a>  dt.bar[Sex<span class="sc">==</span><span class="st">"Female"</span> <span class="sc">&amp;</span> Counts.x<span class="sc">&lt;</span>Counts.y] <span class="co"># more counts for Trimmed Salmon</span></span>
<span id="cb4-38"><a href="#cb4-38"></a></span>
<span id="cb4-39"><a href="#cb4-39"></a>  my.limit<span class="ot">&lt;-</span>dt.bar[,<span class="fu">max</span>(Counts.x,Counts.y)]  <span class="sc">%&gt;%</span> ceiling</span>
<span id="cb4-40"><a href="#cb4-40"></a>  p1.chrY.trim<span class="ot">&lt;-</span><span class="fu">ggplot</span>(dt.bar, <span class="fu">aes</span>(Counts.x<span class="sc">+</span><span class="dv">1</span>,Counts.y<span class="sc">+</span><span class="dv">1</span>)) <span class="sc">+</span> </span>
<span id="cb4-41"><a href="#cb4-41"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>dt.bar, <span class="at">size=</span><span class="dv">5</span>, <span class="at">position=</span><span class="fu">position_jitter</span>(<span class="at">width=</span><span class="fl">0.2</span>,<span class="at">height=</span><span class="fl">0.2</span>),<span class="at">shape=</span><span class="dv">21</span>) <span class="sc">+</span></span>
<span id="cb4-42"><a href="#cb4-42"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">trans =</span> scales<span class="sc">::</span><span class="fu">log2_trans</span>(),</span>
<span id="cb4-43"><a href="#cb4-43"></a>                     <span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">trans_breaks</span>(<span class="st">"log2"</span>,<span class="cf">function</span>(x) <span class="dv">2</span><span class="sc">^</span>x),</span>
<span id="cb4-44"><a href="#cb4-44"></a>                     <span class="at">labels =</span> scales<span class="sc">::</span><span class="fu">trans_format</span>(<span class="st">"log2"</span>, <span class="fu">math_format</span>(<span class="dv">2</span><span class="sc">^</span>.x)),</span>
<span id="cb4-45"><a href="#cb4-45"></a>                     <span class="at">limits=</span><span class="fu">c</span>(.<span class="dv">5</span>,my.limit<span class="sc">*</span><span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb4-46"><a href="#cb4-46"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">trans =</span> scales<span class="sc">::</span><span class="fu">log2_trans</span>(),</span>
<span id="cb4-47"><a href="#cb4-47"></a>                     <span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">trans_breaks</span>(<span class="st">"log2"</span>,<span class="cf">function</span>(x) <span class="dv">2</span><span class="sc">^</span>x),</span>
<span id="cb4-48"><a href="#cb4-48"></a>                     <span class="at">labels =</span> scales<span class="sc">::</span><span class="fu">trans_format</span>(<span class="st">"log2"</span>, <span class="fu">math_format</span>(<span class="dv">2</span><span class="sc">^</span>.x)),</span>
<span id="cb4-49"><a href="#cb4-49"></a>                     <span class="at">limits=</span><span class="fu">c</span>(.<span class="dv">5</span>,my.limit<span class="sc">*</span><span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb4-50"><a href="#cb4-50"></a>  <span class="co">#coord_fixed() + </span></span>
<span id="cb4-51"><a href="#cb4-51"></a>  <span class="fu">facet_grid</span>(Sex<span class="sc">~</span>GA,<span class="at">scales=</span><span class="st">"free_x"</span>) <span class="sc">+</span></span>
<span id="cb4-52"><a href="#cb4-52"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Sum of read count (No trimmed reads)"</span>, <span class="at">y=</span><span class="st">"Sum of read count (trimmed reads)"</span>) <span class="sc">+</span></span>
<span id="cb4-53"><a href="#cb4-53"></a>  <span class="fu">theme_Publication</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"black"</span>))</span>
<span id="cb4-54"><a href="#cb4-54"></a></span>
<span id="cb4-55"><a href="#cb4-55"></a>  <span class="fu">pdf</span>(<span class="at">file=</span><span class="st">"Figures/Suppl/cfRNA.Suppl.Text.Fig1D.v2.pdf"</span>, <span class="at">width=</span><span class="dv">11</span>, <span class="at">height=</span><span class="dv">7</span>, <span class="at">title=</span><span class="st">"Trim vs No trim"</span>)</span>
<span id="cb4-56"><a href="#cb4-56"></a>  cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(p1.chrY.trim, <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"D"</span>),<span class="at">label_size=</span><span class="dv">27</span>)</span>
<span id="cb4-57"><a href="#cb4-57"></a>  <span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>


<!-- -->

</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = true;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./main_figure.html" class="pagination-link" aria-label="Main Figures">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Main Figures</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb5" data-shortcodes="false"><pre class="sourceCode numberSource markdown number-lines code-with-copy"><code class="sourceCode markdown"><span id="cb5-1"><a href="#cb5-1"></a><span class="fu"># Supplementary Text Figures {#sec-suppl-figure}</span></span>
<span id="cb5-2"><a href="#cb5-2"></a></span>
<span id="cb5-3"><a href="#cb5-3"></a>This chapter demonstrates the <span class="in">`R`</span> codes that make supplementary text figures shown in the paper. </span>
<span id="cb5-4"><a href="#cb5-4"></a>Making the main figures are shown in @sec-main-figure.</span>
<span id="cb5-5"><a href="#cb5-5"></a></span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="fu">## Supplementary Text Figure A {#sec-si-figure-a}</span></span>
<span id="cb5-7"><a href="#cb5-7"></a><span class="al">![](static/figure/cfRNA.SI.FigA.png)</span></span>
<span id="cb5-8"><a href="#cb5-8"></a></span>
<span id="cb5-9"><a href="#cb5-9"></a><span class="in">```{r }</span></span>
<span id="cb5-10"><a href="#cb5-10"></a><span class="in">#| label: fig-si-a</span></span>
<span id="cb5-11"><a href="#cb5-11"></a><span class="in">#| eval: false</span></span>
<span id="cb5-12"><a href="#cb5-12"></a><span class="in">  lazyLoad("downsample.chrY.TPM_cache/beamer/load_ds_salmon_4b8dd727729fd0ae5deb423d69e7c16d") # dl.chrY.TPM and dl.stat</span></span>
<span id="cb5-13"><a href="#cb5-13"></a></span>
<span id="cb5-14"><a href="#cb5-14"></a><span class="in">  dt.ds.chrY &lt;- lapply(names(dl.stat), function(DS) dl.stat[[DS]][,DS:=DS]) %&gt;% rbindlist</span></span>
<span id="cb5-15"><a href="#cb5-15"></a><span class="in">  dt.ds.chrY$DS&lt;-factor(dt.ds.chrY$DS, level=c("1M","5M","10M"))</span></span>
<span id="cb5-16"><a href="#cb5-16"></a><span class="in">  dt.ds.chrY[,GA:=gsub(" weeks","wk",GA)]</span></span>
<span id="cb5-17"><a href="#cb5-17"></a><span class="in">  setnames(dt.ds.chrY, "Type", "Method")</span></span>
<span id="cb5-18"><a href="#cb5-18"></a></span>
<span id="cb5-19"><a href="#cb5-19"></a><span class="in">  p1.chrY&lt;-ggplot(dt.ds.chrY, aes(Method,SumCount)) +</span></span>
<span id="cb5-20"><a href="#cb5-20"></a><span class="in">  geom_rect(data = dt.ds.chrY[DS=="5M"],fill="grey60",xmin = -Inf,xmax = Inf, ymin = -Inf,ymax = Inf,alpha = 0.2) +</span></span>
<span id="cb5-21"><a href="#cb5-21"></a><span class="in">  geom_bar(aes(fill=Method), stat="identity",width=.5, alpha=.7) +</span></span>
<span id="cb5-22"><a href="#cb5-22"></a><span class="in">  ggsci::scale_fill_jama() +</span></span>
<span id="cb5-23"><a href="#cb5-23"></a><span class="in">  facet_grid(Sex~DS+GA,scales="free") +</span></span>
<span id="cb5-24"><a href="#cb5-24"></a><span class="in">  labs(y="Sum of read count on chrY genes") +</span></span>
<span id="cb5-25"><a href="#cb5-25"></a><span class="in">  theme_Publication() +</span></span>
<span id="cb5-26"><a href="#cb5-26"></a><span class="in">  theme(legend.position="top",</span></span>
<span id="cb5-27"><a href="#cb5-27"></a><span class="in">        panel.border=element_rect(colour="black"),</span></span>
<span id="cb5-28"><a href="#cb5-28"></a><span class="in">        axis.title.x=element_blank(),</span></span>
<span id="cb5-29"><a href="#cb5-29"></a><span class="in">        axis.text.x=element_blank(),</span></span>
<span id="cb5-30"><a href="#cb5-30"></a><span class="in">        axis.ticks.x=element_blank())</span></span>
<span id="cb5-31"><a href="#cb5-31"></a></span>
<span id="cb5-32"><a href="#cb5-32"></a><span class="in">  #</span></span>
<span id="cb5-33"><a href="#cb5-33"></a><span class="in">  pdf(file="Figures/Suppl/cfRNA.Suppl.Text.Fig1A.pdf", width=11, height=8, title="Abundance of chrY")</span></span>
<span id="cb5-34"><a href="#cb5-34"></a><span class="in">  p1.chrY</span></span>
<span id="cb5-35"><a href="#cb5-35"></a><span class="in">  dev.off()</span></span>
<span id="cb5-36"><a href="#cb5-36"></a><span class="in">```</span></span>
<span id="cb5-37"><a href="#cb5-37"></a></span>
<span id="cb5-38"><a href="#cb5-38"></a><span class="fu">## Supplementary Text Figure B {#sec-si-figure-b}</span></span>
<span id="cb5-39"><a href="#cb5-39"></a><span class="al">![](static/figure/cfRNA.SI.FigB.png)</span></span>
<span id="cb5-40"><a href="#cb5-40"></a></span>
<span id="cb5-41"><a href="#cb5-41"></a><span class="in">```{r }</span></span>
<span id="cb5-42"><a href="#cb5-42"></a><span class="in">#| label: fig-si-b</span></span>
<span id="cb5-43"><a href="#cb5-43"></a><span class="in">#| eval: false</span></span>
<span id="cb5-44"><a href="#cb5-44"></a><span class="in">  lazyLoad("downsample.chrY.TPM_cache/beamer/load_ds_salmon_4b8dd727729fd0ae5deb423d69e7c16d") # dl.chrY.TPM and dl.stat</span></span>
<span id="cb5-45"><a href="#cb5-45"></a></span>
<span id="cb5-46"><a href="#cb5-46"></a><span class="in">  lapply(dl.chrY.TPM, function(i) i[,.(sum(Counts),sum(TPM))])</span></span>
<span id="cb5-47"><a href="#cb5-47"></a></span>
<span id="cb5-48"><a href="#cb5-48"></a><span class="in">  dl.chrY.TPM&lt;-lapply(dl.chrY.TPM, function(DT) {</span></span>
<span id="cb5-49"><a href="#cb5-49"></a><span class="in">                  #DT[Sex=="Male",Class:=ifelse(TPM&gt;0,"TP","FN")]</span></span>
<span id="cb5-50"><a href="#cb5-50"></a><span class="in">                  #DT[Sex=="Female",Class:=ifelse(TPM==0,"TN","FP")] </span></span>
<span id="cb5-51"><a href="#cb5-51"></a><span class="in">                  DT[,`Predicted Sex`:=ifelse(TPM&gt;0, "Male","Female"),Type]</span></span>
<span id="cb5-52"><a href="#cb5-52"></a><span class="in">                  DT$Sex&lt;-factor(DT$Sex, level=c("Male","Female"))</span></span>
<span id="cb5-53"><a href="#cb5-53"></a><span class="in">                  DT$`Predicted Sex`&lt;-factor(DT$`Predicted Sex`, level=c("Male","Female"))</span></span>
<span id="cb5-54"><a href="#cb5-54"></a><span class="in">                  DT </span></span>
<span id="cb5-55"><a href="#cb5-55"></a><span class="in">  }) </span></span>
<span id="cb5-56"><a href="#cb5-56"></a></span>
<span id="cb5-57"><a href="#cb5-57"></a><span class="in">  #####################################</span></span>
<span id="cb5-58"><a href="#cb5-58"></a><span class="in">  # Predictive performance by DS &amp; GA #</span></span>
<span id="cb5-59"><a href="#cb5-59"></a><span class="in">  #####################################</span></span>
<span id="cb5-60"><a href="#cb5-60"></a><span class="in">  dt.chrY.con.ga&lt;-lapply(names(dl.chrY.TPM), function(DS){</span></span>
<span id="cb5-61"><a href="#cb5-61"></a><span class="in">    lapply(split(dl.chrY.TPM[[DS]], dl.chrY.TPM[[DS]]$GA), function(DT){</span></span>
<span id="cb5-62"><a href="#cb5-62"></a><span class="in">    my.GA&lt;-DT[,.N,GA]$GA</span></span>
<span id="cb5-63"><a href="#cb5-63"></a></span>
<span id="cb5-64"><a href="#cb5-64"></a><span class="in">    cm1&lt;-caret::confusionMatrix(DT[Type=="HiSat2+Salmon"][["Predicted Sex"]], reference=DT[Type=="HiSat2+Salmon"][["Sex"]])</span></span>
<span id="cb5-65"><a href="#cb5-65"></a><span class="in">    cm2&lt;-caret::confusionMatrix(DT[Type=="Salmon (SA mode)"][["Predicted Sex"]], reference=DT[Type=="Salmon (SA mode)"][["Sex"]])</span></span>
<span id="cb5-66"><a href="#cb5-66"></a></span>
<span id="cb5-67"><a href="#cb5-67"></a><span class="in">    rbind(</span></span>
<span id="cb5-68"><a href="#cb5-68"></a><span class="in">    data.table(DS=DS, GA=my.GA,Type="HiSat2+Salmon", Measure=names(cm1$byClass), Value=cm1$byClass),</span></span>
<span id="cb5-69"><a href="#cb5-69"></a><span class="in">    data.table(DS=DS, GA=my.GA,Type="Salmon (SA mode)", Measure=names(cm2$byClass), Value=cm2$byClass)</span></span>
<span id="cb5-70"><a href="#cb5-70"></a><span class="in">    )[Measure %in% c("F1","Precision", "Recall")]</span></span>
<span id="cb5-71"><a href="#cb5-71"></a><span class="in">    }) %&gt;% rbindlist </span></span>
<span id="cb5-72"><a href="#cb5-72"></a><span class="in">  }) %&gt;% rbindlist </span></span>
<span id="cb5-73"><a href="#cb5-73"></a></span>
<span id="cb5-74"><a href="#cb5-74"></a><span class="in">  dt.chrY.con.ga$DS&lt;-factor(dt.chrY.con.ga$DS, level=c("1M","5M","10M"))</span></span>
<span id="cb5-75"><a href="#cb5-75"></a><span class="in">  dt.chrY.con.ga[,GA:=gsub(" weeks","wk",GA)]</span></span>
<span id="cb5-76"><a href="#cb5-76"></a><span class="in">  setnames(dt.chrY.con.ga, "Type", "Method")</span></span>
<span id="cb5-77"><a href="#cb5-77"></a></span>
<span id="cb5-78"><a href="#cb5-78"></a><span class="in">  p1.chrY.con&lt;-ggplot(dt.chrY.con.ga, aes(Method,Value)) +</span></span>
<span id="cb5-79"><a href="#cb5-79"></a><span class="in">  geom_rect(data = dt.chrY.con.ga[DS=="5M"],fill="grey60",xmin = -Inf,xmax = Inf, ymin = -Inf,ymax = Inf,alpha = 0.2) +</span></span>
<span id="cb5-80"><a href="#cb5-80"></a><span class="in">  geom_bar(aes(fill=Method), stat="identity",width=.5, alpha=.7) +</span></span>
<span id="cb5-81"><a href="#cb5-81"></a><span class="in">  ggsci::scale_fill_jama() +</span></span>
<span id="cb5-82"><a href="#cb5-82"></a><span class="in">  facet_grid(Measure~DS+GA,scales="free") +</span></span>
<span id="cb5-83"><a href="#cb5-83"></a><span class="in">  labs(y="Score") +</span></span>
<span id="cb5-84"><a href="#cb5-84"></a><span class="in">  theme_Publication() +</span></span>
<span id="cb5-85"><a href="#cb5-85"></a><span class="in">  theme(legend.position="none",</span></span>
<span id="cb5-86"><a href="#cb5-86"></a><span class="in">        panel.border=element_rect(colour="black"),</span></span>
<span id="cb5-87"><a href="#cb5-87"></a><span class="in">        axis.title.x=element_blank(),</span></span>
<span id="cb5-88"><a href="#cb5-88"></a><span class="in">        axis.text.x=element_blank(),</span></span>
<span id="cb5-89"><a href="#cb5-89"></a><span class="in">        axis.ticks.x=element_blank())</span></span>
<span id="cb5-90"><a href="#cb5-90"></a></span>
<span id="cb5-91"><a href="#cb5-91"></a></span>
<span id="cb5-92"><a href="#cb5-92"></a><span class="in">  # merge SI.Fig1A + SI.Fig1B</span></span>
<span id="cb5-93"><a href="#cb5-93"></a><span class="in">  pdf(file="Figures/Suppl/cfRNA.Suppl.Text.Fig1AB.pdf", width=11, height=14, title="Abundance of chrY")</span></span>
<span id="cb5-94"><a href="#cb5-94"></a><span class="in">  cowplot::plot_grid(p1.chrY, p1.chrY.con, labels="AUTO", ncol=1, label_size=27, align="v", axis="l")</span></span>
<span id="cb5-95"><a href="#cb5-95"></a><span class="in">  dev.off()</span></span>
<span id="cb5-96"><a href="#cb5-96"></a><span class="in">```</span></span>
<span id="cb5-97"><a href="#cb5-97"></a></span>
<span id="cb5-98"><a href="#cb5-98"></a><span class="fu">## Supplementary Text Figure C {#sec-si-figure-c}</span></span>
<span id="cb5-99"><a href="#cb5-99"></a><span class="al">![](static/figure/cfRNA.SI.FigC.png)</span></span>
<span id="cb5-100"><a href="#cb5-100"></a></span>
<span id="cb5-101"><a href="#cb5-101"></a><span class="in">```{r }</span></span>
<span id="cb5-102"><a href="#cb5-102"></a><span class="in">#| label: fig-si-c</span></span>
<span id="cb5-103"><a href="#cb5-103"></a><span class="in">#| eval: false</span></span>
<span id="cb5-104"><a href="#cb5-104"></a><span class="in">  # load the transcript definition used in Salmon (SA)</span></span>
<span id="cb5-105"><a href="#cb5-105"></a><span class="in">  load("RData/gr.ensg.Homo_sapiens.GRCh38.88.cdna.all.ncrna.fa.gz.RData") %&gt;% system.time # isa GRanges</span></span>
<span id="cb5-106"><a href="#cb5-106"></a><span class="in">  gr.ensg[seqnames(gr.ensg)=="Y" &amp; gr.ensg$gene_biotype=="protein_coding",] # chrY protein-coding; n=42 </span></span>
<span id="cb5-107"><a href="#cb5-107"></a><span class="in">  this.gene&lt;-(gr.ensg[seqnames(gr.ensg)=="Y" &amp; gr.ensg$gene_biotype=="protein_coding",] %&gt;% names) # chrY protein-coding; n=42 (without the training version e.g. ENSG00000176679)</span></span>
<span id="cb5-108"><a href="#cb5-108"></a></span>
<span id="cb5-109"><a href="#cb5-109"></a><span class="in">  # load the TPM based om 10M-ds dataset</span></span>
<span id="cb5-110"><a href="#cb5-110"></a><span class="in">  load("RData/li.TPM.10M.RData")</span></span>
<span id="cb5-111"><a href="#cb5-111"></a><span class="in">  li.TPM[["Salmon_aln"]][["TPM"]] %&gt;% dim # 62803 genes x 100 samples</span></span>
<span id="cb5-112"><a href="#cb5-112"></a></span>
<span id="cb5-113"><a href="#cb5-113"></a><span class="in">  ## Heatamp of TPM using `HiSat2+Salmon` (based on 10M)</span></span>
<span id="cb5-114"><a href="#cb5-114"></a><span class="in">  li.TPM[["Salmon_aln"]][["counts"]][this.gene,] %&gt;% rowSums %&gt;% sort %&gt;% rev # count of 42 chrY genes</span></span>
<span id="cb5-115"><a href="#cb5-115"></a><span class="in">  filter.g2&lt;-li.TPM[["Salmon_aln"]][["counts"]][this.gene,] %&gt;% rowSums %&gt;% sort %&gt;% rev &gt;0 # filter for 42 chrY genes with TPM &gt;0</span></span>
<span id="cb5-116"><a href="#cb5-116"></a><span class="in">  this.gene2&lt;-(li.TPM[["Salmon_aln"]][["counts"]][this.gene,] %&gt;% rowSums %&gt;% sort %&gt;% rev)[filter.g2] %&gt;% names # n=24 genes</span></span>
<span id="cb5-117"><a href="#cb5-117"></a><span class="in">  this.sample2&lt;-li.TPM[["Salmon_aln"]][["counts"]][this.gene,] %&gt;% colSums %&gt;% sort %&gt;% rev %&gt;% names # n=100 samples sorted </span></span>
<span id="cb5-118"><a href="#cb5-118"></a><span class="in">  filter.s2&lt;-li.TPM[["Salmon_aln"]][["counts"]][this.gene2,] %&gt;% colSums %&gt;% sort %&gt;% rev &gt;0 # filter for samples with TPM &gt;0 </span></span>
<span id="cb5-119"><a href="#cb5-119"></a><span class="in">  predicted.male2&lt;-(li.TPM[["Salmon_aln"]][["counts"]][this.gene,] %&gt;% colSums %&gt;% sort %&gt;% rev)[filter.s2] %&gt;% names # n=60 samples </span></span>
<span id="cb5-120"><a href="#cb5-120"></a></span>
<span id="cb5-121"><a href="#cb5-121"></a><span class="in">  my.mat2&lt;-log2(li.TPM[["Salmon_aln"]][["counts"]][this.gene2,this.sample2]+1)</span></span>
<span id="cb5-122"><a href="#cb5-122"></a><span class="in">  rownames(my.mat2)&lt;-gr.ensg[this.gene2]$gene_name</span></span>
<span id="cb5-123"><a href="#cb5-123"></a></span>
<span id="cb5-124"><a href="#cb5-124"></a><span class="in">  col.anno&lt;-mat.samples[this.sample2,c("GA","Sex"),drop=F] # isa `data.frame`</span></span>
<span id="cb5-125"><a href="#cb5-125"></a><span class="in">  col.anno$"Predicted Sex"&lt;-ifelse(this.sample2 %in% predicted.male2,"Male","Female")</span></span>
<span id="cb5-126"><a href="#cb5-126"></a><span class="in">  col.anno&lt;-col.anno[,c("Predicted Sex","Sex","GA")]</span></span>
<span id="cb5-127"><a href="#cb5-127"></a></span>
<span id="cb5-128"><a href="#cb5-128"></a><span class="in">  GA.color&lt;-ggsci::pal_jama("default",alpha=.9)(4)</span></span>
<span id="cb5-129"><a href="#cb5-129"></a><span class="in">  names(GA.color)&lt;-c("12wk","20wk","28wk","36wk")</span></span>
<span id="cb5-130"><a href="#cb5-130"></a><span class="in">  my.color=list(</span></span>
<span id="cb5-131"><a href="#cb5-131"></a><span class="in">      `Sex`=c(Female="hotpink",Male='skyblue'),</span></span>
<span id="cb5-132"><a href="#cb5-132"></a><span class="in">      `Predicted Sex`=c(Female="hotpink3",Male='skyblue3'),</span></span>
<span id="cb5-133"><a href="#cb5-133"></a><span class="in">      #`GA`=c("12wk"=cbPalette2[1],"20wk"=cbPalette2[2],"28wk"=cbPalette2[3],"36wk"=cbPalette2[4])</span></span>
<span id="cb5-134"><a href="#cb5-134"></a><span class="in">      `GA`=GA.color</span></span>
<span id="cb5-135"><a href="#cb5-135"></a><span class="in">  )</span></span>
<span id="cb5-136"><a href="#cb5-136"></a></span>
<span id="cb5-137"><a href="#cb5-137"></a><span class="in">  cph1&lt;-ComplexHeatmap::pheatmap(my.mat2,</span></span>
<span id="cb5-138"><a href="#cb5-138"></a><span class="in">                                 annotation_col=col.anno, </span></span>
<span id="cb5-139"><a href="#cb5-139"></a><span class="in">                                 annotation_colors=my.color, </span></span>
<span id="cb5-140"><a href="#cb5-140"></a><span class="in">                                 cluster_cols=F,</span></span>
<span id="cb5-141"><a href="#cb5-141"></a><span class="in">                                 cluster_rows=F,</span></span>
<span id="cb5-142"><a href="#cb5-142"></a><span class="in">                                 show_colnames=F,</span></span>
<span id="cb5-143"><a href="#cb5-143"></a><span class="in">                                 border_color=NA,</span></span>
<span id="cb5-144"><a href="#cb5-144"></a><span class="in">                                 fontsize=12,</span></span>
<span id="cb5-145"><a href="#cb5-145"></a><span class="in">                                 name="log2(count)",</span></span>
<span id="cb5-146"><a href="#cb5-146"></a><span class="in">                                 row_title="\nHiSat2+Salmon",</span></span>
<span id="cb5-147"><a href="#cb5-147"></a><span class="in">                                 row_title_gp = gpar(fontsize = 20),</span></span>
<span id="cb5-148"><a href="#cb5-148"></a><span class="in">  )</span></span>
<span id="cb5-149"><a href="#cb5-149"></a></span>
<span id="cb5-150"><a href="#cb5-150"></a><span class="in">  filter.g3&lt;-li.TPM[["Salmon"]][["counts"]][this.gene,] %&gt;% rowSums %&gt;% sort %&gt;% rev &gt;0 # filter for 42 chrY genes with count &gt;0</span></span>
<span id="cb5-151"><a href="#cb5-151"></a><span class="in">  this.gene3&lt;-(li.TPM[["Salmon"]][["counts"]][this.gene,] %&gt;% rowSums %&gt;% sort %&gt;% rev)[filter.g3] %&gt;% names # n=16 genes (out of 42) &gt;0 count </span></span>
<span id="cb5-152"><a href="#cb5-152"></a><span class="in">  this.sample3&lt;-li.TPM[["Salmon"]][["counts"]][this.gene,] %&gt;% colSums %&gt;% sort %&gt;% rev %&gt;% names # 100 sample names</span></span>
<span id="cb5-153"><a href="#cb5-153"></a><span class="in">  filter.s3&lt;-li.TPM[["Salmon"]][["counts"]][this.gene3,] %&gt;% colSums %&gt;% sort %&gt;% rev &gt;0 # samples with count&gt;0 </span></span>
<span id="cb5-154"><a href="#cb5-154"></a><span class="in">  #(li.TPM[["Salmon"]][["counts"]][this.gene,] %&gt;% colSums %&gt;% sort %&gt;% rev)[filter.s3] # n=38 samples with count&gt;0 (predicted as Male)</span></span>
<span id="cb5-155"><a href="#cb5-155"></a><span class="in">  predicted.male3&lt;-(li.TPM[["Salmon"]][["counts"]][this.gene,] %&gt;% colSums %&gt;% sort %&gt;% rev)[filter.s3] %&gt;% names # n=38 samples with count&gt;0 (predicted as Male)</span></span>
<span id="cb5-156"><a href="#cb5-156"></a></span>
<span id="cb5-157"><a href="#cb5-157"></a><span class="in">  my.mat3&lt;-log2(li.TPM[["Salmon"]][["counts"]][this.gene3,this.sample3]+1)</span></span>
<span id="cb5-158"><a href="#cb5-158"></a><span class="in">  rownames(my.mat3)&lt;-gr.ensg[this.gene3]$gene_name</span></span>
<span id="cb5-159"><a href="#cb5-159"></a></span>
<span id="cb5-160"><a href="#cb5-160"></a><span class="in">  col.anno3&lt;-mat.samples[this.sample3,c("GA","Sex"),drop=F]</span></span>
<span id="cb5-161"><a href="#cb5-161"></a><span class="in">  col.anno3$"Predicted Sex"&lt;-ifelse(this.sample3 %in% predicted.male3,"Male","Female")</span></span>
<span id="cb5-162"><a href="#cb5-162"></a><span class="in">  col.anno3&lt;-col.anno3[,c("Predicted Sex","Sex","GA")]</span></span>
<span id="cb5-163"><a href="#cb5-163"></a></span>
<span id="cb5-164"><a href="#cb5-164"></a><span class="in">  # via ComplexHeatmap</span></span>
<span id="cb5-165"><a href="#cb5-165"></a><span class="in">  cph3&lt;-ComplexHeatmap::pheatmap(my.mat3,</span></span>
<span id="cb5-166"><a href="#cb5-166"></a><span class="in">                                 annotation_col=col.anno3, </span></span>
<span id="cb5-167"><a href="#cb5-167"></a><span class="in">                                 annotation_colors=my.color, </span></span>
<span id="cb5-168"><a href="#cb5-168"></a><span class="in">                                 cluster_cols=F,</span></span>
<span id="cb5-169"><a href="#cb5-169"></a><span class="in">                                 cluster_rows=F,</span></span>
<span id="cb5-170"><a href="#cb5-170"></a><span class="in">                                 show_colnames=F,</span></span>
<span id="cb5-171"><a href="#cb5-171"></a><span class="in">                                 border_color=NA,</span></span>
<span id="cb5-172"><a href="#cb5-172"></a><span class="in">                                 fontsize=12,</span></span>
<span id="cb5-173"><a href="#cb5-173"></a><span class="in">                                 name="log2(count)",</span></span>
<span id="cb5-174"><a href="#cb5-174"></a><span class="in">                                 row_title="\nSalmon (SA mode)",</span></span>
<span id="cb5-175"><a href="#cb5-175"></a><span class="in">                                 row_title_gp = gpar(fontsize = 20),</span></span>
<span id="cb5-176"><a href="#cb5-176"></a><span class="in">                                 heatmap_legend_param = list(</span></span>
<span id="cb5-177"><a href="#cb5-177"></a><span class="in">                                                            title_gp=gpar(fontsize=13,fontface="bold"),</span></span>
<span id="cb5-178"><a href="#cb5-178"></a><span class="in">                                                            labels_gp=gpar(fontsize=11),</span></span>
<span id="cb5-179"><a href="#cb5-179"></a><span class="in">                            annotation_legend_param = list(title_gp=gpar(fontsize=13,fontface="bold"),labels_gp = gpar(fontsize = 11))</span></span>
<span id="cb5-180"><a href="#cb5-180"></a><span class="in">                                )</span></span>
<span id="cb5-181"><a href="#cb5-181"></a><span class="in">  )</span></span>
<span id="cb5-182"><a href="#cb5-182"></a></span>
<span id="cb5-183"><a href="#cb5-183"></a></span>
<span id="cb5-184"><a href="#cb5-184"></a><span class="in">  pdf(file="Figures/Suppl/cfRNA.Suppl.Text.Fig1C.pdf", width=11, height=14, title="Abundance of chrY")</span></span>
<span id="cb5-185"><a href="#cb5-185"></a><span class="in">  draw(cph1 %v% cph3, main_heatmap="log2(count)", ht_gap=unit(.7,"cm"), auto_adjust=F)</span></span>
<span id="cb5-186"><a href="#cb5-186"></a><span class="in">  grid.text(label="C",gp=gpar(fontsize=27,fontface="bold"),x=0.02,y=.99,just="top")</span></span>
<span id="cb5-187"><a href="#cb5-187"></a><span class="in">  dev.off()</span></span>
<span id="cb5-188"><a href="#cb5-188"></a><span class="in">```</span></span>
<span id="cb5-189"><a href="#cb5-189"></a></span>
<span id="cb5-190"><a href="#cb5-190"></a><span class="fu">## Supplementary Text Figure D {#sec-si-figure-d}</span></span>
<span id="cb5-191"><a href="#cb5-191"></a><span class="al">![](static/figure/cfRNA.SI.FigD.png)</span></span>
<span id="cb5-192"><a href="#cb5-192"></a></span>
<span id="cb5-193"><a href="#cb5-193"></a><span class="in">```{r }</span></span>
<span id="cb5-194"><a href="#cb5-194"></a><span class="in">#| label: fig-si-d</span></span>
<span id="cb5-195"><a href="#cb5-195"></a><span class="in">#| eval: false</span></span>
<span id="cb5-196"><a href="#cb5-196"></a><span class="in">  dl.chrY.TPM %&gt;% names</span></span>
<span id="cb5-197"><a href="#cb5-197"></a></span>
<span id="cb5-198"><a href="#cb5-198"></a><span class="in">  # load the TPM based om 10M-ds dataset</span></span>
<span id="cb5-199"><a href="#cb5-199"></a><span class="in">  load("RData/li.TPM.10M.trim.RData")</span></span>
<span id="cb5-200"><a href="#cb5-200"></a><span class="in">  li.TPM[["Salmon"]][["TPM"]] %&gt;% dim # 59354 genes x 100 samples</span></span>
<span id="cb5-201"><a href="#cb5-201"></a></span>
<span id="cb5-202"><a href="#cb5-202"></a><span class="in">  dt.trim.10M&lt;-data.table(</span></span>
<span id="cb5-203"><a href="#cb5-203"></a><span class="in">      `Type`="Salmon (SA mode)",</span></span>
<span id="cb5-204"><a href="#cb5-204"></a><span class="in">      `SampleID`=li.TPM[["Salmon"]][["counts"]][this.gene,] %&gt;% colSums %&gt;% names,</span></span>
<span id="cb5-205"><a href="#cb5-205"></a><span class="in">      `Counts`=li.TPM[["Salmon"]][["counts"]][this.gene,] %&gt;% colSums, </span></span>
<span id="cb5-206"><a href="#cb5-206"></a><span class="in">      `TPM`=li.TPM[["Salmon"]][["TPM"]][this.gene,] %&gt;% colSums </span></span>
<span id="cb5-207"><a href="#cb5-207"></a><span class="in">      )</span></span>
<span id="cb5-208"><a href="#cb5-208"></a><span class="in">  # add Sex and GA</span></span>
<span id="cb5-209"><a href="#cb5-209"></a><span class="in">  dt.trim.10M&lt;-merge(dt.trim.10M, dt.samples[,.(SampleID,GA,Sex,pn_female)])</span></span>
<span id="cb5-210"><a href="#cb5-210"></a></span>
<span id="cb5-211"><a href="#cb5-211"></a><span class="in">  dt.trim.10M[,`Predicted Sex`:=ifelse(TPM&gt;0, "Male","Female"),Type]</span></span>
<span id="cb5-212"><a href="#cb5-212"></a><span class="in">  dt.trim.10M$Sex&lt;-factor(dt.trim.10M$Sex, level=c("Male","Female"))</span></span>
<span id="cb5-213"><a href="#cb5-213"></a><span class="in">  dt.trim.10M$`Predicted Sex`&lt;-factor(dt.trim.10M$`Predicted Sex`, level=c("Male","Female"))</span></span>
<span id="cb5-214"><a href="#cb5-214"></a></span>
<span id="cb5-215"><a href="#cb5-215"></a><span class="in">  dt.chrY.trim&lt;-rbind(</span></span>
<span id="cb5-216"><a href="#cb5-216"></a><span class="in">        dl.chrY.TPM[["10M"]][Type=="Salmon (SA mode)"][,Trim:="No"],</span></span>
<span id="cb5-217"><a href="#cb5-217"></a><span class="in">        dt.trim.10M[,Trim:="Yes"]</span></span>
<span id="cb5-218"><a href="#cb5-218"></a><span class="in">        )</span></span>
<span id="cb5-219"><a href="#cb5-219"></a><span class="in">  ## Number of false positives: chrY signal detected from females </span></span>
<span id="cb5-220"><a href="#cb5-220"></a><span class="in">  dt.chrY.trim[,.(Trim,Sex,`Predicted Sex`)] %&gt;% ftable</span></span>
<span id="cb5-221"><a href="#cb5-221"></a></span>
<span id="cb5-222"><a href="#cb5-222"></a><span class="in">  ##</span></span>
<span id="cb5-223"><a href="#cb5-223"></a><span class="in">  ## Read counts by trim or not</span></span>
<span id="cb5-224"><a href="#cb5-224"></a><span class="in">  dt.bar&lt;-merge(</span></span>
<span id="cb5-225"><a href="#cb5-225"></a><span class="in">    dl.chrY.TPM[["10M"]][Type=="Salmon (SA mode)",.(SampleID,GA=gsub(" weeks","wk",GA),Sex,Counts,TPM)], # no-trim</span></span>
<span id="cb5-226"><a href="#cb5-226"></a><span class="in">    dt.trim.10M[Type=="Salmon (SA mode)",.(SampleID,GA,Sex,Counts,TPM)], # trim</span></span>
<span id="cb5-227"><a href="#cb5-227"></a><span class="in">    by=c("SampleID","Sex","GA")</span></span>
<span id="cb5-228"><a href="#cb5-228"></a><span class="in">  )</span></span>
<span id="cb5-229"><a href="#cb5-229"></a><span class="in">  cor.test(~Counts.x + Counts.y, dt.bar)</span></span>
<span id="cb5-230"><a href="#cb5-230"></a><span class="in">  dt.bar[Counts.x&gt;Counts.y] # no-trim &gt; trim</span></span>
<span id="cb5-231"><a href="#cb5-231"></a><span class="in">  dt.bar[Counts.x&lt;Counts.y] # no-trim &lt; trim</span></span>
<span id="cb5-232"><a href="#cb5-232"></a><span class="in">  dt.bar[Sex=="Female" &amp; Counts.x&lt;Counts.y] # more counts for Trimmed Salmon</span></span>
<span id="cb5-233"><a href="#cb5-233"></a></span>
<span id="cb5-234"><a href="#cb5-234"></a><span class="in">  my.limit&lt;-dt.bar[,max(Counts.x,Counts.y)]  %&gt;% ceiling</span></span>
<span id="cb5-235"><a href="#cb5-235"></a><span class="in">  p1.chrY.trim&lt;-ggplot(dt.bar, aes(Counts.x+1,Counts.y+1)) + </span></span>
<span id="cb5-236"><a href="#cb5-236"></a><span class="in">  geom_point(data=dt.bar, size=5, position=position_jitter(width=0.2,height=0.2),shape=21) +</span></span>
<span id="cb5-237"><a href="#cb5-237"></a><span class="in">  scale_x_continuous(trans = scales::log2_trans(),</span></span>
<span id="cb5-238"><a href="#cb5-238"></a><span class="in">                     breaks = scales::trans_breaks("log2",function(x) 2^x),</span></span>
<span id="cb5-239"><a href="#cb5-239"></a><span class="in">                     labels = scales::trans_format("log2", math_format(2^.x)),</span></span>
<span id="cb5-240"><a href="#cb5-240"></a><span class="in">                     limits=c(.5,my.limit*2)) +</span></span>
<span id="cb5-241"><a href="#cb5-241"></a><span class="in">  scale_y_continuous(trans = scales::log2_trans(),</span></span>
<span id="cb5-242"><a href="#cb5-242"></a><span class="in">                     breaks = scales::trans_breaks("log2",function(x) 2^x),</span></span>
<span id="cb5-243"><a href="#cb5-243"></a><span class="in">                     labels = scales::trans_format("log2", math_format(2^.x)),</span></span>
<span id="cb5-244"><a href="#cb5-244"></a><span class="in">                     limits=c(.5,my.limit*2)) +</span></span>
<span id="cb5-245"><a href="#cb5-245"></a><span class="in">  #coord_fixed() + </span></span>
<span id="cb5-246"><a href="#cb5-246"></a><span class="in">  facet_grid(Sex~GA,scales="free_x") +</span></span>
<span id="cb5-247"><a href="#cb5-247"></a><span class="in">  labs(x="Sum of read count (No trimmed reads)", y="Sum of read count (trimmed reads)") +</span></span>
<span id="cb5-248"><a href="#cb5-248"></a><span class="in">  theme_Publication() + theme(panel.border = element_rect(colour = "black"))</span></span>
<span id="cb5-249"><a href="#cb5-249"></a></span>
<span id="cb5-250"><a href="#cb5-250"></a><span class="in">  pdf(file="Figures/Suppl/cfRNA.Suppl.Text.Fig1D.v2.pdf", width=11, height=7, title="Trim vs No trim")</span></span>
<span id="cb5-251"><a href="#cb5-251"></a><span class="in">  cowplot::plot_grid(p1.chrY.trim, labels=c("D"),label_size=27)</span></span>
<span id="cb5-252"><a href="#cb5-252"></a><span class="in">  dev.off()</span></span>
<span id="cb5-253"><a href="#cb5-253"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




</body></html>